releases: add TiDB 5.4.0 release notes (#8061)
pingcap/docs-cn: releases/release-5.4.0.md

| :---------- | :----------- | :----------- |
|  |  |  |
|  |  |  |
|  |  |  |

### 配置文件参数

|  配置文件    |  配置项    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- | :----------- |
|  |  |  |
|  |  |  |
|  |  |  |

### 其他

-
-
-
-

## 新功能

### SQL

- **功能 1**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **功能 2**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 事务

- **功能 3**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **功能 4**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 安全

- **功能 5**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 性能

- **功能 6**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

> 

Reviewer: - **Raft Engine (experimental feature)**

Support using [Raft Engine](https://github.com/tikv/raft-engine) as the Raft Log storage of TiKV. Compared to deployment using RocksDB, Raft Engine can reduce up to 40% write I/O throughput and 10% CPU usage of TiKV server. and brings around 5% TPS improvement and 20% tail latency reduction for certain workloads.

As it involves data format changes, Raft Engine is currently an experimental feature and disabled by default. Be aware that the latest Raft Engine is not compatible with Raft Engine older than v5.4.0. Therefore you should make sure Raft Engine is disabled on any existing TiKV nodes before performing upgrade or downgrade crossing v5.4.0.

[User document](./tikv-configuration-file.md#raft-engine)

Reviewer: 支持使用 Raft Engine 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。

由于 Raft Engine 涉及数据格式改动，目前仍属于实验特性，并默认关闭。同时请注意最新的 Raft Engine 不与 v5.4.0 版本前的 Raft Engine 兼容。因此在进行跨越 v5.4.0 版本的升级和降级之前，需要确保已有 TiKV 节点上的 Raft Engine 已被关闭。

### 稳定性

- **功能 7**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

## 高可用和容灾

- **功能 8**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

> 

Reviewer 2: ```suggestion
- **备份影响优化**

   BR 增加了备份线程自动调节功能。该功能通过监控集群资源的使用率自动调节备份的线程数的手段，降低备份过程对集群的影响。 在某些 Case 验证中，通过增加集群用于备份的资源和开启备份线程自动调节功能，备份影响可以降低到 10% 以下。

   该功能默认开启，但是如果你是在离线环境中进行备份，可以关闭该功能来获得更高的备份速度。

   详细文档请阅读 [BR 自动调节](/br/br-features.md#自动调节-从-v54-版本开始引入)。

- ** 支持 Azure Blob Storage 作为备份目标存储 **

   BR 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

   该功能目前是实验特性，

   详细情况阅读 [BR 支持 Azure Blob Storage 远端存储](/br/backup-and-restore-azblob.md )
```

### 数据迁移

- **功能 9**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

> 

Reviewer 3: ```suggestion
**[Lightning] 增加已存在数据表是否允许导入的开关**

增加 `incremental-import ` 开关。默认`false`表明目标表已存在数据时将不会执行导入，改为`true`则继续导入。
注意，当使用并行导入特性时，此项必须为`true`

**[Lightning] 重复数据检测**

在`backend=local`模式下。数据导入完成之前会输出冲突数据，然后从数据库中删除这些冲突数据。用户可以在导入完成后解析冲突数据，并根据业务规则选择适合的数据进行插入。
建议根据冲突数据清洗上游数据源，避免在后续增量数据迁移阶段遇到冲突数据而造成数据不一致。

**[DM] relay log 使用方式优化**

- 恢复 source 配置中 enable-relay 开关
- 增加 start-relay/stop-relay 命令动态开启/关闭 relay log
- relay log 的开启与 source 绑定， source 迁移到任意 worker 均保持原有开启/关闭状态
- relay log 的存放路径移至 worker 配置文件

**[DM] collation 处理方式优化**

增加`collation_compatible`开关，支持`strict`和`loose`两种模式。默认`loose`模式适用对 collation 要求不严格，不一致亦可正常同步的场景；`strict`模式则适用对 collation 要求严格，不一致则报错的场景。

**[DM] transfer source 平滑执行**

当 worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 source 迁移到其他节点。优化后的 `transfer source` 简化了使用者操作步骤，不再要求先暂停所有关联 task 而是直接执行，DM 将在内部完成所需操作。

```

### 问题诊断效率

> - **功能 10**

Reviewer 4: ** 持续性能分析 **
支持 TiFlash 组件查看 CPU Profiling
支持火焰图形式查看 CPU Profiling 和 Goroutine 结果。
支持在 TiDB Operator 部署环境下启用持续性能分析功能。

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### TiDB 数据共享订阅

- **功能 11**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 部署及运维

> - **功能 12**

Reviewer 4: ** 持续性能分析（实验特性） **
支持更多组件：支持 TiFlash 组件查看 CPU Profiling。
支持更方便的查看形式：支持火焰图形式查看 CPU Profiling 和 Goroutine 结果。
支持更多部署环境：支持在 TiDB Operator 部署环境下启用持续性能分析功能。

该功能默认关闭，需进入 TiDB Dashboard 持续性能分析页面开启。

持续性能分析功能必须使用 TiUP v1.9.0 及以上版本，TiDB Operator vx.x.x 及以上版本升级或安装的集群才可使用。

---
TiDB 5.4 Release Notes
---

# TiDB 5.4 Release Notes

发版日期：2022 年 x 月 x 日

TiDB 版本：5.4.0

在 v5.4.0 版本中，你可以获得以下关键特性：

+
+
+

## 兼容性变化

> **注意：**
>
> 当从一个早期的 TiDB 版本升级到 TiDB v5.4.0 时，如需了解所有中间版本对应的兼容性更改说明，请查看对应版本的 [Release Notes](/releases/release-notes.md)。

### 系统变量

|  变量名    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- |
|  `tidb_backoff_lock_fast` | 修改 | 默认值由 `100` 修改为 `10` |

> |  |  |  |

Reviewer 5: ```suggestion
| `tidb_enable_paging`  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用 paging 方式发送 coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。在这种情况下，由于 `Limit` 算子限制只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，降低延迟，减少资源消耗。 |
```

- `character_set_client` 在处理 `prepare` 语句时可能出现兼容性问题。
- TiDB 不支持 `_gbk"xxx"` 的用法，但是支持 `_utf8mb4"xxx"` 的用法。而 MySQL 对于 `_charset"xxx"` 的用法都支持。
- 从 TiDB v4.0（不包含 v4.0.0）升级到 v5.4.0 及更新的集群，`tidb_enable_index_merge` 变量默认关闭。

## 新功能

### SQL

- **功能 1**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **功能 2**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 事务

- **功能 3**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **功能 4**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 安全

- **功能 5**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 性能

- **功能 6**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

> 

Reviewer 6: - TiFlash
  - 新增支持更多函数下推至 MPP 引擎
      - 字符串函数：LPAD(), RPAD(), STRCMP()
      - 日期时间函数：ADDDATE(), DATE_ADD(), DATE_SUB(), SUBDATE(), QUARTER()
  - 引入动态线程池，提升资源利用率
  - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性

- **在 DM 中新增 `transfer source` 支持平滑执行同步任务**

    当 DM-worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 `source` 配置迁移到其他节点。优化后的 `transfer source` 简化了用户操作步骤，不再要求先暂停所有关联 task 而是直接执行平滑迁移，DM 将在内部完成所需操作。

### 问题诊断效率

- **功能 10**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### TiDB 数据共享订阅

- **功能 11**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 部署及运维

- **功能 12**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **持续性能分析（实验特性）**

    - 支持更多组件：支持 TiFlash 组件查看 CPU Profiling
    - 支持更方便的查看形式：支持以火焰图形式查看 CPU Profiling 和 Goroutine 结果。
    - 支持更多部署环境：支持在 TiDB Operator 部署环境下启用持续性能分析功能。

    该功能默认关闭，需进入 TiDB Dashboard 持续性能分析页面开启，开启方法见[用户文档](/dashboard/continuous-profiling.md)。

    要使用持续性能分析功能，集群须由 TiUP v1.9.0 及以上版本或 TiDB Operator vx.x.x（TBD）及以上版本升级或安装。

## 遥测

- **功能 13**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

## 提升改进

> 

Reviewer 5: - TiDB
    - 新增系统变量 `tidb_enable_paging`，开启该功能可显著降低使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 数据较小且无法下推到 `IndexScan` 上的读请求的延迟 https://github.com/pingcap/tidb/issues/30578

- `character_set_client` 在处理 `prepare` 语句时可能出现兼容性问题。
- TiDB 不支持 `_gbk"xxx"` 的用法，但是支持 `_utf8mb4"xxx"` 的用法。而 MySQL 对于 `_charset"xxx"` 的用法都支持。
- 从 TiDB v4.0（不包含 v4.0.0）升级到 v5.4.0 及更新的集群，`tidb_enable_index_merge` 变量默认关闭。

## 新功能

### SQL

- **功能 1**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **功能 2**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 事务

- **功能 3**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **功能 4**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 安全

- **功能 5**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 性能

- **功能 6**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **新增 Raft Engine（实验特性）**

>     支持使用 Raft Engine 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。
> 
>     由于 Raft Engine 涉及数据格式改动，目前仍属于实验特性，并默认关闭。同时请注意最新的 Raft Engine 不与 v5.4.0 版本前的 Raft Engine 兼容。因此在进行跨越 v5.4.0 版本的升级和降级之前，需要确保已有 TiKV 节点上的 Raft Engine 已被关闭。

Reviewer: ```suggestion
    支持使用 [Raft Engine](https://github.com/tikv/raft-engine) 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。

    由于 Raft Engine 涉及数据格式改动，目前仍属于实验特性，并默认关闭。同时请注意最新的 Raft Engine 不与 v5.4.0 版本前的 Raft Engine 兼容。因此在进行跨越 v5.4.0 版本的升级和降级之前，需要确保已有 TiKV 节点上的 Raft Engine 已被关闭。

    [用户文档](./tikv-configuration-file.md#raft-engine)
```

- TiFlash

  - 支持将更多函数下推至 MPP 引擎
      - 字符串函数：`LPAD()`、`RPAD()`、`STRCMP()`
      - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
  - 引入动态线程池，提升资源利用率
  - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性

> 

Reviewer 7: ```suggestion
- TiCDC
  - 大幅降低在 TiCDC 启用时，对上游 TiDB 集群的性能影响。
```

### 问题诊断效率

- **功能 10**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### TiDB 数据共享订阅

- **功能 11**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

### 部署及运维

- **功能 12**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **持续性能分析（实验特性）**

    - 支持更多组件：支持 TiFlash 组件查看 CPU Profiling
    - 支持更方便的查看形式：支持以火焰图形式查看 CPU Profiling 和 Goroutine 结果。
    - 支持更多部署环境：支持在 TiDB Operator 部署环境下启用持续性能分析功能。

    该功能默认关闭，需进入 TiDB Dashboard 持续性能分析页面开启，开启方法见[用户文档](/dashboard/continuous-profiling.md)。

    要使用持续性能分析功能，集群须由 TiUP v1.9.0 及以上版本或 TiDB Operator vx.x.x（TBD）及以上版本升级或安装。

## 遥测

- **功能 13**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

## 提升改进

+ TiDB

    - 新增系统变量 `tidb_enable_paging`，开启该功能可显著降低使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 数据较小且无法下推到 `IndexScan` 上的读请求的延迟 [#30578](https://github.com/pingcap/tidb/issues/30578)

> 

Reviewer 6: - TiFlash
  - 新增支持更多函数下推至 MPP 引擎
      - 字符串函数：LPAD(), RPAD(), STRCMP()
      - 日期时间函数：ADDDATE(), DATE_ADD(), DATE_SUB(), SUBDATE(), QUARTER()
  - 引入动态线程池，提升资源利用率
  - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性

Reviewer 8: - Index Merge GA：提供了对诸如 
```sql
 SELECT * FROM table WHERE key1 <= 100 OR key2 = 200; 
 ```
（DNF:X~1~ ⋁ X~2~ ⋁ …X~n~） 过滤条件为多列组合（**OR** 连接）的查询的合并索引访问优化。（目前仅支持 **OR**，后继开发将继续支持 CNF:  Y~1~ ⋀ Y~2~ ⋀ …Y~n~ 查询条件组合）
- 提升由行存到列存数据同步处理时对 raft log 的解码（decoding）效率， 此环节的CPU使用率最多可降低 90%
- 实现了一个内置的文件读写压力测试工具： Stress test framework for DeltaTree，以方便模拟各种不同的工作负载
- Dashboard里的“continuous profiling“ 正式支持对 TiFlash 的监控
- TiFlash 调整了文件系统中 Delta Tree 相关的默认参数，使之更适合一般的生产环境

Reviewer 8: (DNF:X<sub>1</sub> ⋁ X<sub>2</sub> ⋁ …X<sub>n</sub>)
CNF:  Y<sub>1</sub> ⋀ Y<sub>2</sub> ⋀ …Y<sub>3</sub>
:joy: 不好意思，格式没注意，弄坏了，请麻烦用这两个替换一下上面的相应内容（DNF和CNF那里看起来比较怪的地方

Reviewer 8: > * Index Merge GA：提供了对诸如
> 
> ```sql
>  SELECT * FROM table WHERE key1 <= 100 OR key2 = 200; 
> ```
> 
> (DNF:X1 ⋁ X2 ⋁ …Xn) 过滤条件为多列组合（**OR** 连接）的查询的合并索引访问优化。（目前仅支持 **OR**，后继开发将继续支持 CNF: Y1 ⋀ Y2 ⋀ …Y3 形式的查询条件组合）
> 
> * 提升由行存到列存数据同步处理时对 raft log 的解码（decoding）效率， 此环节的CPU使用率最多可降低 90%
> * 实现了一个内置的文件读写压力测试工具： Stress test framework for DeltaTree，以方便模拟各种不同的工作负载
> * Dashboard里的“continuous profiling“ 正式支持对 TiFlash 的监控
> * TiFlash 调整了文件系统中 Delta Tree 相关的默认参数，使之更适合一般的生产环境

这一段我挪个地方。另外 @LiSong0214 在前面提到的一句重复了，可以删除。

### 数据迁移

- **功能 9**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

- **在 TiDB Lightning 中添加重复数据的检测**

    在 `backend=local` 模式下，数据导入完成之前 TiDB Lightning 会输出冲突数据，然后从数据库中删除这些冲突数据。用户可以在导入完成后解析冲突数据，并根据业务规则选择适合的数据进行插入。建议根据冲突数据清洗上游数据源，避免在后续增量数据迁移阶段遇到冲突数据而造成数据不一致。

- **在 TiDB Data Migration (DM) 中 优化 relay log 的使用方式**

    - 恢复 `source` 配置中 `enable-relay` 开关
    - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能
    - relay log 的开启状态与 `source` 绑定，source 迁移到任意 DM-worker 均保持原有开启或关闭状态
    - relay log 的存放路径移至 DM-worker 配置文件

- **在 DM 中优化排序规则的处理方式**

    增加 `collation_compatible` 开关，支持 `strict` 和 `loose`（默认）两种模式。如果对排序规则要求不严格，允许排序规则不一致，使用默认的 `loose` 模式可使同步正常进行；如果对排序规则要求严格，排序规则不一致导致报错，则可以使用 `strict` 模式。

- **在 DM 中新增 `transfer source` 支持平滑执行同步任务**

    当 DM-worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 `source` 配置迁移到其他节点。优化后的 `transfer source` 简化了用户操作步骤，不再要求先暂停所有关联 task 而是直接执行平滑迁移，DM 将在内部完成所需操作。

> 

Reviewer 9: ```suggestion
- **优化 TiCDC 对集群的影响**

    大幅降低了 TiCDC 启用后，对 TiDB 集群的性能影响。在实验室环境中，TiCDC 对 TiDB 的性能影响可以降低到 5% 以下。
```

| `tidb_read_staleness` | 新增 |  |
| `tidb_stats_load_sync_wait` | 新增 |  |
| `tidb_stats_load_pseudo_timeout` | 新增 |  |

### 配置文件参数

|  配置文件    |  配置项    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- | :----------- |
| TiDB | `stats-load-concurrency` | 新增 |               |
| TiDB | `stats-load-queue-size`   | 新增 |               |
| TiKV | `backup.enable-auto-tune` | 修改 | 在 v5.3.0 中默认值为 `false`，自 v5.4.0 起默认值改为 `true`。表示在集群资源占用率较高的情况下，是否允许 BR 自动限制备份使用的资源以求减少对集群的影响。在默认配置下，备份速度可能下降。 |
| TiKV | `log.level`、`log.format`、`log.enable-timestamp`、`log.file.filename`、`log.file.max-size`、`log.file.max-days`、`log.file.max-backups` | 新增  | 参数说明见[统一各组件的日志格式和日志归档轮转规则](#统一各组件的日志格式和日志归档轮转规则)。 |
| TiKV | `log-level`、`log-format`、`log-file`、`log-rotation-size` | 变更 | 将 TiKV log 参数名改为与 TiDB log 参数一致的命名方式，即 `log.level`、`log.format`、`log.enable-timestamp`。如果 TiKV log 参数为非默认值则保持兼容；如果同时配置 TiKV log 参数和 TiDB log 命名方式的参数，使用 TiDB log 命名方式的参数。详情参见[统一各组件的日志格式和日志归档轮转规则](#统一各组件的日志格式和日志归档轮转规则)。 |

> | TiKV  |  `log-rotation-timespan`  | 删除 |  轮换日志的时间跨度。当超过该时间跨度，日志文件会被轮换，即在当前日志文件的文件名后附加一个时间戳，并创建一个新文件。 |

Reviewer 10: ```suggestion
| TiKV  |  `log-rotation-timespan`  | 删除 |  轮换日志的时间跨度。当超过该时间跨度，日志文件会被轮换，即在当前日志文件的文件名后附加一个时间戳，并创建一个新文件。 |
| TiKV | `raft_store.apply_batch_system.pool_size` | 新增 | apply 线程池大小，替代 `raftstore.apply-pool-size`。 |
| TiKV | `raft_store.store_batch_system.pool_size` | 新增 | store 线程池大小，替代 `raftstore.store-pool-size`。 |
| TiKV | `snap-generator-pool-size` | 新增 | snap-generator 线程池大小 |
```

max-days = 0
设置日志最大保留天数，默认全保留不清理。

max-backups = 0
设置日志备份的最大保留文件数，默认全保留不清理。
```

### 其他

- TiDB Dashboard 默认不再使用 `root` + 空密码登录。

    从 v5.4.0开始，使用 TiUP 启动集群时推荐使用 `start --initial`。执行该操作启动集群后，会为 `root` 账号自动生成一个随机密码，`root` 账号登录 Dashboard 需要使用这个密码。

- 为 TiDB 和 PD 之间新增接口。使用 `information_schema.TIDB_HOT_REGIONS_HISTORY` 系统表时，TiDB 需要使用匹配的 PD 版本。

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

    在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。为了更好的支持中文用户，TiDB 从 v5.4.0 版本开始支持 GBK 字符集，同时支持 `gbk_bin` 和 `gbk_chinese_ci` 两种排序规则。

    在使用 GBK 字符集时，请注意以下兼容性限制：

    - TiFlash 暂不支持 GBK 字符。
    - TiCDC 暂不支持 GBK 字符。
    - 如果 `character_set_client` 和 `character_set_connection` 都是 `gbk` 时，处理非法 GBK 字符与 MySQL 存在兼容性问题。
    - `character_set_client` 在处理 `prepare` 语句时可能出现兼容性问题。
    - TiDB 不支持 `_gbk"xxx"` 的用法，但是支持 `_utf8mb4"xxx"` 的用法。而 MySQL 对于 `_charset"xxx"` 的用法都支持。
    - TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.3.0 之前不支持恢复 `charset=GBK` 的表。

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

> 

Reviewer 6: - **TiFlash Index Merge GA**
提供了对诸如
 ```SELECT * FROM table WHERE key1 <= 100 OR key2 = 200; ```

   (DNF:X1 ⋁ X2 ⋁ …Xn) 过滤条件为多列组合（OR 连接）的查询的合并索引访问优化。（目前仅支持 OR，后继开发将继续支持 CNF: Y1 ⋀ Y2 ⋀ …Y3 形式的查询条件组合）

Reviewer 8: 这个放这里不太合适，第一句标题内容有误。我重新写到前面 _新功能_ 那边。

Reviewer 8: 这一段看上去是 **安全** 相关的内容

### 性能

- **新增 Raft Engine（实验特性）**

    支持使用 [Raft Engine](https://github.com/tikv/raft-engine) 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。

    由于 Raft Engine 涉及数据格式改动，目前仍属于实验特性，并默认关闭。同时请注意最新的 Raft Engine 不与 v5.4.0 版本前的 Raft Engine 兼容。因此在进行跨越 v5.4.0 版本的升级和降级之前，需要确保已有 TiKV 节点上的 Raft Engine 已被关闭。

    [用户文档](/tikv-configuration-file.md#raft-engine)

- TiFlash

  - 支持将更多函数下推至 MPP 引擎
      - 字符串函数：`LPAD()`、`RPAD()`、`STRCMP()`
      - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
  - 引入动态线程池，提升资源利用率
  - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性

> 

Reviewer 6: - 提升由行存到列存数据同步处理时对 raft log 的解码（decoding）效率， 此环节的CPU使用率最多可降低 90%
- TiFlash 调整了文件系统中 Delta Tree 相关的默认参数，使之更适合一般的生产环境

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

- **在 TiDB Lightning 中添加重复数据的检测**

    在 `backend=local` 模式下，数据导入完成之前 TiDB Lightning 会输出冲突数据，然后从数据库中删除这些冲突数据。用户可以在导入完成后解析冲突数据，并根据业务规则选择适合的数据进行插入。建议根据冲突数据清洗上游数据源，避免在后续增量数据迁移阶段遇到冲突数据而造成数据不一致。

- **在 TiDB Data Migration (DM) 中 优化 relay log 的使用方式**

    - 恢复 `source` 配置中 `enable-relay` 开关
    - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能
    - relay log 的开启状态与 `source` 绑定，source 迁移到任意 DM-worker 均保持原有开启或关闭状态
    - relay log 的存放路径移至 DM-worker 配置文件

- **在 DM 中优化排序规则的处理方式**

    增加 `collation_compatible` 开关，支持 `strict` 和 `loose`（默认）两种模式。如果对排序规则要求不严格，允许排序规则不一致，使用默认的 `loose` 模式可使同步正常进行；如果对排序规则要求严格，排序规则不一致导致报错，则可以使用 `strict` 模式。

- **在 DM 中 优化 `transfer source`，支持平滑执行同步任务**

    当 DM-worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 `source` 配置迁移到其他节点。优化后的 `transfer source` 简化了用户操作步骤，不再要求先暂停所有关联 task 而是直接执行平滑迁移，DM 将在内部完成所需操作。

- **DM OpenAPI 特性 GA**

    DM 支持通过 API 的方式进行日常管理，包括增加数据源、管理任务等。本次更新 OpenAPI 从实验特性转为正式特性。

### 问题诊断效率

- **Top SQL（实验特性）**

    新推出实验性特性 Top SQL（默认关闭），帮助用户轻松找到节点中负载贡献较大的查询。

    [用户文档](/dashboard/top-sql.md)

> 

Reviewer 6: -  实现了一个内置的文件读写压力测试工具： Stress test framework for DeltaTree，以方便模拟各种不同的工作负载

### TiDB 数据共享订阅

- **优化 TiCDC 对集群的影响**

    大幅降低了 TiCDC 启用后，对 TiDB 集群的性能影响。在实验室环境中，TiCDC 对 TiDB 的性能影响可以降低到 5% 以下。

### 部署及运维

- **持续性能分析（实验特性）**

    - 支持更多组件：支持 TiFlash 组件查看 CPU Profiling
    - 支持更方便的查看形式：支持以火焰图形式查看 CPU Profiling 和 Goroutine 结果。
    - 支持更多部署环境：支持在 TiDB Operator 部署环境下启用持续性能分析功能。

    该功能默认关闭，需进入 TiDB Dashboard 持续性能分析页面开启，开启方法见[用户文档](/dashboard/continuous-profiling.md)。

    要使用持续性能分析功能，集群须由 TiUP v1.9.0 及以上版本或 TiDB Operator vx.x.x（TBD）及以上版本升级或安装。

## 提升改进

+ TiDB

    - 新增系统变量 `tidb_enable_paging`，开启该功能可显著降低使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 数据较小且无法下推到 `IndexScan` 上的读请求的延迟 [#30578](https://github.com/pingcap/tidb/issues/30578)

> + TiKV

Reviewer 10: ```suggestion
+ TiKV
  + coprocessor 支持分页 API 进行流式处理。
  + 支持 read-through-lock，使读操作不需要等待清理 secondary lock。
  + 增加了磁盘保护机制，尽量避免磁盘空间耗尽导致 panic
  + 日志支持存档和轮替
  + raft 客户端减少系统调用并提高 CPU 效率 
  + coprocessor 支持下推 substring 到 tikv 
  + 通过跳过读锁提高在 RC 隔离级别中扫描的性能
  + 减少备份使用的默认线程池大小，并在压力大时限制其使用 
  + 支持动态调整 apply 和 store 线程池大小
  + 支持配置 snap-generator 线程池大小
```

+ PD

+ TiFlash

+ Tools

    + Backup & Restore (BR)

    + TiCDC

    + TiDB Data Migration (DM)

    + TiDB Lightning

    + Dumpling

    + TiDB Binlog

## Bug 修复

+ TiDB

> + TiKV

Reviewer 10: ```suggestion
+ TiKV
  + 修复 mvcc 删除记录可能不会被 GC 删除的问题
  + 修复悲观事务中 prewrite 请求重试在极少数情况下影响数据一致性的风险。 
  + 修复 GC 扫描导致的 OOM 
  + 修复当达到磁盘容量满时 RocksDB flush 或 compaction 导致的 panic 
```

+ 为 TiDB Lightning 增加已存在数据表是否允许导入的开关
+ 优化持续性能分析实验特性
+ TiSpark 支持用户认证与鉴权

## 兼容性变化

> **注意：**
>
> 当从一个早期的 TiDB 版本升级到 TiDB v5.4.0 时，如需了解所有中间版本对应的兼容性更改说明，请查看对应版本的 [Release Notes](/releases/release-notes.md)。

### 系统变量

|  变量名    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- |
|  `tidb_backoff_lock_fast` | 修改 | 默认值由 `100` 修改为 `10` |
| `tidb_enable_index_merge` | 修改 | 默认值由 `OFF` 改为 `ON`。如果从 TiDB v4.0（不包含 v4.0.0）升级到 v5.4.0 及更新的集群，该变量值默认保持 `OFF`。 |
| `tidb_enable_paging`  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用 paging 方式发送 coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。在这种情况下，由于 `Limit` 算子限制只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，降低延迟，减少资源消耗。 |
| `tidb_read_staleness` | 新增 | 用于设置当前会话期待读取的历史数据的所处时刻，默认值为 `0` |
| [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `0` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时。 |
| [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败(`false`) 还是退回使用 pseudo 的统计信息 (`true`) |

> | [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 这个变量用于控制是否持久化执行 `ANALYZE` 语句时指定的配置项 |

Reviewer 11: ```suggestion
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 这个变量用于控制是否开启[ ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性 |
```

---
title: TiDB 5.4 Release Notes
---

# TiDB 5.4 Release Notes

发版日期：2022 年 x 月 x 日

TiDB 版本：5.4.0

在 v5.4.0 版本中，你可以获得以下关键特性：

+ 支持 GBK 字符集
+ 支持索引合并 (Index Merge) 数据访问方法，能够合并多个列上索引的条件过滤结果
+ 支持通过 session 变量实现有界限过期数据读取
+ 支持统计信息采集配置持久化
+ 支持使用 Raft Engine 作为 TiKV 的日志存储引擎
+ 优化备份对集群的影响
+ 支持 Azure Blob Storage 作为备份目标存储（实验特性）
+ 持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能
+ 为 TiDB Lightning 增加已存在数据表是否允许导入的开关
+ 优化持续性能分析实验特性
+ TiSpark 支持用户认证与鉴权

## 兼容性变化

> **注意：**
>
> 当从一个早期的 TiDB 版本升级到 TiDB v5.4.0 时，如需了解所有中间版本对应的兼容性更改说明，请查看对应版本的 [Release Notes](/releases/release-notes.md)。

### 系统变量

|  变量名    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- |

> |  `tidb_backoff_lock_fast` | 修改 | 默认值由 `100` 修改为 `10` |

Reviewer 11: ```suggestion
|  `tidb_backoff_lock_fast` | 修改 | 默认值由 `100` 修改为 `10` |
|  [`tidb_enable_column_tracking`](/system-variables.md#tidb_enable_column_tracking-从-v540-版本开始引入) | 新增 | 用于控制是否开启 TiDB 对 `PREDICATE COLUMNS` 的收集，默认值为 `OFF` |
```

    从 v5.4.0 版本开始 TiDB 支持通过 session 变量设置有界限过期读，进一步提升易用性，具体设置如下：

    ```sql
    set @@tidb_replica_read=leader_and_follower
    set @@tidb_read_staleness="-5"
    ```

    通过该设置，可以实现就近选取 leader 或 follower 节点，并读取 5 秒钟前的最新过期数据，满足准实时场景下低延迟高吞吐数据访问的业务诉求，降低研发门槛，提升易用性。

- **TiDB 正式发布索引合并 (Index Merge) 功能**

    [索引合并](/explain-index-merge.md) 是在 TiDB v4.0 版本中作为实验特性引入的一种查询执行方式的优化，可以大幅提高查询在扫描多列数据时条件过滤的效率。例如对以下的查询，若 `WHERE` 子句中两个 `OR` 连接的过滤条件在各自包含的 _key1_ 与 _key2_ 两个列上都存在索引，则 _索引合并_ 可以同时利用  _key1_ 与 _key2_ 上的索引分别进行过滤，然后合并出最终的结果。

    ```sql
    SELECT * FROM table WHERE key1 <= 100 OR key2 = 200;
    ```

    以往 TiDB 在一个表上的查询只能使用一个索引，无法同时使用多个索引进行条件过滤。相较以往，_索引合并_ 避免了此情况下可能不必要的大量数据扫描，也可以使得需要灵活查询不特定多列数据组合的用户利用单列上的索引达到高效稳定的查询，无需大量构建多列复合索引。

    本版本正式发布了 _索引合并 (Index Merge)_ 特性，但仍存在以下的使用条件和限制（详情请参见[用户文档](/explain-index-merge.md)）：

    目前 TiDB 的 _索引合并_ 优化只限于 _析取范式_ (X<sub>1</sub> ⋁ X<sub>2</sub> ⋁ …X<sub>n</sub>)，即 `WHERE` 子句中过滤条件连接词为 `OR`。

    如果全新部署的集群版本为 v5.4.0 或以上，此特性默认开启。如果从 v5.4.0 以前的版本升级到 v5.4.0 或以上，默认保持升级前此特性的开关状态（v4.0.0 之前无此项特性的版本默认关闭），由用户决定是否开启。

> 

Reviewer 11: ```suggestion

- **支持收集部分列的统计信息（实验特性）**
    
    执行 SQL 语句时，优化器在大多数情况下只会用到部分列（例如， `WHERE`、`JOIN`、`ORDER BY`、`GROUP BY` 子句中用到的列）的统计信息。这些被优化器用到的列称为 `PREDICATE COLUMNS`。
    
    从 v5.4.0 开始，TiDB 引入了收集部分列的统计信息的特性（默认关闭），支持只收集定列或者 `PREDICATE COLUMNS` 的统计信息，极大地降低了收集统计信息的开销。
    
    [用户文档](/statistics.md#收集部分列的统计信息)
    
```

- **支持统计信息的同步加载（实验特性）**

>     从 v5.4.0 开始，TiDB 引入了统计信息同步加载的特性，支持执行当前 SQL 语句时将直方图、TopN、CMSketch 等占用空间较大的统计信息同步加载到内存，提高该 SQL 语句优化时统计信息的完整性。

Reviewer 11: ```suggestion
    从 v5.4.0 开始，TiDB 引入了统计信息同步加载的特性（默认关闭），支持执行当前 SQL 语句时将直方图、TopN、CMSketch 等占用空间较大的统计信息同步加载到内存，提高该 SQL 语句优化时统计信息的完整性。
    
    [用户文档](/statistics.md#统计信息的加载)
```

    - 开启采集配置项持久化

        设置 `tidb_analyze_version = 2` 且 `tidb_persist_analyze_options = true` 会开启配置项持久化。开启后，手动 analyze 指定的所有配置项会被持久化并覆盖已有选项。后续的手动或自动 analyze 任务会沿用已有配置项进行统计信息采集，直到持久化功能关闭或用户手动指定新的采集配置项。

    - 关闭采集配置项持久化

        设置 `tidb_analyze_version = 1` 或 `tidb_persist_analyze_options = false` 会关闭采集配置项持久化功能。持久化关闭后，已有配置项不会被删除，但不记录新增配置项。新采集任务不会沿用已有的持久化配置项，再次开启采集配置项持久化将会直接使用已有的配置项进行统计信息采集。如果需要更新已有的配置项，请手动执行 `analyze` 命令并指定新的采集配置项。

## 高可用和容灾

- **优化备份对集群的影响**

    Backup & Restore (BR) 增加了备份线程自动调节功能。该功能通过监控集群资源的使用率自动调节备份的线程数的方式，降低备份过程对集群的影响。在某些 Case 验证中，通过增加集群用于备份的资源和开启备份线程自动调节功能，备份的影响可以降低到 10% 以下。

    该功能默认开启，但是如果你在离线环境中进行备份，可以关闭该功能来获得更高的备份速度。

    详细文档请阅读 [BR 自动调节](/br/br-features.md#自动调节-从-v54-版本开始引入)。

- **支持 Azure Blob Storage 作为备份目标存储（实验特性）**

    Backup & Restore (BR) 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

    该功能目前是实验特性，详细情况参考 [BR 支持 Azure Blob Storage 远端存储](/br/backup-and-restore-azblob.md)。

### 数据迁移

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

> - **在 TiDB Lightning 中添加重复数据的检测**

Reviewer 12: ```suggestion
- **在 TiDB Lightning 中添加重复数据的检测（实验特性）**
```

---
title: TiDB 5.4 Release Notes
---

# TiDB 5.4 Release Notes

发版日期：2022 年 x 月 x 日

TiDB 版本：5.4.0

在 v5.4.0 版本中，你可以获得以下关键特性：

+ 支持 GBK 字符集
+ 支持索引合并 (Index Merge) 数据访问方法，能够合并多个列上索引的条件过滤结果
+ 支持通过 session 变量实现有界限过期数据读取
+ 支持统计信息采集配置持久化
+ 支持使用 Raft Engine 作为 TiKV 的日志存储引擎
+ 优化备份对集群的影响
+ 支持 Azure Blob Storage 作为备份目标存储（实验特性）
+ 持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能
+ 为 TiDB Lightning 增加已存在数据表是否允许导入的开关
+ 优化持续性能分析实验特性
+ TiSpark 支持用户认证与鉴权

## 兼容性变化

> **注意：**
>
> 当从一个早期的 TiDB 版本升级到 TiDB v5.4.0 时，如需了解所有中间版本对应的兼容性更改说明，请查看对应版本的 [Release Notes](/releases/release-notes.md)。

### 系统变量

|  变量名    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- |
|  `tidb_backoff_lock_fast` | 修改 | 默认值由 `100` 修改为 `10` |

> | `tidb_enable_index_merge` | 修改 | 默认值由 `OFF` 改为 `ON`。如果从 TiDB v4.0（不包含 v4.0.0）升级到 v5.4.0 及更新的集群，该变量值默认保持 `OFF`。 |

Reviewer 13: ```suggestion
| `tidb_enable_index_merge` | 修改 | 默认值由 `OFF` 改为 `ON`。如果从低于 v4.0.0 版本升级到 v5.4.0 及以上版本的集群，该变量值默认保持 `OFF`。如果从 v4.0.0 及以上版本升级到 v5.4.0 及以上版本的集群时，该变量开关保持升级前的状态。对于 v5.4.0 及以上版本的新建集群，该变量开关默认保持 `ON`。 |
```

| `tidb_enable_paging`  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用 paging 方式发送 coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。在这种情况下，由于 `Limit` 算子限制只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，降低延迟，减少资源消耗。 |
| `tidb_read_staleness` | 新增 | 用于设置当前会话期待读取的历史数据的所处时刻，默认值为 `0` |
| [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `0` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时。 |
| [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败(`false`) 还是退回使用 pseudo 的统计信息 (`true`) |
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启[ ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性 |

> 

Reviewer 14: | `tidb_store_limit` | 修改 | 之前支持实例级别及集群级别的设置，现在只支持集群级别的设置 |

Author: Added

| TiKV | `log.level`、`log.format`、`log.enable-timestamp`、`log.file.filename`、`log.file.max-size`、`log.file.max-days`、`log.file.max-backups` | 新增  | 参数说明见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。 |
| TiKV | `log-level`、`log-format`、`log-file`、`log-rotation-size` | 变更 | 将 TiKV log 参数名改为与 TiDB log 参数一致的命名方式，即 `log.level`、`log.format`、`log.enable-timestamp`。如果 TiKV log 参数为非默认值则保持兼容；如果同时配置 TiKV log 参数和 TiDB log 命名方式的参数，使用 TiDB log 命名方式的参数。详情参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。 |
| TiKV  |  `log-rotation-timespan`  | 删除 |  轮换日志的时间跨度。当超过该时间跨度，日志文件会被轮换，即在当前日志文件的文件名后附加一个时间戳，并创建一个新文件。 |
| TiKV | `raft-engine` | 新增 | 包含 `enable`、`dir`、`batch-compression-threshold`、`bytes-per-sync`、`target-file-size`、`purge-threshold`、`recovery-mode`、`recovery-read-block-size`、`recovery-read-block-size`、`recovery-threads`，详情参见 [TiKV 配置文件：raft-engine](/tikv-configuration-file.md#raft-engine)。|
| TiKV | `snap-generator-pool-size` | 新增 | `snap-generator` 线程池大小，默认值为 `2` |
| TiKV | `raftstore.raft-log-gc-tick-interval` | 修改 | 默认值修改为 `3s` |

> | TiKV | `raftstore.raft-log-compact-sync-interval` | 新增 | 控制 CompactLog 命令的间隔，默认值为 `2s` |

Reviewer 10: ```suggestion
| TiKV | `raftstore.raft-log-compact-sync-interval` | 新增 | 控制 CompactLog 命令的间隔，默认值为 `2s` |
| TiKV | `backup.num-threads` | 变更 | 默认值修改为 `min(CPU_NUM * 0.5, 8)` |
| TiKV | `backup.auto-tune-remain-threads` | 变更 | 默认值修改为 `CPU_NUM * 0.5` |
| TiKV | `backup.io-thread-size` | 新增 | backup 使用的 I/O 线程池大小，默认值为 `2` |
```

Author: 和 BR 的 PM 确认，这两个配置项无需添加

| PD | `log.level` | 修改 | 默认值由 "INFO" 改为 "info"，保证大小写不敏感 |
| PD | `hot-regions-write-interval` | 新增 |  设置 PD 存储 Hot Region 信息的时间间隔。默认值为 `10m`。 |
| PD | `hot-regions-reserved-days` | 新增 | 设置 PD 保留的 Hot Region 信息的最长时间。默认值为 `7`。
| TiFlash | `profile.default.enable_elastic_threadpool` | 新增  |  表示是否启用可自动扩展的线程池。打开该配置项可以显著提高 TiFlash 在高并发场景的 CPU 利用率。默认值为 `false`。|
| TiDB Data Migration (DM) | `collation_compatible` | 同步 CREATE 语句中缺省 Collation 的方式，可选 "loose" 和 "strict"，默认为 "loose"。 |

> | TiFlash | `storage.format_version` | 新增可选值 | 表示 DTFile 储存文件格式，默认值为 `2`。|

Reviewer 15: @SchrodingerZhu `storage.format_version` 是本次新增配置项吗？

> | TiFlash | `storage.format_version` | 新增可选值 | 表示 DTFile 储存文件格式，默认值为 `2`。|

Reviewer 16: 之前版本也支持，但并未doc

| TiFlash | `logger.count` | 修改 | 默认值修改为 `10` |
| TiFlash | `status.metrics_port` | 修改 | 默认值修改为 `8234` |
| TiFlash | `raftstore.apply-pool-size` | 新增 | 处理 Raft 数据落盘的线程池中线程的数量，默认值为 `4`。 |
| TiFlash | `raftstore.store-pool-size` | 新增 | 处理 Raft 的线程池中线程的数量，即 Raftstore 线程池的大小，默认值为 `4`。 |
| TiCDC | `enable-tidb-extension` | 新增 |  开启该配置项后，TiCDC 在 Canal-JSON 协议格式中附加 TiDB 扩展字段。                  |
| TiCDC | `max-message-bytes` | 修改 | 将 Kafka sink 模块的 `max-message-bytes` 默认值设置为 `10M`  |
| TiCDC | `partition-num`      | 修改 | 将 Kafka Sink `partition-num` 的默认值改由 `4` 为 `3`，使 TiCDC 更加平均地分发消息到各个 Kafka partition |

> 

Reviewer 12: ```suggestion
| TiDB Lightning | `meta-schema-name` |  新增  | 在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认值为 "lightning_metadata"。 | 
| TiDB Lightning | `incremental-import` | 新增 | 是否允许向已存在数据的表导入数据。默认值为 false。| 
```

Author: Added

### 其他

- TiDB Dashboard 默认不再使用 `root` + 空密码登录。

    从 v5.4.0开始，使用 TiUP 启动集群时推荐使用 `start --initial`。执行该操作启动集群后，会为 `root` 账号自动生成一个随机密码，`root` 账号登录 Dashboard 需要使用这个密码。

- 为 TiDB 和 PD 之间新增接口。使用 `information_schema.TIDB_HOT_REGIONS_HISTORY` 系统表时，TiDB 需要使用匹配的 PD 版本。
- TiDB Server、PD Server 和 TiKV Server 将采用统一的参数命名方式来管理日志命名、输出格式、轮转和过期的规则。参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

>     在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。为了更好的支持中文用户，TiDB 从 v5.4.0 版本开始支持 GBK 字符集，同时支持 `gbk_bin` 和 `gbk_chinese_ci` 两种排序规则。

Reviewer 17: 支持这两种排序规则的前提是开启 TiDB 配置项 `new_collations_enabled_on_first_bootstrap`，需要说明一下？

Author: Done

    在使用 GBK 字符集时，请注意以下兼容性限制：

    - TiFlash 暂不支持 GBK 字符。
    - TiCDC 暂不支持 GBK 字符。
    - 如果 `character_set_client` 和 `character_set_connection` 都是 `gbk` 时，处理非法 GBK 字符与 MySQL 存在兼容性问题。

>     - `character_set_client` 在处理 `prepare` 语句时可能出现兼容性问题。

Reviewer 17: ```suggestion
```

这个兼容性说明我觉得可以删除，这本来是 MySQL 存在的问题，这边没有必要特殊说明。

    - TiDB 不支持 `_gbk"xxx"` 的用法，但是支持 `_utf8mb4"xxx"` 的用法。而 MySQL 对于 `_charset"xxx"` 的用法都支持。
    - TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.3.0 之前不支持恢复 `charset=GBK` 的表。

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

### 性能

- **新增 Raft Engine（实验特性）**

    支持使用 [Raft Engine](https://github.com/tikv/raft-engine) 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。

    由于 Raft Engine 涉及数据格式改动，目前仍属于实验特性，并默认关闭。同时请注意最新的 Raft Engine 不与 v5.4.0 版本前的 Raft Engine 兼容。因此在进行跨越 v5.4.0 版本的升级和降级之前，需要确保已有 TiKV 节点上的 Raft Engine 已被关闭。

    [用户文档](/tikv-configuration-file.md#raft-engine)

- **持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能**

    - 支持将更多函数下推至 MPP 引擎
        - 字符串函数：`LPAD()`、`RPAD()`、`STRCMP()`
        - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
    - 引入动态线程池，提升资源利用率（实验特性）
    - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性

>     - 提升由行存到列存数据同步处理时对 raft log 的解码 (decoding) 效率，此环节的 CPU 使用率最多可降低 90%

Reviewer 13: ```suggestion
    - 提升由行存到列存数据同步处理时对 raft log 的解码 (decoding) 效率，使 CPU 使用率降低达 90%
```

    在对数据实时性要求不高的业务场景下，可以选择过期读模式。使用该模式可以降低延迟和提升吞吐。TiDB 目前支持通过显示只读事务或 SQL 语句的方式实现过期读。两种方式均支持指定时间的精确过期读和指定时间边界的过期读两种模式，详细用法请参考[过期读文档](/read-historical-data.md)。

    从 v5.4.0 版本开始 TiDB 支持通过 session 变量设置有界限过期读，进一步提升易用性，具体设置如下：

    ```sql
    set @@tidb_replica_read=leader_and_follower
    set @@tidb_read_staleness="-5"
    ```

    通过该设置，可以实现就近选取 leader 或 follower 节点，并读取 5 秒钟前的最新过期数据，满足准实时场景下低延迟高吞吐数据访问的业务诉求，降低研发门槛，提升易用性。

- **TiDB 正式发布索引合并 (Index Merge) 功能**

    [索引合并](/explain-index-merge.md) 是在 TiDB v4.0 版本中作为实验特性引入的一种查询执行方式的优化，可以大幅提高查询在扫描多列数据时条件过滤的效率。例如对以下的查询，若 `WHERE` 子句中两个 `OR` 连接的过滤条件在各自包含的 _key1_ 与 _key2_ 两个列上都存在索引，则 _索引合并_ 可以同时利用  _key1_ 与 _key2_ 上的索引分别进行过滤，然后合并出最终的结果。

    ```sql
    SELECT * FROM table WHERE key1 <= 100 OR key2 = 200;
    ```

    以往 TiDB 在一个表上的查询只能使用一个索引，无法同时使用多个索引进行条件过滤。相较以往，_索引合并_ 避免了此情况下可能不必要的大量数据扫描，也可以使得需要灵活查询不特定多列数据组合的用户利用单列上的索引达到高效稳定的查询，无需大量构建多列复合索引。

    本版本正式发布了 _索引合并 (Index Merge)_ 特性，但仍存在以下的使用条件和限制（详情请参见[用户文档](/explain-index-merge.md)）：

    目前 TiDB 的 _索引合并_ 优化只限于 _析取范式_ (X<sub>1</sub> ⋁ X<sub>2</sub> ⋁ …X<sub>n</sub>)，即 `WHERE` 子句中过滤条件连接词为 `OR`。

>     如果全新部署的集群版本为 v5.4.0 或以上，此特性默认开启。如果是从 v5.4.0 以前的版本升级到 v5.4.0 或以上的，默认保持升级前此特性的开闭状态（v4.0 之前无此项特性的版本也视同关闭），需要由用户自己决定是否开启。

Reviewer 13: ```suggestion
    如果全新部署的集群版本为 v5.4.0 或以上，此特性默认开启。如果从 v5.4.0 以前的版本升级到 v5.4.0 或以上，默认保持升级前此特性的开关状态（v4.0.0 之前无此项特性的版本默认关闭），由用户决定是否开启。
```

- **支持统计信息的同步加载（实验特性）**

    从 v5.4.0 开始，TiDB 引入了统计信息同步加载的特性，支持执行当前 SQL 语句时将直方图、TopN、CMSketch 等占用空间较大的统计信息同步加载到内存，提高该 SQL 语句优化时统计信息的完整性。

> 

Reviewer: - **优化 RocksDB 在大数据量下的锁争用问题**

RocksDB 依赖全局锁来维护多版本的文件信息，并在销毁读快照时对过期文件进行清理。在文件数较多且读写频繁的场景下，这一操作显著影响前台延迟。本版本针对这一问题进行优化，详见 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)。

Author: Added

    - 开启采集配置项持久化

        设置 `tidb_analyze_version = 2` 且 `tidb_persist_analyze_options = true` 会开启配置项持久化。开启后，手动 analyze 指定的所有配置项会被持久化并覆盖已有选项。后续的手动或自动 analyze 任务会沿用已有配置项进行统计信息采集，直到持久化功能关闭或用户手动指定新的采集配置项。

    - 关闭采集配置项持久化

        设置 `tidb_analyze_version = 1` 或 `tidb_persist_analyze_options = false` 会关闭采集配置项持久化功能。持久化关闭后，已有配置项不会被删除，但不记录新增配置项。新采集任务不会沿用已有的持久化配置项，再次开启采集配置项持久化将会直接使用已有的配置项进行统计信息采集。如果需要更新已有的配置项，请手动执行 `analyze` 命令并指定新的采集配置项。

## 高可用和容灾

- **优化备份对集群的影响**

    Backup & Restore (BR) 增加了备份线程自动调节功能。该功能通过监控集群资源的使用率自动调节备份的线程数的方式，降低备份过程对集群的影响。在某些 Case 验证中，通过增加集群用于备份的资源和开启备份线程自动调节功能，备份的影响可以降低到 10% 以下。

    该功能默认开启，但是如果你在离线环境中进行备份，可以关闭该功能来获得更高的备份速度。

    详细文档请阅读 [BR 自动调节](/br/br-features.md#自动调节-从-v54-版本开始引入)。

- **支持 Azure Blob Storage 作为备份目标存储（实验特性）**

    Backup & Restore (BR) 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

    该功能目前是实验特性，详细情况参考 [BR 支持 Azure Blob Storage 远端存储](/br/backup-and-restore-azblob.md)。

### 数据迁移

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

> 

Reviewer 12: ```suggestion

- **为 TiDB Lightning 增加并行导入模式下保存各个 TiDB Lightning 实例元信息的 schema 名字的参数 **

    在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认为 "lightning_metadata"。对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。

```

- **在 TiDB Lightning 中添加重复数据的检测**

    在 `backend=local` 模式下，数据导入完成之前 TiDB Lightning 会输出冲突数据，然后从数据库中删除这些冲突数据。用户可以在导入完成后解析冲突数据，并根据业务规则选择适合的数据进行插入。建议根据冲突数据清洗上游数据源，避免在后续增量数据迁移阶段遇到冲突数据而造成数据不一致。

- **在 TiDB Data Migration (DM) 中 优化 relay log 的使用方式**

    - 恢复 `source` 配置中 `enable-relay` 开关
    - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能
    - relay log 的开启状态与 `source` 绑定，source 迁移到任意 DM-worker 均保持原有开启或关闭状态
    - relay log 的存放路径移至 DM-worker 配置文件

- **在 DM 中优化排序规则的处理方式**

>     增加 `collation_compatible` 开关，支持 `strict` 和 `loose`（默认）两种模式。如果对排序规则要求不严格，允许排序规则不一致，使用默认的 `loose` 模式可使同步正常进行；如果对排序规则要求严格，排序规则不一致导致报错，则可以使用 `strict` 模式。

Reviewer 18: ```suggestion
    增加 `collation_compatible` 开关，支持 `strict` 和 `loose`（默认）两种模式。如果对排序规则要求不严格，允许排序规则在上下游不一致，使用默认的 `loose` 模式可使同步正常进行；如果对排序规则要求严格，排序规则在上下游不一致会导致业务报错，则可以使用 `strict` 模式。
```

cc @sunzhaoyang

Reviewer 3: 以田原老师的为准

Reviewer 3: 增加 `collation_compatible` 开关，支持 `strict` 和 `loose`（默认）两种模式。
如果对排序规则要求不严格，允许上下游查询结果排序规则不一致，使用默认的 `loose` 模式比较方便维护；
如果对排序规则要求严格，业务要求排序规则必须一致，则应当使用 `strict` 模式，但如果下游不支持上游缺省的 collation，同步可能会报错，需要手动处理。

cc @glorv @lance6716

> - **在 DM 中 优化 `transfer source`，支持平滑执行同步任务**

Reviewer 18: ```suggestion
- **在 DM 中优化 `transfer source`，支持平滑执行同步任务**
```

    当 DM-worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 `source` 配置迁移到其他节点。优化后的 `transfer source` 简化了用户操作步骤，不再要求先暂停所有关联 task 而是直接执行平滑迁移，DM 将在内部完成所需操作。

- **DM OpenAPI 特性 GA**

    DM 支持通过 API 的方式进行日常管理，包括增加数据源、管理任务等。本次更新 OpenAPI 从实验特性转为正式特性。

### 问题诊断效率

- **Top SQL（实验特性）**

    新推出实验性特性 Top SQL（默认关闭），帮助用户轻松找到节点中负载贡献较大的查询。

    [用户文档](/dashboard/top-sql.md)

### TiDB 数据共享订阅

- **优化 TiCDC 对集群的影响**

    大幅降低了 TiCDC 启用后，对 TiDB 集群的性能影响。在实验室环境中，TiCDC 对 TiDB 的性能影响可以降低到 5% 以下。

### 部署及运维

- **持续性能分析（实验特性）**

>     - 支持更多组件：支持 TiFlash 组件查看 CPU Profiling

Reviewer 13: ```suggestion
    - 支持更多组件：除了 TiDB、PD 和 TiKV，5.4.0 版本还支持查看 TiFlash CPU Profiling
```

    - 支持更方便的查看形式：支持以火焰图形式查看 CPU Profiling 和 Goroutine 结果。
    - 支持更多部署环境：支持在 TiDB Operator 部署环境下启用持续性能分析功能。

    该功能默认关闭，需进入 TiDB Dashboard 持续性能分析页面开启，开启方法见[用户文档](/dashboard/continuous-profiling.md)。

>     要使用持续性能分析功能，集群须由 TiUP v1.9.0 及以上版本或 TiDB Operator vx.x.x（TBD）及以上版本升级或安装。

Reviewer 13: ```suggestion
    持续性能分析仅支持由 TiUP v1.9.0 及以上版本或 TiDB Operator vx.x.x（TBD）及以上版本升级或安装的集群。
```

## 提升改进

+ TiDB

    - 新增系统变量 `tidb_enable_paging`，开启该功能可显著降低使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 数据较小且无法下推到 `IndexScan` 上的读请求的延迟 [#30578](https://github.com/pingcap/tidb/issues/30578)

> 

Reviewer 14: - 支持 `ADMIN {SESSION | INSTANCE | GLOBAL} PLAN_CACHE` 语法，用于清空缓存的查询计划 [#30370](https://github.com/pingcap/tidb/pull/30370)

Author: Added

+ TiKV

    - Coprocessor 支持分页 API 进行流式处理 [#11448](https://github.com/tikv/tikv/issues/11448)
    - 支持 `read-through-lock`，使读操作不需要等待清理 secondary lock [#11402](https://github.com/tikv/tikv/issues/11402)

>     - 增加了磁盘保护机制，尽量避免磁盘空间耗尽导致 panic [#10960](https://github.com/tikv/tikv/issues/10960) [#10959](https://github.com/tikv/tikv/issues/10959) [#10926](https://github.com/tikv/tikv/issues/10926) [#10925](https://github.com/tikv/tikv/issues/10925) [#10895](https://github.com/tikv/tikv/issues/10895) [#10894](https://github.com/tikv/tikv/issues/10894) [#10868](https://github.com/tikv/tikv/issues/10868) [#10840](https://github.com/tikv/tikv/issues/10840) [#10839](https://github.com/tikv/tikv/issues/10839) [#10819](https://github.com/tikv/tikv/issues/10819) [#10801](https://github.com/tikv/tikv/issues/10801)

Reviewer: I think one tracking issue is enough. cc @tier-cap 
```suggestion
    - 增加了磁盘保护机制，尽量避免磁盘空间耗尽导致 panic [#10537](https://github.com/tikv/tikv/issues/10537)
```

    - 日志支持存档和轮替 [#11651](https://github.com/tikv/tikv/issues/11651)
    - 减少 Raft 客户端的系统调用并提高 CPU 效率 [#11309](https://github.com/tikv/tikv/issues/11309)
    - Coprocessor 支持下推 substring 到 TiKV [#11495](https://github.com/tikv/tikv/issues/11495)
    - 通过跳过读锁提高在 RC 隔离级别中扫描的性能 [#11485](https://github.com/tikv/tikv/issues/11485)
    - 减少备份使用的默认线程池大小，并在压力大时限制其使用 [#11000](https://github.com/tikv/tikv/issues/11000)
    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)

+ PD

+ TiFlash

+ Tools

    + Backup & Restore (BR)

    + TiCDC

    + TiDB Data Migration (DM)

    + TiDB Lightning

    + Dumpling

    + TiDB Binlog

## Bug 修复

+ TiDB

> 

Reviewer 14: + 修复当从 v4.x 版本升级到 v5.x 版本后 `tidb_analyze_version` 的值变化的问题 [#25422](https://github.com/pingcap/tidb/issues/25422)
+ 修复在子查询中使用不同的 collation 导致查询结果错误的问题 [#30748](https://github.com/pingcap/tidb/issues/30748)
+ 修复 `concat(ifnull(time(3))` 的结果与 MySQL 不一致的问题 [#29498](https://github.com/pingcap/tidb/issues/29498)
+ 修复乐观事务下数据索引可能不一致的问题 [#30410](https://github.com/pingcap/tidb/issues/30410)
+ 修复当表达式不能下推给 TiKV 时 IndexMerge 查询计划错误的问题 [#30200](https://github.com/pingcap/tidb/issues/30200)
+ 修复并发的列类型变更导致 schema 与数据不一致的问题 [#31048](https://github.com/pingcap/tidb/issues/31048)
+ 修复当有子查询时 IndexMerge 的查询结果错误的问题 [#30913](https://github.com/pingcap/tidb/issues/30913)
+ 修复当客户端设置过大的 FetchSize 后 TiDB 执行会 panic 的问题 [#30896](https://github.com/pingcap/tidb/issues/30896)
+ 修复当 Lightning 没有权限访问 `mysql.tidb` 表时，导入的结果不正确的问题 [#31088](https://github.com/pingcap/tidb/issues/31088)
+ 修复 LEFT JOIN 有时会被错误地转换成 INNER JOIN 的问题 [#20510](https://github.com/pingcap/tidb/issues/20510)
+ 修复当 `CASE-WHEN` 表达式与 collation 同时使用时可能 panic 的问题 [#30245](https://github.com/pingcap/tidb/issues/30245)
+ 修复当 `IN` 的值中带有二进制常量时查询结果错误的问题 [#31261](https://github.com/pingcap/tidb/issues/31261)
+ 修复当 CTE 中带有子查询时查询结果错误的问题 [#31255](https://github.com/pingcap/tidb/issues/31255)
+ 修复 `INSERT ... SELECT ... ON DUPLICATE KEY UPDATE` 语句 panic 的问题 [#28078](https://github.com/pingcap/tidb/issues/28078)
+ 修复 INDEX HASH JOIN 报 `send on closed channel` 的问题 [#31129](https://github.com/pingcap/tidb/issues/31129)

Author: Added

---
title: TiDB 5.4 Release Notes
---

# TiDB 5.4 Release Notes

发版日期：2022 年 x 月 x 日

TiDB 版本：5.4.0

在 v5.4.0 版本中，你可以获得以下关键特性：

> 

Reviewer 8: 支持 _索引合并 (Index Merge)_ 数据访问方法，使查询能够利用到多个列上的独立索引，提高过滤效果

Author: 已添加

+ 支持 GBK 字符集
+ 支持通过 session 变量实现有界限过期数据读取
+ 支持统计信息采集配置持久化
+ 支持使用 Raft Engine 作为 TiKV 的日志存储引擎
+ 优化备份对集群的影响
+ 支持 Azure Blob Storage 作为备份目标存储（实验特性）

> + 提升 TiFlash 性能和稳定性

Reviewer 8: 持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能

Author: 已添加

+ 为 TiDB Lightning 增加已存在数据表是否允许导入的开关
+ 优化持续性能分析实验特性
+ TiSpark 支持用户认证与鉴权

## 兼容性变化

> **注意：**
>
> 当从一个早期的 TiDB 版本升级到 TiDB v5.4.0 时，如需了解所有中间版本对应的兼容性更改说明，请查看对应版本的 [Release Notes](/releases/release-notes.md)。

### 系统变量

|  变量名    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- |
|  `tidb_backoff_lock_fast` | 修改 | 默认值由 `100` 修改为 `10` |
| `tidb_enable_index_merge` | 修改 | 默认值由 `OFF` 改为 `ON`。如果从 TiDB v4.0（不包含 v4.0.0）升级到 v5.4.0 及更新的集群，该变量值默认保持 `OFF`。 |
| `tidb_enable_paging`  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用 paging 方式发送 coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。在这种情况下，由于 `Limit` 算子限制只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，降低延迟，减少资源消耗。 |
| `tidb_read_staleness` | 新增 | 用于设置当前会话期待读取的历史数据的所处时刻，默认值为 `0` |

> | `tidb_stats_load_sync_wait` | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `0` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时 |

Reviewer 11: ```suggestion
| [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `0` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时。 |
```

> | `tidb_stats_load_pseudo_timeout` | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败(`false`) 还是退回使用 pseudo 的统计信息 (`true`) |

Reviewer 11: ```suggestion
| [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败(`false`) 还是退回使用 pseudo 的统计信息 (`true`) |
```

> | `tidb_persist_analyze_options`  | 新增  | 这个变量用于控制是否持久化执行 `ANALYZE` 语句时指定的配置项 |

Reviewer 11: ```suggestion
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 这个变量用于控制是否持久化执行 `ANALYZE` 语句时指定的配置项 |
```

### 配置文件参数

|  配置文件    |  配置项    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- | :----------- |

> | TiDB | `stats-load-concurrency` | 新增 |  用于设置 TiDB 统计信息同步加载功能最多可以并发处理多少列，默认值为 `5`             |

Reviewer 11: ```suggestion
| TiDB | [`stats-load-concurrency`](/tidb-configuration-file.md#stats-load-concurrency-从-v540-版本开始引入) | 新增 |  用于设置 TiDB 统计信息同步加载功能可以并发处理的最大列数，默认值为 `5`             |
```

> | TiDB | `stats-load-queue-size`   | 新增 |  用于设置 TiDB 统计信息同步加载功能最多可以缓存多少列的请求，默认值为 `1000`             |

Reviewer 11: ```suggestion
| TiDB | [`stats-load-queue-size`](/tidb-configuration-file.md#stats-load-queue-size-从-v540-版本开始引入)   | 新增 |  用于设置 TiDB 统计信息同步加载功能最多可以缓存多少列的请求，默认值为 `1000`             |
```

format = "text"
设置日志输出格式，默认为 text，支持 text 和 json 两种格式。

enable-timestamp = true
设置时间戳输出开关，默认为 true。

filename = ""
设置日志文件名前缀，默认为无前缀。

max-size = 300
设置日志分割大小，默认为 300 MB，最大支持 4096 MB。

max-days = 0
设置日志最大保留天数，默认全保留不清理。

max-backups = 0
设置日志备份的最大保留文件数，默认全保留不清理。
```

### 其他

- TiDB Dashboard 默认不再使用 `root` + 空密码登录。

    从 v5.4.0开始，使用 TiUP 启动集群时推荐使用 `start --initial`。执行该操作启动集群后，会为 `root` 账号自动生成一个随机密码，`root` 账号登录 Dashboard 需要使用这个密码。

- 为 TiDB 和 PD 之间新增接口。使用 `information_schema.TIDB_HOT_REGIONS_HISTORY` 系统表时，TiDB 需要使用匹配的 PD 版本。

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

    在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。为了更好的支持中文用户，TiDB 从 v5.4.0 版本开始支持 GBK 字符集，同时支持 `gbk_bin` 和 `gbk_chinese_ci` 两种排序规则。

    在使用 GBK 字符集时，请注意以下兼容性限制：

    - TiFlash 暂不支持 GBK 字符。
    - TiCDC 暂不支持 GBK 字符。
    - 如果 `character_set_client` 和 `character_set_connection` 都是 `gbk` 时，处理非法 GBK 字符与 MySQL 存在兼容性问题。
    - `character_set_client` 在处理 `prepare` 语句时可能出现兼容性问题。
    - TiDB 不支持 `_gbk"xxx"` 的用法，但是支持 `_utf8mb4"xxx"` 的用法。而 MySQL 对于 `_charset"xxx"` 的用法都支持。
    - TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.3.0 之前不支持恢复 `charset=GBK` 的表。

> - **TiFlash IndexMerge GA**

Reviewer 8: 这一条应当放到性能里面。并不是对SQL的语意变更或语法扩充。

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

### 性能

> 

Reviewer 8: - TiDB 正式发布 **索引合并 (Index Merge)** 数据访问方法
    _索引合并_ 是 TiDB 在 4.0 版本作为实验功能引入的一种查询执行方式的优化，可以大幅提高查询在扫描多列数据时条件过滤的效率。例如对以下的查询 _索引合并_ 可以同时利用 _key1_ 与_key2_ 两个过滤条件所在列上的索引上进行过滤，然后合并出最终的结果。以往 TiDB 在一个表上的查询只能使用一个索引，无法同时使用多个索引分别进行条件扫描。相较以往 _索引合并_ 避免了此类不必要的大量扫描，也可以使得需要灵活查询不特定多列数据组合的用户不必大量构建多列复合索引。 
```sql
 SELECT * FROM table WHERE key1 <= 100 OR key2 = 200; 
 ```

5.4.0 版本正式发布 _索引合并(Index Merge)_ ，但仍存在以下的使用条件和限制（详情请参见 [用户文档](https://docs.pingcap.com/tidb/stable/explain-index-merge)）：
目前 TiDB 的 索引合并 优化只限于 _析取范式_ (如：X<sub>1</sub> ⋁ X<sub>2</sub> ⋁ …X<sub>n</sub>), `WHERE` 子句中过滤条件连接词为 `OR` 的查询的合并索引访问优化。（后继开发将继续支持 _和取范式_ 如:  Y<sub>1</sub> ⋀ Y<sub>2</sub> ⋀ …Y<sub>3</sub> 形式即接词为 `AND` 的查询条件组合）。
5.4.0 之后全新的安装的 TiDB 将会默认开启此项功能，而对于旧系统升级的情况则默认保持就系统中的开闭状态（4.0 之前无此项功能的版本也视同关闭）需要由用户自己决定是否开启。

Author: 已添加至 PR 中

- **支持 Azure Blob Storage 作为备份目标存储（实验特性）**

    Backup & Restore (BR) 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

    该功能目前是实验特性，详细情况参考 [BR 支持 Azure Blob Storage 远端存储](/br/backup-and-restore-azblob.md)。

### 数据迁移

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

- **为 TiDB Lightning 增加新参数，用于在并行导入模式下保存各个 TiDB Lightning 实例元信息的 schema 名字 **

     为 TiDB Lightning 增加 `meta-schema-name` 参数。在并行导入模式下，该参数用于在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认值为 "lightning_metadata"。对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。

- **在 TiDB Lightning 中添加重复数据的检测**

    在 `backend=local` 模式下，数据导入完成之前 TiDB Lightning 会输出冲突数据，然后从数据库中删除这些冲突数据。用户可以在导入完成后解析冲突数据，并根据业务规则选择适合的数据进行插入。建议根据冲突数据清洗上游数据源，避免在后续增量数据迁移阶段遇到冲突数据而造成数据不一致。

- **在 TiDB Data Migration (DM) 中 优化 relay log 的使用方式**

    - 恢复 `source` 配置中 `enable-relay` 开关
    - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能
    - relay log 的开启状态与 `source` 绑定，source 迁移到任意 DM-worker 均保持原有开启或关闭状态
    - relay log 的存放路径移至 DM-worker 配置文件

> - **在 DM 中优化排序规则的处理方式**
> 
>     增加 `collation_compatible` 开关，支持 `strict` 和 `loose`（默认）两种模式。如果对排序规则要求不严格，允许排序规则在上下游不一致，使用默认的 `loose` 模式可使同步正常进行；如果对排序规则要求严格，排序规则在上下游不一致会导致业务报错，则可以使用 `strict` 模式。

Reviewer 15: ```suggestion
- **在 DM 中优化[排序规则](/character-set-and-collation.md)的处理方式**

    增加 collation_compatible 开关，支持 strict 和 loose（默认）两种模式：

    - 如果对排序规则要求不严格，允许上下游查询结果排序规则不一致，使用默认的 loose 模式可以避免报错。
    - 如果对排序规则要求严格，业务要求排序规则必须一致，则应当使用 strict 模式。但如果下游不支持上游缺省的 collation，同步可能会报错。
```

| TiKV | `raft-engine` | 新增 | 包含 `enable`、`dir`、`batch-compression-threshold`、`bytes-per-sync`、`target-file-size`、`purge-threshold`、`recovery-mode`、`recovery-read-block-size`、`recovery-read-block-size`、`recovery-threads`，详情参见 [TiKV 配置文件：raft-engine](/tikv-configuration-file.md#raft-engine)。|
| TiKV | `raftstore.raft-log-compact-sync-interval` | 新增 | 控制 CompactLog 命令的间隔，默认值为 `2s` |
| TiKV | [`backup.enable-auto-tune`](/tikv-configuration-file.md#enable-auto-tune-从-v54-版本开始引入) | 修改 | 在 v5.3.0 中默认值为 `false`，自 v5.4.0 起默认值改为 `true`。表示在集群资源占用率较高的情况下，是否允许 BR 自动限制备份使用的资源，减少对集群的影响。在默认配置下，备份速度可能下降。 |
| TiKV | `log-level`、`log-format`、`log-file`、`log-rotation-size` | 修改 | 将 TiKV log 参数名替换为与 TiDB log 参数一致的参数名，即 `log.level`、`log.format`、`log.file.filename`、`log.enable-timestamp`。如果只设置了原参数、且把其值设为非默认值，原参数与新参数会保持兼容；如果同时设置了原参数和新参数，则会使用新参数。详情参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。 |
| TiKV | `raftstore.raft-log-gc-tick-interval` | 修改 | 默认值修改为 `3s` |
| TiKV  |  `log-rotation-timespan`  | 删除 |  轮换日志的时间跨度。当超过该时间跨度，日志文件会被轮换，即在当前日志文件的文件名后附加一个时间戳，并创建一个新文件。 |
| PD | `log.level` | 修改 | 默认值由 "INFO" 改为 "info"，保证大小写不敏感 |

> | TiFlash | `profile.default.enable_elastic_threadpool` | 新增  |  表示是否启用可自动扩展的线程池。打开该配置项可以显著提高 TiFlash 在高并发场景的 CPU 利用率。默认值为 `false`。|
> | TiDB Data Migration (DM) | [`collation_compatible`](/dm/task-configuration-file-full.md#完整配置文件示例) | 同步 CREATE 语句中缺省 Collation 的方式，可选 "loose" 和 "strict"，默认为 "loose"。 |
> | TiFlash | [`storage.format_version`](/tiflash/tiflash-configuration.md#配置文件-tiflashtoml) | 新增 | 表示 DTFile 储存文件格式，默认值为 `2`，该格式在数据文件中内嵌哈希值。也可以设置为 `3`，该格式包含元数据，标记数据校验，支持多种哈希算法。|
> | TiFlash | `logger.count` | 修改 | 默认值修改为 `10` |
> | TiFlash | `status.metrics_port` | 修改 | 默认值修改为 `8234` |
> | TiFlash | `raftstore.apply-pool-size` | 新增 | 处理 Raft 数据落盘的线程池中线程的数量，默认值为 `4`。 |
> | TiFlash | `raftstore.store-pool-size` | 新增 | 处理 Raft 的线程池中线程的数量，即 Raftstore 线程池的大小，默认值为 `4`。 |
> | TiCDC | `enable-tidb-extension` | 新增 |  开启该配置项后，TiCDC 在 Canal-JSON 协议格式中附加 TiDB 扩展字段。                  |

Reviewer 13: ```suggestion
| TiFlash | `profile.default.enable_elastic_threadpool` | 新增  |  表示是否启用可自动扩展的线程池。打开该配置项可以显著提高 TiFlash 在高并发场景的 CPU 利用率。默认值为 `false`。|
| TiFlash | [`storage.format_version`](/tiflash/tiflash-configuration.md#配置文件-tiflashtoml) | 新增 | 表示 DTFile 储存文件格式，默认值为 `2`，该格式在数据文件中内嵌哈希值。也可以设置为 `3`，该格式包含元数据，标记数据校验，支持多种哈希算法。|
| TiFlash | `logger.count` | 修改 | 默认值修改为 `10` |
| TiFlash | `status.metrics_port` | 修改 | 默认值修改为 `8234` |
| TiFlash | `raftstore.apply-pool-size` | 新增 | 处理 Raft 数据落盘的线程池中线程的数量，默认值为 `4`。 |
| TiFlash | `raftstore.store-pool-size` | 新增 | 处理 Raft 的线程池中线程的数量，即 Raftstore 线程池的大小，默认值为 `4`。 |
| TiDB Data Migration (DM) | [`collation_compatible`](/dm/task-configuration-file-full.md#完整配置文件示例) | 同步 CREATE 语句中缺省 Collation 的方式，可选 "loose" 和 "strict"，默认为 "loose"。 |
| TiCDC | `enable-tidb-extension` | 新增 |  开启该配置项后，TiCDC 在 Canal-JSON 协议格式中附加 TiDB 扩展字段。                  |
```
Moved the DM configuration under TiFlash, next to TiCDC.

| TiCDC | `partition-num`      | 修改 | 将 Kafka Sink `partition-num` 的默认值改由 `4` 为 `3`，使 TiCDC 更加平均地分发消息到各个 Kafka partition |
| TiDB Lightning | `meta-schema-name` | 修改 | 此配置项控制 TiDB Lightning 在目标 TiDB 中保存 metadata 对应的 schema name。从 v5.4.0 开始，只在开启了并行导入功能时（对应配置为 `tikv-importer.incremental-import = true` ），才会在目标 TiDB 中创建此库。 |
| TiDB Lightning | `task-info-schema-name` |  新增  | 用于配置当 Lightning 检测到冲突数据时，对应冲突数据存储的库名，默认值为 "lightning_task_info"。如果没有开启冲突检测功能，则无需配置此参数。  |
| TiDB Lightning | `incremental-import` | 新增 | 是否允许向已存在数据的表导入数据。默认值为 false |

### 其他

- TiDB Dashboard 默认不再使用 `root` + 空密码登录。

    从 v5.4.0开始，使用 TiUP 启动集群时推荐使用 `start --initial`。执行该操作启动集群后，会为 `root` 账号自动生成一个随机密码，`root` 账号登录 Dashboard 需要使用这个密码。

- 为 TiDB 和 PD 之间新增接口。使用 `information_schema.TIDB_HOT_REGIONS_HISTORY` 系统表时，TiDB 需要使用匹配的 PD 版本。
- TiDB Server、PD Server 和 TiKV Server 将采用统一的参数命名方式来管理日志命名、输出格式、轮转和过期的规则。参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

>     在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。为了更好的支持中文用户，TiDB 从 v5.4.0 版本开始支持 GBK 字符集，同时支持 `gbk_bin` 和 `gbk_chinese_ci` 两种排序规则（在使用这两种排序规则前需要开启 TiDB 配置项 `new_collations_enabled_on_first_bootstrap`）。

Reviewer 11: @zimulala PTAL Thanks

>     在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。为了更好的支持中文用户，TiDB 从 v5.4.0 版本开始支持 GBK 字符集，同时支持 `gbk_bin` 和 `gbk_chinese_ci` 两种排序规则（在使用这两种排序规则前需要开启 TiDB 配置项 `new_collations_enabled_on_first_bootstrap`）。

Reviewer 17: LGTM

    
    [用户文档](/statistics.md#统计信息的加载)

### 稳定性

- **支持统计信息采集配置持久化**

    统计信息是优化器生成执行计划时所参考的基础信息之一，统计信息的准确性直接影响生成的执行计划是否合理。为了保证统计信息的准确性，有时候需要针对不同的表、分区、索引设置不同的采集配置项。

    TiDB 从 v5.4.0 版本开始支持 `ANALYZE` 配置持久化功能，方便后续收集统计信息时沿用已有配置项。
    
    `ANALYZE` 配置持久化功能默认开启（系统变量 `tidb_analyze_version = 2` 且 `tidb_persist_analyze_options = true`），用于记录手动执行 `ANALYZE` 语句时指定的持久化配置项。记录后，当 TiDB 下一次自动更新统计信息或者你手动收集统计信息但未指定配置项时，TiDB 会按照记录的配置项收集统计信息。
    
    [用户文档](/statistics.md#analyze-配置持久化)

## 高可用和容灾

- **优化备份对集群的影响**

    Backup & Restore (BR) 增加了备份线程自动调节功能（默认开启）。该功能通过监控集群资源的使用率自动调节备份的线程数的方式，降低备份过程对集群的影响。在某些 Case 验证中，通过增加集群用于备份的资源和开启备份线程自动调节功能，备份的影响可以降低到 10% 以下。
    详细文档请阅读 [BR 自动调节](/br/br-features.md#自动调节-从-v54-版本开始引入)。

- **支持 Azure Blob Storage 作为备份目标存储（实验特性）**

    Backup & Restore (BR) 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

    该功能目前是实验特性，详细情况参考 [BR 支持 Azure Blob Storage 远端存储](/br/backup-and-restore-azblob.md)。

### 数据迁移

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

> - **为 TiDB Lightning 增加新参数，用于在并行导入模式下保存各个 TiDB Lightning 实例元信息的 schema 名字 **

Reviewer 12: ```suggestion
- **为 TiDB Lightning 增加新参数，用于在并行导入模式下保存各个 TiDB Lightning 实例元信息的 schema 名字，确保导入数据的正确性 **
```

| [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败(`false`) 还是退回使用 pseudo 的统计信息 (`true`) |
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启[ ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性 |
| `tidb_store_limit` | 修改 | v5.4.0 前支持实例级别及集群级别的设置，现在只支持集群级别的设置 |

### 配置文件参数

|  配置文件    |  配置项    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- | :----------- |
| TiDB | [`stats-load-concurrency`](/tidb-configuration-file.md#stats-load-concurrency-从-v540-版本开始引入) | 新增 |  用于设置 TiDB 统计信息同步加载功能可以并发处理的最大列数，默认值为 `5`             |
| TiDB | [`stats-load-queue-size`](/tidb-configuration-file.md#stats-load-queue-size-从-v540-版本开始引入)   | 新增 |  用于设置 TiDB 统计信息同步加载功能最多可以缓存多少列的请求，默认值为 `1000`             |
| TiKV | [`backup.enable-auto-tune`](/tikv-configuration-file.md#enable-auto-tune-从-v54-版本开始引入) | 修改 | 在 v5.3.0 中默认值为 `false`，自 v5.4.0 起默认值改为 `true`。表示在集群资源占用率较高的情况下，是否允许 BR 自动限制备份使用的资源，减少对集群的影响。在默认配置下，备份速度可能下降。 |

> | TiKV | `log.level`、`log.format`、`log.enable-timestamp`、`log.file.filename`、`log.file.max-size`、`log.file.max-days`、`log.file.max-backups` | 新增  | 参数说明见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。 |
> | TiKV | `log-level`、`log-format`、`log-file`、`log-rotation-size` | 变更 | 将 TiKV log 参数名改为与 TiDB log 参数一致的命名方式，即 `log.level`、`log.format`、`log.enable-timestamp`。如果 TiKV log 参数为非默认值则保持兼容；如果同时配置 TiKV log 参数和 TiDB log 命名方式的参数，使用 TiDB log 命名方式的参数。详情参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。 |
> | TiKV  |  `log-rotation-timespan`  | 删除 |  轮换日志的时间跨度。当超过该时间跨度，日志文件会被轮换，即在当前日志文件的文件名后附加一个时间戳，并创建一个新文件。 |

Reviewer 19: ```suggestion
| TiKV | `log.file.max-size`、`log.file.max-days`、`log.file.max-backups` | 新增  | 参数说明见 [TiKV 配置文件 - log.file](/tikv-configuration-file.md#logfile-从-v540-版本开始引入)。 |
| TiKV | `log-level`、`log-format`、`log-file`、`log-rotation-size` | 修改 | 将 TiKV log 参数名替换为与 TiDB log 参数一致的参数名，即 `log.level`、`log.format`、`log.file.filename`、`log.enable-timestamp`。如果只设置了原参数、且把其值设为非默认值，原参数与新参数会保持兼容；如果同时设置了原参数和新参数，则会使用新参数。详情参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。 |
| TiKV  |  `log-rotation-timespan`  | 删除 |  轮换日志的时间跨度。当超过该时间跨度，日志文件会被轮换，即在当前日志文件的文件名后附加一个时间戳，并创建一个新文件。 |
```

    支持使用 [Raft Engine](https://github.com/tikv/raft-engine) 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。

    由于 Raft Engine 涉及数据格式改动，目前仍属于实验特性，并默认关闭。同时请注意最新的 Raft Engine 不与 v5.4.0 版本前的 Raft Engine 兼容。因此在进行跨越 v5.4.0 版本的升级和降级之前，需要确保已有 TiKV 节点上的 Raft Engine 已被关闭。

    [用户文档](/tikv-configuration-file.md#raft-engine)

- **持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能**

    - 支持将更多函数下推至 MPP 引擎
        - 字符串函数：`LPAD()`、`RPAD()`、`STRCMP()`
        - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
    - 引入动态线程池，提升资源利用率（实验特性）
    - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性
    - 提升由行存到列存数据同步处理时对 raft log 的解码 (decoding) 效率，使 CPU 使用率降低达 90%
    - TiFlash 调整了文件系统中 Delta Tree 相关的默认参数，使之更适合一般的生产环境

- **通过 session 变量实现有界限过期数据读取**

    TiDB 是基于 Raft 协议的多副本分布式数据库。面对高并发，高吞吐业务场景，可以通过follower 节点实现读性能扩展，构建读写分离架构。

    针对不同的业务场景，follower 提供强一致读和弱一致过期读两种读模式。强一致读满足数据实时性要求严格的业务场景，但是因为 leader 和 follower 的数据同步延迟、吞吐较低、延迟较高，特别是在跨机房架构下延迟问题被进一步放大。

    在对数据实时性要求不高的业务场景下，可以选择过期读模式。使用该模式可以降低延迟和提升吞吐。TiDB 目前支持通过显示只读事务或 SQL 语句的方式实现过期读。两种方式均支持指定时间的精确过期读和指定时间边界的过期读两种模式，详细用法请参考[过期读文档](/read-historical-data.md)。

>     从 v5.4.0 版本开始 TiDB 支持通过 session 变量设置有界限过期读，进一步提升易用性，具体设置如下：

Reviewer 19: ```suggestion
    从 v5.4.0 版本开始 TiDB 支持通过 session 变量设置有界限过期读，进一步提升易用性，具体设置示例如下：
```

    ```sql
    set @@tidb_replica_read=leader_and_follower
    set @@tidb_read_staleness="-5"
    ```

>     通过该设置，可以实现就近选取 leader 或 follower 节点，并读取 5 秒钟前的最新过期数据，满足准实时场景下低延迟高吞吐数据访问的业务诉求，降低研发门槛，提升易用性。

Reviewer 19: ```suggestion
    通过该设置，TiDB 可以实现就近选取 leader 或 follower 节点，并读取 5 秒钟前的最新过期数据，满足准实时场景下低延迟高吞吐数据访问的业务诉求，降低研发门槛，提升易用性。
```

>     通过该设置，TiDB 可以实现就近选取 leader 或 follower 节点，并读取 5 秒钟前的最新过期数据，满足准实时场景下低延迟高吞吐数据访问的业务诉求，降低研发门槛，提升易用性。

Reviewer 19: ```suggestion
    通过该设置，TiDB 可以实现就近选取 leader 或 follower 节点，并读取 5 秒钟内的最新过期数据，满足准实时场景下低延迟高吞吐数据访问的业务诉求，降低研发门槛，提升易用性。
```

- **TiDB 正式发布索引合并 (Index Merge) 功能**

    [索引合并](/explain-index-merge.md) 是在 TiDB v4.0 版本中作为实验特性引入的一种查询执行方式的优化，可以大幅提高查询在扫描多列数据时条件过滤的效率。例如对以下的查询，若 `WHERE` 子句中两个 `OR` 连接的过滤条件在各自包含的 _key1_ 与 _key2_ 两个列上都存在索引，则 _索引合并_ 可以同时利用  _key1_ 与 _key2_ 上的索引分别进行过滤，然后合并出最终的结果。

    ```sql
    SELECT * FROM table WHERE key1 <= 100 OR key2 = 200;
    ```

    以往 TiDB 在一个表上的查询只能使用一个索引，无法同时使用多个索引进行条件过滤。相较以往，_索引合并_ 避免了此情况下可能不必要的大量数据扫描，也可以使得需要灵活查询不特定多列数据组合的用户利用单列上的索引达到高效稳定的查询，无需大量构建多列复合索引。

    本版本正式发布了 _索引合并 (Index Merge)_ 特性，但仍存在以下的使用条件和限制（详情请参见[用户文档](/explain-index-merge.md)）：

    目前 TiDB 的 _索引合并_ 优化只限于 _析取范式_ (X<sub>1</sub> ⋁ X<sub>2</sub> ⋁ …X<sub>n</sub>)，即 `WHERE` 子句中过滤条件连接词为 `OR`。

    如果全新部署的集群版本为 v5.4.0 或以上，此特性默认开启。如果从 v5.4.0 以前的版本升级到 v5.4.0 或以上，默认保持升级前此特性的开关状态（v4.0.0 之前无此项特性的版本默认关闭），由用户决定是否开启。

> - **支持收集部分列的统计信息（实验特性）**
>     
>     执行 SQL 语句时，优化器在大多数情况下只会用到部分列（例如， `WHERE`、`JOIN`、`ORDER BY`、`GROUP BY` 子句中用到的列）的统计信息。这些被优化器用到的列称为 `PREDICATE COLUMNS`。
>     
>     从 v5.4.0 开始，TiDB 引入了收集部分列的统计信息的特性（默认关闭），支持只收集指定列或者 `PREDICATE COLUMNS` 的统计信息，极大地降低了收集统计信息的开销。
>     
>     [用户文档](/statistics.md#收集部分列的统计信息)

Reviewer 11: ```suggestion
- **支持收集 `PREDICATE COLUMNS` 的统计信息（实验特性）**

    执行 SQL 语句时，优化器在大多数情况下只会用到部分列（例如， `WHERE`、`JOIN`、`ORDER BY`、`GROUP BY` 子句中出现的列）的统计信息，这些用到的列称为 `PREDICATE COLUMNS`。
    
    从 v5.4.0 开始，你可以设置系统变量 [`tidb_enable_column_tracking`](/system-variables.md#tidb_enable_column_tracking-从-v540-版本开始引入) 的值为 `ON` 开启 TiDB 对 `PREDICATE COLUMNS` 的收集。
    
    开启后，TiDB 将每隔 100 * [`stats-lease`](/tidb-configuration-file.md#stats-lease) 时间将 `PREDICATE COLUMNS` 信息写入系统表 `mysql.column_stats_usage`。等到业务的查询模式稳定以后，使用 `ANALYZE TABLE TableName PREDICATE COLUMNS` 语法收集 `PREDICATE COLUMNS` 列的统计信息，可以极大地降低收集统计信息的开销。
    
    [用户文档](/statistics.md#收集部分列的统计信息)
```

    从 v5.4.0 开始，TiDB 引入了统计信息同步加载的特性，支持执行当前 SQL 语句时将直方图、TopN、CMSketch 等占用空间较大的统计信息同步加载到内存，提高该 SQL 语句优化时统计信息的完整性。

- **优化 RocksDB 在大数据量下的锁争用问题**

    RocksDB 依赖全局锁来维护多版本的文件信息，并当销毁读快照时对过期文件进行清理。在文件数较多且读写频繁的场景下，这一操作显著影响前台延迟。本版本针对这一问题进行优化，详见 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)。

### 稳定性

- **支持统计信息采集配置持久化**

    统计信息是优化器生成执行计划时所参考的基础信息之一，统计信息的准确性直接影响生成的执行计划是否合理。为了保证统计信息的准确性，有时候需要针对不同的表、分区、索引设置不同的采集配置项。

>     TiDB 从 v5.4.0 版本开始支持通过 `analyze` 命令采集统计信息并持久化指定的配置项，方便后续的统计信息采集沿用已有配置项。具体配置项信息请参考 [`Analyze` 文档](/sql-statements/sql-statement-analyze-table.md#analyze)的 `AnalyzeOption` 章节。
> 
>     - 开启采集配置项持久化
> 
>         设置 `tidb_analyze_version = 2` 且 `tidb_persist_analyze_options = true` 会开启配置项持久化。开启后，手动 analyze 指定的所有配置项会被持久化并覆盖已有选项。后续的手动或自动 analyze 任务会沿用已有配置项进行统计信息采集，直到持久化功能关闭或用户手动指定新的采集配置项。
> 
>     - 关闭采集配置项持久化
> 
>         设置 `tidb_analyze_version = 1` 或 `tidb_persist_analyze_options = false` 会关闭采集配置项持久化功能。持久化关闭后，已有配置项不会被删除，但不记录新增配置项。新采集任务不会沿用已有的持久化配置项，再次开启采集配置项持久化将会直接使用已有的配置项进行统计信息采集。如果需要更新已有的配置项，请手动执行 `analyze` 命令并指定新的采集配置项。

Reviewer 11: ```suggestion
    TiDB 从 v5.4.0 版本开始支持 `ANALYZE` 配置持久化功能，方便后续的统计信息采集沿用已有配置项。
    
    `ANALYZE` 配置持久化功能默认开启（系统变量 `tidb_analyze_version = 2` 且 `tidb_persist_analyze_options = true`），用于记录手动执行 `ANALYZE` 语句时指定的持久化配置项。记录后，当 TiDB 下一次自动更新统计信息或者你手动收集统计信息但未指定配置项时，TiDB 会按照记录的配置项收集统计信息。
    
    多次手动执行 `ANALYZE` 语句并指定持久化配置项时，TiDB 会使用最新一次 `ANALYZE` 指定的配置项覆盖上一次记录的持久化配置。
    
    [用户文档](/statistics.md#analyze-配置持久化)
```

## 提升改进

+ TiDB

    - 新增系统变量 `tidb_enable_paging`，开启该功能可显著降低使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 数据较小且无法下推到 `IndexScan` 上的读请求的延迟 [#30578](https://github.com/pingcap/tidb/issues/30578)
    - 支持 `ADMIN {SESSION | INSTANCE | GLOBAL} PLAN_CACHE` 语法，用于清空缓存的查询计划 [#30370](https://github.com/pingcap/tidb/pull/30370)

+ TiKV

    - Coprocessor 支持分页 API 进行流式处理 [#11448](https://github.com/tikv/tikv/issues/11448)
    - 支持 `read-through-lock`，使读操作不需要等待清理 secondary lock [#11402](https://github.com/tikv/tikv/issues/11402)
    - 增加了磁盘保护机制，尽量避免磁盘空间耗尽导致 panic [#10537](https://github.com/tikv/tikv/issues/10537)
    - 日志支持存档和轮替 [#11651](https://github.com/tikv/tikv/issues/11651)
    - 减少 Raft 客户端的系统调用并提高 CPU 效率 [#11309](https://github.com/tikv/tikv/issues/11309)
    - Coprocessor 支持下推 substring 到 TiKV [#11495](https://github.com/tikv/tikv/issues/11495)
    - 通过跳过读锁提高在 RC 隔离级别中扫描的性能 [#11485](https://github.com/tikv/tikv/issues/11485)
    - 减少备份使用的默认线程池大小，并在压力大时限制其使用 [#11000](https://github.com/tikv/tikv/issues/11000)
    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)

> - 优化在文件数较多且读写频繁的场景下 RocksDB 的全局锁争用问题 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)

Reviewer: ```suggestion
    - 优化在文件数较多且读写频繁的场景下 RocksDB 的全局锁争用问题 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)
```

| TiDB Lightning | `incremental-import` | 新增 | 是否允许向已存在数据的表导入数据。默认值为 false |

### 其他

- TiDB Dashboard 默认不再使用 `root` + 空密码登录。

    从 v5.4.0开始，使用 TiUP 启动集群时推荐使用 `start --initial`。执行该操作启动集群后，会为 `root` 账号自动生成一个随机密码，`root` 账号登录 Dashboard 需要使用这个密码。

- 为 TiDB 和 PD 之间新增接口。使用 `information_schema.TIDB_HOT_REGIONS_HISTORY` 系统表时，TiDB 需要使用匹配的 PD 版本。
- TiDB Server、PD Server 和 TiKV Server 将采用统一的参数命名方式来管理日志命名、输出格式、轮转和过期的规则。参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

    在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。为了更好的支持中文用户，TiDB 从 v5.4.0 版本开始支持 GBK 字符集，同时支持 `gbk_bin` 和 `gbk_chinese_ci` 两种排序规则（在使用这两种排序规则前需要开启 TiDB 配置项 `new_collations_enabled_on_first_bootstrap`）。

在使用 GBK 字符集时，需要注意兼容性限制，详情参考[字符集和排序 - GBK](/character-set-gbk.md)。

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

### 性能

- **新增 Raft Engine（实验特性）**

>     支持使用 [Raft Engine](https://github.com/tikv/raft-engine) 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。
> 
>     由于 Raft Engine 涉及数据格式改动，目前仍属于实验特性，并默认关闭。同时请注意最新的 Raft Engine 不与 v5.4.0 版本前的 Raft Engine 兼容。因此在进行跨越 v5.4.0 版本的升级和降级之前，需要确保已有 TiKV 节点上的 Raft Engine 已被关闭。

Reviewer: ```suggestion
    支持使用 [Raft Engine](https://github.com/tikv/raft-engine) 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。此外，Raft Engine 提升了日志回收效率，修复了极端条件下日志堆积的问题。

    Raft Engine 目前仍属于实验特性，并默认关闭。另外请注意 v5.4 版本的 Raft Engine 数据格式与之前版本不同，对集群做升级或者降级操作之前，需要确保所有 TiKV 节点上的 Raft Engine 已被关闭。我们只建议在 v5.4 及以后的版本使用 Raft Engine。
```

    如果全新部署的集群版本为 v5.4.0 或以上，此特性默认开启。如果从 v5.4.0 以前的版本升级到 v5.4.0 或以上，默认保持升级前此特性的开关状态（v4.0.0 之前无此项特性的版本默认关闭），由用户决定是否开启。

- **支持收集 `PREDICATE COLUMNS` 的统计信息（实验特性）**

    执行 SQL 语句时，优化器在大多数情况下只会用到部分列（例如， `WHERE`、`JOIN`、`ORDER BY`、`GROUP BY` 子句中出现的列）的统计信息，这些用到的列称为 `PREDICATE COLUMNS`。
    
    从 v5.4.0 开始，你可以设置系统变量 [`tidb_enable_column_tracking`](/system-variables.md#tidb_enable_column_tracking-从-v540-版本开始引入) 的值为 `ON` 开启 TiDB 对 `PREDICATE COLUMNS` 的收集。
    
    开启后，TiDB 将每隔 100 * [`stats-lease`](/tidb-configuration-file.md#stats-lease) 时间将 `PREDICATE COLUMNS` 信息写入系统表 `mysql.column_stats_usage`。等到业务的查询模式稳定以后，使用 `ANALYZE TABLE TableName PREDICATE COLUMNS` 语法收集 `PREDICATE COLUMNS` 列的统计信息，可以极大地降低收集统计信息的开销。
    
    [用户文档](/statistics.md#收集部分列的统计信息)
    
- **支持统计信息的同步加载（实验特性）**

    从 v5.4.0 开始，TiDB 引入了统计信息同步加载的特性（默认关闭），支持执行当前 SQL 语句时将直方图、TopN、CMSketch 等占用空间较大的统计信息同步加载到内存，提高该 SQL 语句优化时统计信息的完整性。
    
    [用户文档](/statistics.md#统计信息的加载)

### 稳定性

- **支持统计信息采集配置持久化**

    统计信息是优化器生成执行计划时所参考的基础信息之一，统计信息的准确性直接影响生成的执行计划是否合理。为了保证统计信息的准确性，有时候需要针对不同的表、分区、索引设置不同的采集配置项。

    TiDB 从 v5.4.0 版本开始支持 `ANALYZE` 配置持久化功能，方便后续收集统计信息时沿用已有配置项。
    

>     `ANALYZE` 配置持久化功能默认开启（系统变量 `tidb_analyze_version = 2` 且 `tidb_persist_analyze_options = true`），用于记录手动执行 `ANALYZE` 语句时指定的持久化配置项。记录后，当 TiDB 下一次自动更新统计信息或者你手动收集统计信息但未指定配置项时，TiDB 会按照记录的配置项收集统计信息。

Reviewer 11: ```suggestion
    `ANALYZE` 配置持久化功能默认开启（系统变量 `tidb_analyze_version` 为默认值 `2`，`tidb_persist_analyze_options` 为默认值 `ON`），用于记录手动执行 `ANALYZE` 语句时指定的持久化配置项。记录后，当 TiDB 下一次自动更新统计信息或者你手动收集统计信息但未指定配置项时，TiDB 会按照记录的配置项收集统计信息。
```

    
    [用户文档](/statistics.md#analyze-配置持久化)

## 高可用和容灾

- **优化备份对集群的影响**

    Backup & Restore (BR) 增加了备份线程自动调节功能（默认开启）。该功能通过监控集群资源的使用率自动调节备份的线程数的方式，降低备份过程对集群的影响。在某些 Case 验证中，通过增加集群用于备份的资源和开启备份线程自动调节功能，备份的影响可以降低到 10% 以下。
    详细文档请阅读 [BR 自动调节](/br/br-features.md#自动调节-从-v54-版本开始引入)。

- **支持 Azure Blob Storage 作为备份目标存储（实验特性）**

    Backup & Restore (BR) 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

    该功能目前是实验特性，详细情况参考 [BR 支持 Azure Blob Storage 远端存储](/br/backup-and-restore-azblob.md)。

### 数据迁移

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

- **为 TiDB Lightning 增加新参数，用于在并行导入模式下保存各个 TiDB Lightning 实例元信息的 schema 名字 **

>      为 TiDB Lightning 增加 `meta-schema-name` 参数。在并行导入模式下，该参数用于在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认值为 "lightning_metadata"。对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。

Reviewer 3: ```suggestion
     为 TiDB Lightning 增加 `meta-schema-name` 配置。在并行导入模式下，该参数用于在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名称，默认值为 "lightning_metadata"。对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。
```

- **在 TiDB Lightning 中添加重复数据的检测**

    在 `backend=local` 模式下，数据导入完成之前 TiDB Lightning 会输出冲突数据，然后从数据库中删除这些冲突数据。用户可以在导入完成后解析冲突数据，并根据业务规则选择适合的数据进行插入。建议根据冲突数据清洗上游数据源，避免在后续增量数据迁移阶段遇到冲突数据而造成数据不一致。

- **在 TiDB Data Migration (DM) 中 优化 relay log 的使用方式**

    - 恢复 `source` 配置中 `enable-relay` 开关
    - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能
    - relay log 的开启状态与 `source` 绑定，source 迁移到任意 DM-worker 均保持原有开启或关闭状态
    - relay log 的存放路径移至 DM-worker 配置文件

- **在 DM 中优化[排序规则](/character-set-and-collation.md)的处理方式**

    增加 collation_compatible 开关，支持 strict 和 loose（默认）两种模式：

    - 如果对排序规则要求不严格，允许上下游查询结果排序规则不一致，使用默认的 loose 模式可以避免报错。
    - 如果对排序规则要求严格，业务要求排序规则必须一致，则应当使用 strict 模式。但如果下游不支持上游缺省的 collation，同步可能会报错。

- **在 DM 中优化 `transfer source`，支持平滑执行同步任务**

    当 DM-worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 `source` 配置迁移到其他节点。优化后的 `transfer source` 简化了用户操作步骤，不再要求先暂停所有关联 task 而是直接执行平滑迁移，DM 将在内部完成所需操作。

- **DM OpenAPI 特性 GA**

    DM 支持通过 API 的方式进行日常管理，包括增加数据源、管理任务等。本次更新 OpenAPI 从实验特性转为正式特性。

### 问题诊断效率

- **Top SQL（实验特性）**

>     新推出实验性特性 Top SQL（默认关闭），帮助用户轻松找到节点中负载贡献较大的查询。

Reviewer 11: ```suggestion
    新推出实验性特性 Top SQL（默认关闭），帮助用户轻松找到节点中消耗负载较大的查询。
```

在使用 GBK 字符集时，需要注意兼容性限制，详情参考[字符集和排序 - GBK](/character-set-gbk.md)。

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

### 性能

- **新增 Raft Engine（实验特性）**

    支持使用 [Raft Engine](https://github.com/tikv/raft-engine) 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。此外，Raft Engine 提升了日志回收效率，修复了极端条件下日志堆积的问题。

    Raft Engine 目前仍属于实验特性，并默认关闭。另外请注意 v5.4.0 版本的 Raft Engine 数据格式与之前版本不兼容，对集群做升级或者降级操作之前，需要确保所有 TiKV 节点上的 Raft Engine 已被关闭。只建议在 v5.4.0 及以后的版本使用 Raft Engine。

    [用户文档](/tikv-configuration-file.md#raft-engine)

- **持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能**

    - 支持将更多函数下推至 MPP 引擎
        - 字符串函数：`LPAD()`、`RPAD()`、`STRCMP()`
        - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
    - 引入动态线程池，提升资源利用率（实验特性）
    - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性

>     - 提升由行存到列存数据同步处理时对 raft log 的解码 (decoding) 效率，使 CPU 使用率降低达 90%

Reviewer 20: ```suggestion
    - 提升由行存到列存数据同步处理时对 raft log 的解码 (decoding) 效率，数据同步性能提升 50%
```

>     - TiFlash 调整了文件系统中 Delta Tree 相关的默认参数，使之更适合一般的生产环境

Reviewer 20: ```suggestion
```
这一项已经被上面 “新增或修改一些 TiFlash 已有配置的默认值” 包括了

    ```sql
    SELECT * FROM table WHERE key1 <= 100 OR key2 = 200;
    ```

    以往 TiDB 在一个表上的查询只能使用一个索引，无法同时使用多个索引进行条件过滤。相较以往，_索引合并_ 避免了此情况下可能不必要的大量数据扫描，也可以使得需要灵活查询不特定多列数据组合的用户利用单列上的索引达到高效稳定的查询，无需大量构建多列复合索引。

    本版本正式发布了 _索引合并 (Index Merge)_ 特性，但仍存在以下的使用条件和限制（详情请参见[用户文档](/explain-index-merge.md)）：

    目前 TiDB 的 _索引合并_ 优化只限于 _析取范式_ (X<sub>1</sub> ⋁ X<sub>2</sub> ⋁ …X<sub>n</sub>)，即 `WHERE` 子句中过滤条件连接词为 `OR`。

    如果全新部署的集群版本为 v5.4.0 或以上，此特性默认开启。如果从 v5.4.0 以前的版本升级到 v5.4.0 或以上，默认保持升级前此特性的开关状态（v4.0.0 之前无此项特性的版本默认关闭），由用户决定是否开启。

- **支持收集部分列的统计信息（实验特性）**
    
    执行 SQL 语句时，优化器在大多数情况下只会用到部分列（例如， `WHERE`、`JOIN`、`ORDER BY`、`GROUP BY` 子句中用到的列）的统计信息。这些被优化器用到的列称为 `PREDICATE COLUMNS`。
    
    从 v5.4.0 开始，TiDB 引入了收集部分列的统计信息的特性（默认关闭），支持只收集指定列或者 `PREDICATE COLUMNS` 的统计信息，极大地降低了收集统计信息的开销。
    
    [用户文档](/statistics.md#收集部分列的统计信息)
    
- **支持统计信息的同步加载（实验特性）**

    从 v5.4.0 开始，TiDB 引入了统计信息同步加载的特性（默认关闭），支持执行当前 SQL 语句时将直方图、TopN、CMSketch 等占用空间较大的统计信息同步加载到内存，提高该 SQL 语句优化时统计信息的完整性。
    
    [用户文档](/statistics.md#统计信息的加载)

> - **优化 RocksDB 在大数据量下的锁争用问题**
> 
>     RocksDB 依赖全局锁来维护多版本的文件信息，并当销毁读快照时对过期文件进行清理。在文件数较多且读写频繁的场景下，这一操作显著影响前台延迟。本版本针对这一问题进行优化，详见 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)。
> 

Author: ```suggestion
```

    - 支持 `ADMIN {SESSION | INSTANCE | GLOBAL} PLAN_CACHE` 语法，用于清空缓存的查询计划 [#30370](https://github.com/pingcap/tidb/pull/30370)

+ TiKV

    - Coprocessor 支持分页 API 进行流式处理 [#11448](https://github.com/tikv/tikv/issues/11448)
    - 支持 `read-through-lock`，使读操作不需要等待清理 secondary lock [#11402](https://github.com/tikv/tikv/issues/11402)
    - 增加了磁盘保护机制，尽量避免磁盘空间耗尽导致 panic [#10537](https://github.com/tikv/tikv/issues/10537)
    - 日志支持存档和轮替 [#11651](https://github.com/tikv/tikv/issues/11651)
    - 减少 Raft 客户端的系统调用并提高 CPU 效率 [#11309](https://github.com/tikv/tikv/issues/11309)
    - Coprocessor 支持下推 substring 到 TiKV [#11495](https://github.com/tikv/tikv/issues/11495)
    - 通过跳过读锁提高在 RC 隔离级别中扫描的性能 [#11485](https://github.com/tikv/tikv/issues/11485)
    - 减少备份使用的默认线程池大小，并在压力大时限制其使用 [#11000](https://github.com/tikv/tikv/issues/11000)
    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)
    - 优化在文件数较多且读写频繁的场景下 RocksDB 的全局锁争用问题 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)

+ PD

+ TiFlash

+ Tools

    + Backup & Restore (BR)

>     + TiCDC

Reviewer 21: ## Bug 修复

* 修复当 `min.insync.replicas` 小于 `replication-factor` 时不能同步的问题 [#3994](https://github.com/pingcap/tiflow/issues/3994)
* 修复 cached region 监控指标为负数的问题 [#4300](https://github.com/pingcap/tiflow/issues/4300)
* 修复 mq sink write row 没有监控数据的问题 [#3431](https://github.com/pingcap/tiflow/issues/3431)
* 修复 sql mode 兼容性问题 [#3810](https://github.com/pingcap/tiflow/issues/3810)
* 修复在移除同步任务后潜在的 panic 问题 [#3128](https://github.com/pingcap/tiflow/issues/3128)
* 修复输出默认列值时的 panic 问题和数据不一致问题 [#3929](https://github.com/pingcap/tiflow/issues/3929)
* 修复不支持同步默认值的问题 [#3793](https://github.com/pingcap/tiflow/issues/3793)
* 修复潜在的同步流控死锁问题 [#4055](https://github.com/pingcap/tiflow/issues/4055)
* 修复在磁盘写满时无日志输出的问题 [#3362](https://github.com/pingcap/tiflow/issues/3362)
* 修复 DDL 特殊注释导致的同步停止的问题 [#3755](https://github.com/pingcap/tiflow/issues/3755)
* 修复在某些 RHEL 发行版上因时区问题导致服务无法启动的问题 [#3584](https://github.com/pingcap/tiflow/issues/3584)
* 修复因 checkpoint 不准确导致的潜在的数据丢失问题 [#3545](https://github.com/pingcap/tiflow/issues/3545)
* 修复在容器环境中 OOM 的问题 [#1798](https://github.com/pingcap/tiflow/issues/1798)
* 修复 config.Metadata.Timeout 没有正确配置而导致的同步停止问题 [#3352](https://github.com/pingcap/tiflow/issues/3352)

## 提升改进

* 减少 "EventFeed retry rate limited" 日志 [#4006](https://github.com/pingcap/tiflow/issues/4006)
* 降低在同步大量表时的同步延时 [#3900](https://github.com/pingcap/tiflow/issues/3900)
* 减低 TiKV 节点宕机后 kvclient 恢复的时间 [#3191](https://github.com/pingcap/tiflow/issues/3191)

Reviewer 21: 附上英语版本

## Bug 修复

* Fix the problem that TiCDC cannot send messages when `min.insync.replicas` is less than `replication-factor` [#3994](https://github.com/pingcap/tiflow/issues/3994)
* Fix kv client cached region metric could be negative. [#4300](https://github.com/pingcap/tiflow/issues/4300)
* Fix mq sink write row count metrics. [#3431](https://github.com/pingcap/tiflow/issues/3431)
* Fix issues about sql mode compatibility for mysql sink. [#3810](https://github.com/pingcap/tiflow/issues/3810)
* Fix the potential panic issue that occurs when changefeed info is removed from etcd. [#3128](https://github.com/pingcap/tiflow/issues/3128)
* Fix column default value panic and data inconsistency. [#3929](https://github.com/pingcap/tiflow/issues/3929)
* Fix mounter default date value not support [#3793](https://github.com/pingcap/tiflow/issues/3793)
* Fix a bug that can cause changefeed stuck due to a deadlock occurs. [#4055](https://github.com/pingcap/tiflow/issues/4055)
* Fix TiCDC doesn't report any error when log disk is full. [#3362](https://github.com/pingcap/tiflow/issues/3362)
* Fix syntax error if DDL has a special comment. [#3755](https://github.com/pingcap/tiflow/issues/3755)
* Fix timezone related error that cause cdc server can't run in some RHEL release version (6.8, 6.9 etc). [#3584](https://github.com/pingcap/tiflow/issues/3584)
* Fix data lose due to inaccurate checkpoints. [#3545](https://github.com/pingcap/tiflow/issues/3545)
* Fix OOM in container environment [#1798](https://github.com/pingcap/tiflow/issues/1798)
* Set config.Metadata.Timeout correctly to prevent stuck data synchronization. [#3352](https://github.com/pingcap/tiflow/issues/3352)

## 提升改进

* Reduce "EventFeed retry rate limited" logs [#4006](https://github.com/pingcap/tiflow/issues/4006)
* Reduce checkpoint lag when capturing many tables. [#3900](https://github.com/pingcap/tiflow/issues/3900)
* Reduce kvclient recover time when TiKV is down. [#3191](https://github.com/pingcap/tiflow/issues/3191)

    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)

+ PD

+ TiFlash

+ Tools

    + Backup & Restore (BR)

    + TiCDC

    + TiDB Data Migration (DM)

    + TiDB Lightning

    + Dumpling

    + TiDB Binlog

## Bug 修复

+ TiDB

    - 修复当从 v4.x 版本升级到 v5.x 版本后 `tidb_analyze_version` 的值变化的问题 [#25422](https://github.com/pingcap/tidb/issues/25422)
    - 修复在子查询中使用不同的 collation 导致查询结果错误的问题 [#30748](https://github.com/pingcap/tidb/issues/30748)
    - 修复 `concat(ifnull(time(3))` 的结果与 MySQL 不一致的问题 [#29498](https://github.com/pingcap/tidb/issues/29498)
    - 修复乐观事务下数据索引可能不一致的问题 [#30410](https://github.com/pingcap/tidb/issues/30410)
    - 修复当表达式不能下推给 TiKV 时 IndexMerge 查询计划错误的问题 [#30200](https://github.com/pingcap/tidb/issues/30200)
    - 修复并发的列类型变更导致 schema 与数据不一致的问题 [#31048](https://github.com/pingcap/tidb/issues/31048)
    - 修复当有子查询时 IndexMerge 的查询结果错误的问题 [#30913](https://github.com/pingcap/tidb/issues/30913)
    - 修复当客户端设置过大的 FetchSize 后 TiDB 执行会 panic 的问题 [#30896](https://github.com/pingcap/tidb/issues/30896)

> 

Author: ```suggestion
```

+ 支持统计信息采集配置持久化
+ 支持使用 Raft Engine 作为 TiKV 的日志存储引擎
+ 优化备份对集群的影响
+ 支持 Azure Blob Storage 作为备份目标存储（实验特性）
+ 持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能
+ 为 TiDB Lightning 增加已存在数据表是否允许导入的开关
+ 优化持续性能分析实验特性
+ TiSpark 支持用户认证与鉴权

## 兼容性变化

> **注意：**
>
> 当从一个早期的 TiDB 版本升级到 TiDB v5.4.0 时，如需了解所有中间版本对应的兼容性更改说明，请查看对应版本的 [Release Notes](/releases/release-notes.md)。

### 系统变量

|  变量名    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- |
|  `tidb_backoff_lock_fast` | 修改 | 默认值由 `100` 修改为 `10` |
|  [`tidb_enable_column_tracking`](/system-variables.md#tidb_enable_column_tracking-从-v540-版本开始引入) | 新增 | 用于控制是否开启 TiDB 对 `PREDICATE COLUMNS` 的收集，默认值为 `OFF` |
| `tidb_enable_index_merge` | 修改 | 默认值由 `OFF` 改为 `ON`。如果从低于 v4.0.0 版本升级到 v5.4.0 及以上版本的集群，该变量值默认保持 `OFF`。如果从 v4.0.0 及以上版本升级到 v5.4.0 及以上版本的集群，该变量开关保持升级前的状态。对于 v5.4.0 及以上版本的新建集群，该变量开关默认保持 `ON`。 |
| [`tidb_enable_paging`](/system-variables.md#tidb_enable_paging-从-v540-版本开始引入)  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能会出现读请求的延迟高、TiKV 的 Unified read pool CPU 使用率高的情况。在这种情况下，由于 `Limit` 算子只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，从而降低延迟、减少资源消耗。 |

> | `tidb_read_staleness` | 新增 | 用于设置当前会话允许读取的历史数据时间范围，默认值为 `0` |

Reviewer 22: lgtm

> | `tidb_read_staleness` | 新增 | 用于设置当前会话允许读取的历史数据时间范围，默认值为 `0` |

Reviewer 19: ```suggestion
| `tidb_read_staleness` | 新增 | 用于设置当前会话允许读取的历史数据范围，默认值为 `0` |
```

| TiKV | `raftstore.raft-log-gc-tick-interval` | 修改 | 默认值修改为 `3s` |
| TiKV  |  `log-rotation-timespan`  | 删除 |  轮换日志的时间跨度。当超过该时间跨度，日志文件会被轮换，即在当前日志文件的文件名后附加一个时间戳，并创建一个新文件。 |
| PD | `log.level` | 修改 | 默认值由 "INFO" 改为 "info"，保证大小写不敏感 |
| TiFlash | `profile.default.enable_elastic_threadpool` | 新增  |  表示是否启用可自动扩展的线程池。打开该配置项可以显著提高 TiFlash 在高并发场景的 CPU 利用率。默认值为 `false`。|
| TiDB Data Migration (DM) | [`collation_compatible`](/dm/task-configuration-file-full.md#完整配置文件示例) | 同步 CREATE 语句中缺省 Collation 的方式，可选 "loose" 和 "strict"，默认为 "loose"。 |
| TiFlash | [`storage.format_version`](/tiflash/tiflash-configuration.md#配置文件-tiflashtoml) | 新增 | 表示 DTFile 储存文件格式，默认值为 `2`，该格式在数据文件中内嵌哈希值。也可以设置为 `3`，该格式包含元数据，标记数据校验，支持多种哈希算法。|
| TiFlash | `logger.count` | 修改 | 默认值修改为 `10` |
| TiFlash | `status.metrics_port` | 修改 | 默认值修改为 `8234` |
| TiFlash | `raftstore.apply-pool-size` | 新增 | 处理 Raft 数据落盘的线程池中线程的数量，默认值为 `4`。 |
| TiFlash | `raftstore.store-pool-size` | 新增 | 处理 Raft 的线程池中线程的数量，即 Raftstore 线程池的大小，默认值为 `4`。 |

> | TiCDC | `enable-tidb-extension` | 新增 |  开启该配置项后，TiCDC 在 Canal-JSON 协议格式中附加 TiDB 扩展字段。                  |

Reviewer 12: TiCDC不算是配置文件参数。需要考虑放哪里。@赵一霖

Author: 要是不确定放哪里，可否先删除？
```suggestion
```

> | TiCDC | `max-message-bytes` | 修改 | 将 Kafka sink 模块的 `max-message-bytes` 默认值设置为 `10M`  |

Reviewer 12: ```suggestion
| TiCDC | `max-message-bytes` | 修改 | 将 Kafka sink 模块的 `max-message-bytes` 默认值设置为 `10M`  |
```
@hi-rustin 麻烦确认下默认值。

Reviewer 23: Done!

| TiCDC | `partition-num`      | 修改 | 将 Kafka Sink `partition-num` 的默认值改由 `4` 为 `3`，使 TiCDC 更加平均地分发消息到各个 Kafka partition |

> | TiDB Lightning | `meta-schema-name` |  新增  | 在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认值为 "lightning_metadata"  |

Reviewer 24: `meta-shcema-name` 这个 5.3 就 ga 了。 5.4 对应 duplication detect 的新增参数应该是 `task-info-schema-name`

Reviewer 12: 删掉这行，添加到 v5.3 Release Notes 中。辛苦提个PR改下。

Author: ```suggestion
```

> | TiDB Lightning | `meta-schema-name` |  新增  | 在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认值为 "lightning_metadata"  |

Reviewer 24: ```suggestion
| TiDB Lightning | `task-info-schema-name` |  新增  | 用于配置当 Lightning 检测到冲突数据时，对应冲突数据存储的库名，默认值为 "lightning_task_info"。如果没有开启冲突检测功能，则无需设置此配置。  |
```

| TiDB Lightning | `incremental-import` | 新增 | 是否允许向已存在数据的表导入数据。默认值为 false |

### 其他

- TiDB Dashboard 默认不再使用 `root` + 空密码登录。

    从 v5.4.0开始，使用 TiUP 启动集群时推荐使用 `start --initial`。执行该操作启动集群后，会为 `root` 账号自动生成一个随机密码，`root` 账号登录 Dashboard 需要使用这个密码。

- 为 TiDB 和 PD 之间新增接口。使用 `information_schema.TIDB_HOT_REGIONS_HISTORY` 系统表时，TiDB 需要使用匹配的 PD 版本。
- TiDB Server、PD Server 和 TiKV Server 将采用统一的参数命名方式来管理日志命名、输出格式、轮转和过期的规则。参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

    在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。为了更好的支持中文用户，TiDB 从 v5.4.0 版本开始支持 GBK 字符集，同时支持 `gbk_bin` 和 `gbk_chinese_ci` 两种排序规则（在使用这两种排序规则前需要开启 TiDB 配置项 `new_collations_enabled_on_first_bootstrap`）。

>     在使用 GBK 字符集时，请注意以下兼容性限制：
> 
>     - TiFlash 暂不支持 GBK 字符。
>     - TiCDC 暂不支持 GBK 字符。
>     - 如果 `character_set_client` 和 `character_set_connection` 都是 `gbk` 时，处理非法 GBK 字符与 MySQL 存在兼容性问题。
>     - TiDB 不支持 `_gbk"xxx"` 的用法，但是支持 `_utf8mb4"xxx"` 的用法。而 MySQL 对于 `_charset"xxx"` 的用法都支持。
>     - TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.4.0 之前不支持恢复 `charset=GBK` 的表。

Author: ```suggestion
在使用 GBK 字符集时，需要注意兼容性限制，详情参考[字符集和排序 - GBK](/character-set-gbk.md)。
```

    由于 Raft Engine 涉及数据格式改动，目前仍属于实验特性，并默认关闭。同时请注意最新的 Raft Engine 不与 v5.4.0 版本前的 Raft Engine 兼容。因此在进行跨越 v5.4.0 版本的升级和降级之前，需要确保已有 TiKV 节点上的 Raft Engine 已被关闭。

    [用户文档](/tikv-configuration-file.md#raft-engine)

- **持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能**

    - 支持将更多函数下推至 MPP 引擎
        - 字符串函数：`LPAD()`、`RPAD()`、`STRCMP()`
        - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
    - 引入动态线程池，提升资源利用率（实验特性）
    - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性
    - 提升由行存到列存数据同步处理时对 raft log 的解码 (decoding) 效率，使 CPU 使用率降低达 90%
    - TiFlash 调整了文件系统中 Delta Tree 相关的默认参数，使之更适合一般的生产环境

- **通过 session 变量实现有界限过期数据读取**

    TiDB 是基于 Raft 协议的多副本分布式数据库。面对高并发，高吞吐业务场景，可以通过follower 节点实现读性能扩展，构建读写分离架构。

    针对不同的业务场景，follower 提供强一致读和弱一致过期读两种读模式。强一致读满足数据实时性要求严格的业务场景，但是因为 leader 和 follower 的数据同步延迟、吞吐较低、延迟较高，特别是在跨机房架构下延迟问题被进一步放大。

    在对数据实时性要求不高的业务场景下，可以选择过期读模式。使用该模式可以降低延迟和提升吞吐。TiDB 目前支持通过显示只读事务或 SQL 语句的方式实现过期读。两种方式均支持指定时间的精确过期读和指定时间边界的过期读两种模式，详细用法请参考[过期读文档](/read-historical-data.md)。

    从 v5.4.0 版本开始 TiDB 支持通过 session 变量设置有界限过期读，进一步提升易用性，具体设置如下：

    ```sql
    set @@tidb_replica_read=leader_and_follower
    set @@tidb_read_staleness="-5"
    ```

>     通过该设置，可以实现就近选取 leader 或 follower 节点，并读取 5 秒钟前的最新过期数据，满足准实时场景下低延迟高吞吐数据访问的业务诉求，降低研发门槛，提升易用性。

Author: @Yisaer PTAL, thanks

- **支持收集部分列的统计信息（实验特性）**
    
    执行 SQL 语句时，优化器在大多数情况下只会用到部分列（例如， `WHERE`、`JOIN`、`ORDER BY`、`GROUP BY` 子句中用到的列）的统计信息。这些被优化器用到的列称为 `PREDICATE COLUMNS`。
    
    从 v5.4.0 开始，TiDB 引入了收集部分列的统计信息的特性（默认关闭），支持只收集指定列或者 `PREDICATE COLUMNS` 的统计信息，极大地降低了收集统计信息的开销。
    
    [用户文档](/statistics.md#收集部分列的统计信息)
    
- **支持统计信息的同步加载（实验特性）**

    从 v5.4.0 开始，TiDB 引入了统计信息同步加载的特性（默认关闭），支持执行当前 SQL 语句时将直方图、TopN、CMSketch 等占用空间较大的统计信息同步加载到内存，提高该 SQL 语句优化时统计信息的完整性。
    
    [用户文档](/statistics.md#统计信息的加载)

- **优化 RocksDB 在大数据量下的锁争用问题**

    RocksDB 依赖全局锁来维护多版本的文件信息，并当销毁读快照时对过期文件进行清理。在文件数较多且读写频繁的场景下，这一操作显著影响前台延迟。本版本针对这一问题进行优化，详见 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)。

### 稳定性

- **支持统计信息采集配置持久化**

    统计信息是优化器生成执行计划时所参考的基础信息之一，统计信息的准确性直接影响生成的执行计划是否合理。为了保证统计信息的准确性，有时候需要针对不同的表、分区、索引设置不同的采集配置项。

    TiDB 从 v5.4.0 版本开始支持 `ANALYZE` 配置持久化功能，方便后续收集统计信息时沿用已有配置项。
    
    `ANALYZE` 配置持久化功能默认开启（系统变量 `tidb_analyze_version = 2` 且 `tidb_persist_analyze_options = true`），用于记录手动执行 `ANALYZE` 语句时指定的持久化配置项。记录后，当 TiDB 下一次自动更新统计信息或者你手动收集统计信息但未指定配置项时，TiDB 会按照记录的配置项收集统计信息。
    
    [用户文档](/statistics.md#analyze-配置持久化)

## 高可用和容灾

- **优化备份对集群的影响**

>     Backup & Restore (BR) 增加了备份线程自动调节功能。该功能通过监控集群资源的使用率自动调节备份的线程数的方式，降低备份过程对集群的影响。在某些 Case 验证中，通过增加集群用于备份的资源和开启备份线程自动调节功能，备份的影响可以降低到 10% 以下。

Reviewer 19: ```suggestion
    Backup & Restore (BR) 增加了备份线程自动调节功能（默认开启）。该功能通过监控集群资源的使用率自动调节备份的线程数的方式，降低备份过程对集群的影响。在某些 Case 验证中，通过增加集群用于备份的资源和开启备份线程自动调节功能，备份的影响可以降低到 10% 以下。
```

> 
> 

Reviewer 19: ```suggestion

```

>     该功能默认开启，但是如果你在离线环境中进行备份，可以关闭该功能来获得更高的备份速度。

Reviewer 12: ```suggestion
```

    详细文档请阅读 [BR 自动调节](/br/br-features.md#自动调节-从-v54-版本开始引入)。

- **支持 Azure Blob Storage 作为备份目标存储（实验特性）**

    Backup & Restore (BR) 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

    该功能目前是实验特性，详细情况参考 [BR 支持 Azure Blob Storage 远端存储](/br/backup-and-restore-azblob.md)。

### 数据迁移

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

> - **为 TiDB Lightning 增加新参数，用于在并行导入模式下保存各个 TiDB Lightning 实例元信息的 schema 名字 **

Reviewer 12: @glorv 辛苦挪到5.3。然后把新增的参数添加到这里。

Author: @sunzhaoyang PTAL, thanks!

     为 TiDB Lightning 增加 `meta-schema-name` 参数。在并行导入模式下，该参数用于在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认值为 "lightning_metadata"。对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。

> - **在 TiDB Lightning 中添加重复数据的检测（实验特性）**

Reviewer 12: ```suggestion
- **在 TiDB Lightning 中添加重复数据的检测**
```

    在 `backend=local` 模式下，数据导入完成之前 TiDB Lightning 会输出冲突数据，然后从数据库中删除这些冲突数据。用户可以在导入完成后解析冲突数据，并根据业务规则选择适合的数据进行插入。建议根据冲突数据清洗上游数据源，避免在后续增量数据迁移阶段遇到冲突数据而造成数据不一致。

- **在 TiDB Data Migration (DM) 中 优化 relay log 的使用方式**

    - 恢复 `source` 配置中 `enable-relay` 开关
    - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能
    - relay log 的开启状态与 `source` 绑定，source 迁移到任意 DM-worker 均保持原有开启或关闭状态
    - relay log 的存放路径移至 DM-worker 配置文件

- **在 DM 中优化排序规则的处理方式**

>     增加 `collation_compatible` 开关，支持 `strict` 和 `loose`（默认）两种模式。如果对排序规则要求不严格，允许排序规则在上下游不一致，使用默认的 `loose` 模式可使同步正常进行；如果对排序规则要求严格，排序规则在上下游不一致会导致业务报错，则可以使用 `strict` 模式。

Reviewer 15: release notes 和[用户文档](https://github.com/pingcap/docs-cn/pull/8094/files)在 **“strict”模式的描述** 上有点出入。 cc @sunzhaoyang 

`同步 CREATE 语句中缺省 Collation 的方式，可选 "loose" 和 "strict"，默认为 "loose"。"loose" 模式不会显式补充上游缺省的 Collation，"strict" 会显式补充上游缺省的 Collation。当使用 "strict" 模式，但下游不支持上游缺省的 Collation 时，下游可能会报错。`

+ 支持 Azure Blob Storage 作为备份目标存储（实验特性）
+ 持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能
+ 为 TiDB Lightning 增加已存在数据表是否允许导入的开关
+ 优化持续性能分析实验特性
+ TiSpark 支持用户认证与鉴权

## 兼容性变化

> **注意：**
>
> 当从一个早期的 TiDB 版本升级到 TiDB v5.4.0 时，如需了解所有中间版本对应的兼容性更改说明，请查看对应版本的 [Release Notes](/releases/release-notes.md)。

### 系统变量

|  变量名    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- |
|  `tidb_backoff_lock_fast` | 修改 | 默认值由 `100` 修改为 `10` |
|  [`tidb_enable_column_tracking`](/system-variables.md#tidb_enable_column_tracking-从-v540-版本开始引入) | 新增 | 用于控制是否开启 TiDB 对 `PREDICATE COLUMNS` 的收集，默认值为 `OFF` |
| `tidb_enable_index_merge` | 修改 | 默认值由 `OFF` 改为 `ON`。如果从低于 v4.0.0 版本升级到 v5.4.0 及以上版本的集群，该变量值默认保持 `OFF`。如果从 v4.0.0 及以上版本升级到 v5.4.0 及以上版本的集群，该变量开关保持升级前的状态。对于 v5.4.0 及以上版本的新建集群，该变量开关默认保持 `ON`。 |
| [`tidb_enable_paging`](/system-variables.md#tidb_enable_paging-从-v540-版本开始引入)  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能会出现读请求的延迟高、TiKV 的 Unified read pool CPU 使用率高的情况。在这种情况下，由于 `Limit` 算子只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，从而降低延迟、减少资源消耗。 |
| `tidb_read_staleness` | 新增 | 用于设置当前会话允许读取的历史数据时间范围，默认值为 `0` |

> | [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `0` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时。 |
> | [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败(`false`) 还是退回使用 pseudo 的统计信息 (`true`) |

Reviewer 11: ```suggestion
| [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `OFF` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时。 |
| [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败 (`OFF`) 还是退回使用 pseudo 的统计信息 (`ON`)，默认值为 `ON` |
```

> | [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启[ ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性 |

Reviewer 11: ```suggestion
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启 [ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性，默认值为 `OFF` |
```

| `tidb_store_limit` | 修改 | v5.4.0 前支持实例级别及集群级别的设置，现在只支持集群级别的设置 |

### 配置文件参数

|  配置文件    |  配置项    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- | :----------- |
| TiDB | [`stats-load-concurrency`](/tidb-configuration-file.md#stats-load-concurrency-从-v540-版本开始引入) | 新增 |  用于设置 TiDB 统计信息同步加载功能可以并发处理的最大列数，默认值为 `5`             |
| TiDB | [`stats-load-queue-size`](/tidb-configuration-file.md#stats-load-queue-size-从-v540-版本开始引入)   | 新增 |  用于设置 TiDB 统计信息同步加载功能最多可以缓存多少列的请求，默认值为 `1000`             |

> | TiKV | [`backup.enable-auto-tune`](/tikv-configuration-file.md#enable-auto-tune-从-v54-版本开始引入) | 修改 | 在 v5.3.0 中默认值为 `false`，自 v5.4.0 起默认值改为 `true`。表示在集群资源占用率较高的情况下，是否允许 BR 自动限制备份使用的资源，减少对集群的影响。在默认配置下，备份速度可能下降。 |
> | TiKV | `log.file.max-size`、`log.file.max-days`、`log.file.max-backups` | 新增  | 参数说明见 [TiKV 配置文件 - log.file](/tikv-configuration-file.md#logfile-从-v540-版本开始引入)。 |
> | TiKV | `log-level`、`log-format`、`log-file`、`log-rotation-size` | 修改 | 将 TiKV log 参数名替换为与 TiDB log 参数一致的参数名，即 `log.level`、`log.format`、`log.file.filename`、`log.enable-timestamp`。如果只设置了原参数、且把其值设为非默认值，原参数与新参数会保持兼容；如果同时设置了原参数和新参数，则会使用新参数。详情参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。 |
> | TiKV  |  `log-rotation-timespan`  | 删除 |  轮换日志的时间跨度。当超过该时间跨度，日志文件会被轮换，即在当前日志文件的文件名后附加一个时间戳，并创建一个新文件。 |
> | TiKV | `raft-engine` | 新增 | 包含 `enable`、`dir`、`batch-compression-threshold`、`bytes-per-sync`、`target-file-size`、`purge-threshold`、`recovery-mode`、`recovery-read-block-size`、`recovery-read-block-size`、`recovery-threads`，详情参见 [TiKV 配置文件：raft-engine](/tikv-configuration-file.md#raft-engine)。|
> | TiKV | `snap-generator-pool-size` | 新增 | `snap-generator` 线程池大小，默认值为 `2` |
> | TiKV | `raftstore.raft-log-gc-tick-interval` | 修改 | 默认值修改为 `3s` |
> | TiKV | `raftstore.raft-log-compact-sync-interval` | 新增 | 控制 CompactLog 命令的间隔，默认值为 `2s` |

Reviewer 19: ```suggestion
| TiKV | `snap-generator-pool-size` | 新增 | `snap-generator` 线程池大小，默认值为 `2` |
| TiKV | `log.file.max-size`、`log.file.max-days`、`log.file.max-backups` | 新增  | 参数说明见 [TiKV 配置文件 - log.file](/tikv-configuration-file.md#logfile-从-v540-版本开始引入)。 |
| TiKV | `raft-engine` | 新增 | 包含 `enable`、`dir`、`batch-compression-threshold`、`bytes-per-sync`、`target-file-size`、`purge-threshold`、`recovery-mode`、`recovery-read-block-size`、`recovery-read-block-size`、`recovery-threads`，详情参见 [TiKV 配置文件：raft-engine](/tikv-configuration-file.md#raft-engine)。|
| TiKV | `raftstore.raft-log-compact-sync-interval` | 新增 | 控制 CompactLog 命令的间隔，默认值为 `2s` |
| TiKV | [`backup.enable-auto-tune`](/tikv-configuration-file.md#enable-auto-tune-从-v54-版本开始引入) | 修改 | 在 v5.3.0 中默认值为 `false`，自 v5.4.0 起默认值改为 `true`。表示在集群资源占用率较高的情况下，是否允许 BR 自动限制备份使用的资源，减少对集群的影响。在默认配置下，备份速度可能下降。 |
| TiKV | `log-level`、`log-format`、`log-file`、`log-rotation-size` | 修改 | 将 TiKV log 参数名替换为与 TiDB log 参数一致的参数名，即 `log.level`、`log.format`、`log.file.filename`、`log.enable-timestamp`。如果只设置了原参数、且把其值设为非默认值，原参数与新参数会保持兼容；如果同时设置了原参数和新参数，则会使用新参数。详情参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。 |
| TiKV | `raftstore.raft-log-gc-tick-interval` | 修改 | 默认值修改为 `3s` |
| TiKV  |  `log-rotation-timespan`  | 删除 |  轮换日志的时间跨度。当超过该时间跨度，日志文件会被轮换，即在当前日志文件的文件名后附加一个时间戳，并创建一个新文件。 |
```

> | PD | `log.level` | 修改 | 默认值由 "INFO" 改为 "info"，保证大小写不敏感 |
> | PD | `hot-regions-write-interval` | 新增 |  设置 PD 存储 Hot Region 信息的时间间隔。默认值为 `10m`。 |
> | PD | `hot-regions-reserved-days` | 新增 | 设置 PD 保留的 Hot Region 信息的最长时间。默认值为 `7`。

Reviewer 19: ```suggestion
| PD | `hot-regions-write-interval` | 新增 |  设置 PD 存储 Hot Region 信息的时间间隔。默认值为 `10m`。 |
| PD | `hot-regions-reserved-days` | 新增 | 设置 PD 保留的 Hot Region 信息的最长时间。默认值为 `7`。
| PD | `log.level` | 修改 | 默认值由 "INFO" 改为 "info"，保证大小写不敏感 |
```

> | PD | `hot-regions-write-interval` | 新增 |  设置 PD 存储 Hot Region 信息的时间间隔。默认值为 `10m`。 |
> | PD | `hot-regions-reserved-days` | 新增 | 设置 PD 保留的 Hot Region 信息的最长时间。默认值为 `7`。
> | PD | `log.level` | 修改 | 默认值由 "INFO" 改为 "info"，保证大小写不敏感 |

Reviewer 19: ```suggestion
| PD | `log.level` | 修改 | 默认值由 "INFO" 改为 "info"，保证大小写不敏感 |
```

| TiFlash | `logger.count` | 修改 | 默认值修改为 `10` |
| TiFlash | `status.metrics_port` | 修改 | 默认值修改为 `8234` |
| TiFlash | `raftstore.apply-pool-size` | 新增 | 处理 Raft 数据落盘的线程池中线程的数量，默认值为 `4`。 |
| TiFlash | `raftstore.store-pool-size` | 新增 | 处理 Raft 的线程池中线程的数量，即 Raftstore 线程池的大小，默认值为 `4`。 |
| TiCDC | `enable-tidb-extension` | 新增 |  开启该配置项后，TiCDC 在 Canal-JSON 协议格式中附加 TiDB 扩展字段。                  |
| TiCDC | `max-message-bytes` | 修改 | 将 Kafka sink 模块的 `max-message-bytes` 默认值设置为 `10M`  |
| TiCDC | `partition-num`      | 修改 | 将 Kafka Sink `partition-num` 的默认值改由 `4` 为 `3`，使 TiCDC 更加平均地分发消息到各个 Kafka partition |
| TiDB Lightning | `meta-schema-name` |  新增  | 在并行导入模式下，在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名字，默认值为 "lightning_metadata"  |
| TiDB Lightning | `incremental-import` | 新增 | 是否允许向已存在数据的表导入数据。默认值为 false |

### 其他

- TiDB Dashboard 默认不再使用 `root` + 空密码登录。

    从 v5.4.0开始，使用 TiUP 启动集群时推荐使用 `start --initial`。执行该操作启动集群后，会为 `root` 账号自动生成一个随机密码，`root` 账号登录 Dashboard 需要使用这个密码。

- 为 TiDB 和 PD 之间新增接口。使用 `information_schema.TIDB_HOT_REGIONS_HISTORY` 系统表时，TiDB 需要使用匹配的 PD 版本。

> - TiDB Server、PD Server 和 TiKV Server 将采用统一的参数命名方式来管理日志命名、输出格式、轮转和过期的规则。参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。

Author: PD 和 TiDB 有待确认

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

    在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。为了更好的支持中文用户，TiDB 从 v5.4.0 版本开始支持 GBK 字符集，同时支持 `gbk_bin` 和 `gbk_chinese_ci` 两种排序规则（在使用这两种排序规则前需要开启 TiDB 配置项 `new_collations_enabled_on_first_bootstrap`）。

    在使用 GBK 字符集时，请注意以下兼容性限制：

    - TiFlash 暂不支持 GBK 字符。
    - TiCDC 暂不支持 GBK 字符。
    - 如果 `character_set_client` 和 `character_set_connection` 都是 `gbk` 时，处理非法 GBK 字符与 MySQL 存在兼容性问题。
    - TiDB 不支持 `_gbk"xxx"` 的用法，但是支持 `_utf8mb4"xxx"` 的用法。而 MySQL 对于 `_charset"xxx"` 的用法都支持。
    - TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.4.0 之前不支持恢复 `charset=GBK` 的表。

> 

Author: 兼容性细节链接到用户文档，这里不用写太xiangxi

>     - TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.3.0 之前不支持恢复 `charset=GBK` 的表。

Reviewer 17: ```suggestion
    - TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.4.0 之前不支持恢复 `charset=GBK` 的表。
```

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

### 性能

> - **新增 Raft Engine（实验特性）**

Author: 问 Thomas

    支持使用 [Raft Engine](https://github.com/tikv/raft-engine) 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。

    由于 Raft Engine 涉及数据格式改动，目前仍属于实验特性，并默认关闭。同时请注意最新的 Raft Engine 不与 v5.4.0 版本前的 Raft Engine 兼容。因此在进行跨越 v5.4.0 版本的升级和降级之前，需要确保已有 TiKV 节点上的 Raft Engine 已被关闭。

    [用户文档](/tikv-configuration-file.md#raft-engine)

- **持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能**

    - 支持将更多函数下推至 MPP 引擎
        - 字符串函数：`LPAD()`、`RPAD()`、`STRCMP()`
        - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
    - 引入动态线程池，提升资源利用率（实验特性）
    - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性
    - 提升由行存到列存数据同步处理时对 raft log 的解码 (decoding) 效率，使 CPU 使用率降低达 90%
    - TiFlash 调整了文件系统中 Delta Tree 相关的默认参数，使之更适合一般的生产环境

> - **通过 session 变量实现有界限过期数据读取**

Author: 请研发 review

    - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能
    - relay log 的开启状态与 `source` 绑定，source 迁移到任意 DM-worker 均保持原有开启或关闭状态
    - relay log 的存放路径移至 DM-worker 配置文件

- **在 DM 中优化排序规则的处理方式**

    增加 `collation_compatible` 开关，支持 `strict` 和 `loose`（默认）两种模式。如果对排序规则要求不严格，允许排序规则在上下游不一致，使用默认的 `loose` 模式可使同步正常进行；如果对排序规则要求严格，排序规则在上下游不一致会导致业务报错，则可以使用 `strict` 模式。

- **在 DM 中优化 `transfer source`，支持平滑执行同步任务**

    当 DM-worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 `source` 配置迁移到其他节点。优化后的 `transfer source` 简化了用户操作步骤，不再要求先暂停所有关联 task 而是直接执行平滑迁移，DM 将在内部完成所需操作。

- **DM OpenAPI 特性 GA**

    DM 支持通过 API 的方式进行日常管理，包括增加数据源、管理任务等。本次更新 OpenAPI 从实验特性转为正式特性。

### 问题诊断效率

- **Top SQL（实验特性）**

    新推出实验性特性 Top SQL（默认关闭），帮助用户轻松找到节点中负载贡献较大的查询。

    [用户文档](/dashboard/top-sql.md)

### TiDB 数据共享订阅

- **优化 TiCDC 对集群的影响**

    大幅降低了 TiCDC 启用后，对 TiDB 集群的性能影响。在实验室环境中，TiCDC 对 TiDB 的性能影响可以降低到 5% 以下。

### 部署及运维

- **持续性能分析（实验特性）**

    - 支持更多组件：除了 TiDB、PD 和 TiKV 外，v5.4.0 版本中还支持查看 TiFlash CPU Profiling
    - 支持更方便的查看形式：支持以火焰图形式查看 CPU Profiling 和 Goroutine 结果。
    - 支持更多部署环境：支持在 TiDB Operator 部署环境下启用持续性能分析功能。

    该功能默认关闭，需进入 TiDB Dashboard 持续性能分析页面开启，开启方法见[用户文档](/dashboard/continuous-profiling.md)。

>     持续性能分析仅支持由 v1.9.0 及以上版本 TiUP 或 vx.x.x（TBD）及以上版本 TiDB Operator 升级或安装的集群。

Reviewer 13: ```suggestion
    持续性能分析仅支持由 v1.9.0 及以上版本 TiUP 或 v1.3.0 及以上版本 TiDB Operator 升级或安装的集群。
```

    - 增加了磁盘保护机制，尽量避免磁盘空间耗尽导致 panic [#10960](https://github.com/tikv/tikv/issues/10960) [#10959](https://github.com/tikv/tikv/issues/10959) [#10926](https://github.com/tikv/tikv/issues/10926) [#10925](https://github.com/tikv/tikv/issues/10925) [#10895](https://github.com/tikv/tikv/issues/10895) [#10894](https://github.com/tikv/tikv/issues/10894) [#10868](https://github.com/tikv/tikv/issues/10868) [#10840](https://github.com/tikv/tikv/issues/10840) [#10839](https://github.com/tikv/tikv/issues/10839) [#10819](https://github.com/tikv/tikv/issues/10819) [#10801](https://github.com/tikv/tikv/issues/10801)
    - 日志支持存档和轮替 [#11651](https://github.com/tikv/tikv/issues/11651)
    - 减少 Raft 客户端的系统调用并提高 CPU 效率 [#11309](https://github.com/tikv/tikv/issues/11309)
    - Coprocessor 支持下推 substring 到 TiKV [#11495](https://github.com/tikv/tikv/issues/11495)
    - 通过跳过读锁提高在 RC 隔离级别中扫描的性能 [#11485](https://github.com/tikv/tikv/issues/11485)
    - 减少备份使用的默认线程池大小，并在压力大时限制其使用 [#11000](https://github.com/tikv/tikv/issues/11000)
    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)

> 

Reviewer: remove the entry from feature section, and add this line:
```suggestion
- 优化在文件数较多且读写频繁的场景下 RocksDB 的全局锁争用问题 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)。

```

    BR 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

    该功能目前是实验特性，详细情况参考 [BR 支持 Azure Blob Storage 远端存储](/br/backup-and-restore-azblob.md)。

### 数据迁移

- **功能 9**

    <功能描述 （功能是什么 + 能给用户带来什么好处 + 需要用户注意什么）>

    <功能支持情况，告诉用户 TiDB 默认开始还是关闭此功能，如果默认关闭，如何开启>

    <如果功能限制或此功能特定的兼容性问题，需要提及>

    [用户文档](/)

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

- **在 TiDB Lightning 中添加重复数据的检测**

    在 `backend=local` 模式下，数据导入完成之前 TiDB Lightning 会输出冲突数据，然后从数据库中删除这些冲突数据。用户可以在导入完成后解析冲突数据，并根据业务规则选择适合的数据进行插入。建议根据冲突数据清洗上游数据源，避免在后续增量数据迁移阶段遇到冲突数据而造成数据不一致。

- **在 TiDB Data Migration (DM) 中 优化 relay log 的使用方式**

    - 恢复 `source` 配置中 `enable-relay` 开关
    - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能
    - relay log 的开启状态与 `source` 绑定，source 迁移到任意 DM-worker 均保持原有开启或关闭状态
    - relay log 的存放路径移至 DM-worker 配置文件

- **在 DM 中优化排序规则的处理方式**

    增加 `collation_compatible` 开关，支持 `strict` 和 `loose`（默认）两种模式。如果对排序规则要求不严格，允许排序规则不一致，使用默认的 `loose` 模式可使同步正常进行；如果对排序规则要求严格，排序规则不一致导致报错，则可以使用 `strict` 模式。

> - **在 DM 中新增 `transfer source` 支持平滑执行同步任务**

Reviewer 3: 是优化不是新增。transfer source 一直都有，就是比较难用

    当 DM-worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 `source` 配置迁移到其他节点。优化后的 `transfer source` 简化了用户操作步骤，不再要求先暂停所有关联 task 而是直接执行平滑迁移，DM 将在内部完成所需操作。

> 

Reviewer 3: - **DM OpenAPI 特性 GA**

      DM 支持通过 API 的方式进行日常管理，包括增加数据源，管理任务等。本次更新 OpenAPI 将从实验特性转为正式特性。

- **优化 TiCDC 对集群的影响**

    大幅降低了 TiCDC 启用后，对 TiDB 集群的性能影响。在实验室环境中，TiCDC 对 TiDB 的性能影响可以降低到 5% 以下。

### 问题诊断效率

> 

Reviewer 25: ```suggestion

- **Top SQL**

    新增实验性特性 Top SQL，可以轻松找到节点中负载贡献较大的查询。详见[文档](https://docs.pingcap.com/zh/tidb/stable/top-sql)。
```

Reviewer 25: https://github.com/pingcap/docs-cn/pull/8123/files

| [`tidb_enable_paging`](/system-variables.md#tidb_enable_paging-从-v540-版本开始引入)  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能会出现读请求的延迟高、TiKV 的 Unified read pool CPU 使用率高的情况。在这种情况下，由于 `Limit` 算子只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，从而降低延迟、减少资源消耗。 |
| [`tidb_read_staleness`](/system-variables.md#tidb_read_staleness-从-v540-版本开始引入) | 新增 | 用于设置当前会话允许读取的历史数据范围，默认值为 `0` |
| [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `OFF` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时。 |
| [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败 (`OFF`) 还是退回使用 pseudo 的统计信息 (`ON`)，默认值为 `ON` |
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启 [ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性，默认值为 `OFF` |
| [`tidb_store_limit`](/system-variables.md#tidb_store_limit-从-v304-和-v40-版本开始引入) | 修改 | v5.4.0 前支持实例级别及集群级别的设置，现在只支持集群级别的设置 |

> 

Reviewer 14: | [`tidb_enable_top_sql`](/system-variables.md#tidb_enable_top_sql-从-v540-版本开始引入) | 新增 | 用于控制是否开启 [Top SQL 特性](/dashboard/top-sql.md)，默认值为 `OFF` |

Reviewer 14: | `lc_time_names` | 修改 | 修改为只读变量 |

Reviewer 14: | `lc_time_names` | 修改 | 修改为只读变量 |

Reviewer 14: | `rand_seed1` | 修改 | 之前不生效，现在会生效 |
| `rand_seed2` | 修改 | 之前不生效，现在会生效 |

Author: lc_time_names 和 rand_seed1/2 这两条在变量文档 system-variables.md 中找不到，可否视为不需要用户知道呢？

Reviewer 14: | `tidb_tso_client_batch_max_wait_time` | 修改 | 现在支持设置为浮点数 |

Reviewer 14: 可以看作是个兼容性变更。
它们是继承于 MySQL 的，用户认为 TiDB 默认应该与 MySQL 一致，所以在 system-variables.md 里不提是合理的。
不过这些变量不太重要，release notes 里写不写都行。如果统一不写的话，之后还有一些也不写了。

Author: 如果是 MySQL 的兼容性变更而不是 TiDB 的，而且 TiDB 跟着 MySQL 一起变更的话，感觉也不用写在 TiDB 的 release notes 里的

Reviewer 14: 不是这个意思，MySQL 一直没变，只是 TiDB 在变。比如：
- `rand_seed1/2` 之前 TiDB 不支持，现在支持了
- `lc_time_names` 一直不支持。之前允许设置，为了防止误导用户，现在不允许设置了
如果在 system-variables.md 中本来就没有文档，我倾向于也不写 release notes 了。

Author: OK。那就不写吧

- **在 TiDB Data Migration (DM) 中 优化 relay log 的使用方式**

    - 恢复 `source` 配置中 `enable-relay` 开关
    - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能
    - relay log 的开启状态与 `source` 绑定，source 迁移到任意 DM-worker 均保持原有开启或关闭状态
    - relay log 的存放路径移至 DM-worker 配置文件

- **在 DM 中优化[排序规则](/character-set-and-collation.md)的处理方式**

    增加 collation_compatible 开关，支持 strict 和 loose（默认）两种模式：

    - 如果对排序规则要求不严格，允许上下游查询结果排序规则不一致，使用默认的 loose 模式可以避免报错。
    - 如果对排序规则要求严格，业务要求排序规则必须一致，则应当使用 strict 模式。但如果下游不支持上游缺省的 collation，同步可能会报错。

- **在 DM 中优化 `transfer source`，支持平滑执行同步任务**

    当 DM-worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 `source` 配置迁移到其他节点。优化后的 `transfer source` 简化了用户操作步骤，不再要求先暂停所有关联 task 而是直接执行平滑迁移，DM 将在内部完成所需操作。

- **DM OpenAPI 特性 GA**

    DM 支持通过 API 的方式进行日常管理，包括增加数据源、管理任务等。本次更新 OpenAPI 从实验特性转为正式特性。

### 问题诊断效率

- **Top SQL（实验特性）**

    新推出实验性特性 Top SQL（默认关闭），帮助用户轻松找到节点中消耗负载较大的查询。

    [用户文档](/dashboard/top-sql.md)

### TiDB 数据共享订阅

- **优化 TiCDC 对集群的影响**

    大幅降低了 TiCDC 启用后，对 TiDB 集群的性能影响。在实验室环境中，TiCDC 对 TiDB 的性能影响可以降低到 5% 以下。

### 部署及运维

- **持续性能分析（实验特性）**

>     - 支持更多组件：除了 TiDB、PD 和 TiKV 外，v5.4.0 版本中还支持查看 TiFlash CPU Profiling

Reviewer 13: ```suggestion
    - 支持更多组件：除了 TiDB、PD 和 TiKV 外，v5.4.0 版本中还支持查看 TiFlash CPU Profiling。
```

    - 支持更方便的查看形式：支持以火焰图形式查看 CPU Profiling 和 Goroutine 结果。
    - 支持更多部署环境：支持在 TiDB Operator 部署环境下启用持续性能分析功能。

    该功能默认关闭，需进入 TiDB Dashboard 持续性能分析页面开启，开启方法见[用户文档](/dashboard/continuous-profiling.md)。

    持续性能分析仅支持由 v1.9.0 及以上版本 TiUP 或 v1.3.0 及以上版本 TiDB Operator 升级或安装的集群。

> 

Reviewer 13: ```suggestion

    [用户文档](/dashboard/continuous-profiling.md)

```

    - 支持 `ADMIN {SESSION | INSTANCE | GLOBAL} PLAN_CACHE` 语法，用于清空缓存的查询计划 [#30370](https://github.com/pingcap/tidb/pull/30370)

+ TiKV

    - Coprocessor 支持分页 API 进行流式处理 [#11448](https://github.com/tikv/tikv/issues/11448)
    - 支持 `read-through-lock`，使读操作不需要等待清理 secondary lock [#11402](https://github.com/tikv/tikv/issues/11402)
    - 增加了磁盘保护机制，尽量避免磁盘空间耗尽导致 panic [#10537](https://github.com/tikv/tikv/issues/10537)
    - 日志支持存档和轮替 [#11651](https://github.com/tikv/tikv/issues/11651)
    - 减少 Raft 客户端的系统调用并提高 CPU 效率 [#11309](https://github.com/tikv/tikv/issues/11309)
    - Coprocessor 支持下推 substring 到 TiKV [#11495](https://github.com/tikv/tikv/issues/11495)
    - 通过跳过读锁提高在 RC 隔离级别中扫描的性能 [#11485](https://github.com/tikv/tikv/issues/11485)
    - 减少备份使用的默认线程池大小，并在压力大时限制其使用 [#11000](https://github.com/tikv/tikv/issues/11000)
    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)
    - 优化在文件数较多且读写频繁的场景下 RocksDB 的全局锁争用问题 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)

+ PD

> + TiFlash

Author: 已添加

    - 修复当有子查询时 IndexMerge 的查询结果错误的问题 [#30913](https://github.com/pingcap/tidb/issues/30913)
    - 修复当客户端设置过大的 FetchSize 后 TiDB 执行会 panic 的问题 [#30896](https://github.com/pingcap/tidb/issues/30896)
    - 修复 LEFT JOIN 有时会被错误地转换成 INNER JOIN 的问题 [#20510](https://github.com/pingcap/tidb/issues/20510)
    - 修复当 `CASE-WHEN` 表达式与 collation 同时使用时可能 panic 的问题 [#30245](https://github.com/pingcap/tidb/issues/30245)
    - 修复当 `IN` 的值中带有二进制常量时查询结果错误的问题 [#31261](https://github.com/pingcap/tidb/issues/31261)
    - 修复当 CTE 中带有子查询时查询结果错误的问题 [#31255](https://github.com/pingcap/tidb/issues/31255)
    - 修复 `INSERT ... SELECT ... ON DUPLICATE KEY UPDATE` 语句 panic 的问题 [#28078](https://github.com/pingcap/tidb/issues/28078)
    - 修复 INDEX HASH JOIN 报 `send on closed channel` 的问题 [#31129](https://github.com/pingcap/tidb/issues/31129)

+ TiKV

    - 修复 MVCC 删除记录可能不会被 GC 删除的问题 [#11217](https://github.com/tikv/tikv/issues/11217)
    - 修复悲观事务中 prewrite 请求重试在极少数情况下影响数据一致性的风险 [#11187](https://github.com/tikv/tikv/issues/11187)
    - 修复 GC 扫描导致的内存溢出 [#11410](https://github.com/tikv/tikv/issues/11410)
    - 修复当达到磁盘容量满时 RocksDB flush 或 compaction 导致的 panic [#11224](https://github.com/tikv/tikv/issues/11224)

+ PD

> + TiFlash

Author: 已添加

    [用户文档](/statistics.md#统计信息的加载)

### 稳定性

- **支持统计信息采集配置持久化**

    统计信息是优化器生成执行计划时所参考的基础信息之一，统计信息的准确性直接影响生成的执行计划是否合理。为了保证统计信息的准确性，有时候需要针对不同的表、分区、索引设置不同的采集配置项。

    TiDB 从 v5.4.0 版本开始支持 `ANALYZE` 配置持久化功能，方便后续收集统计信息时沿用已有配置项。

    `ANALYZE` 配置持久化功能默认开启（系统变量 `tidb_analyze_version` 为默认值 `2`，`tidb_persist_analyze_options` 为默认值 `ON`），用于记录手动执行 `ANALYZE` 语句时指定的持久化配置项。记录后，当 TiDB 下一次自动更新统计信息或者你手动收集统计信息但未指定配置项时，TiDB 会按照记录的配置项收集统计信息。

    [用户文档](/statistics.md#analyze-配置持久化)

## 高可用和容灾

- **优化备份对集群的影响**

    Backup & Restore (BR) 增加了备份线程自动调节功能（默认开启）。该功能通过监控集群资源的使用率自动调节备份的线程数的方式，降低备份过程对集群的影响。在某些 Case 验证中，通过增加集群用于备份的资源和开启备份线程自动调节功能，备份的影响可以降低到 10% 以下。
    详细文档请阅读 [BR 自动调节](/br/br-features.md#自动调节-从-v54-版本开始引入)。

- **支持 Azure Blob Storage 作为备份目标存储（实验特性）**

    Backup & Restore (BR) 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

    该功能目前是实验特性，详细情况参考 [BR 支持 Azure Blob Storage 远端存储](/br/backup-and-restore-azblob.md)。

### 数据迁移

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

> 

Reviewer 13: ```suggestion

    [用户文档](/tidb-lightning/tidb-lightning-configuration.md#tidb-lightning-任务配置)

```

### 配置文件参数

|  配置文件    |  配置项    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- | :----------- |
| TiDB | `stats-load-concurrency` | 新增 |               |
| TiDB | `stats-load-queue-size`   | 新增 |               |
| TiKV | `backup.enable-auto-tune` | 修改 | 在 v5.3.0 中默认值为 `false`，自 v5.4.0 起默认值改为 `true`。表示在集群资源占用率较高的情况下，是否允许 BR 自动限制备份使用的资源以求减少对集群的影响。在默认配置下，备份速度可能下降。 |
| TiKV | `log.level`、`log.format`、`log.enable-timestamp`、`log.file.filename`、`log.file.max-size`、`log.file.max-days`、`log.file.max-backups` | 新增  | 参数说明见[统一各组件的日志格式和日志归档轮转规则](#统一各组件的日志格式和日志归档轮转规则)。 |
| TiKV | `log-level`、`log-format`、`log-file`、`log-rotation-size`、`log-rotation-timespan` | 删除 | 废弃 TiKV log 参数，改为使用与 TiDB log 参数一致的命名方式，即 `log.level`、`log.format`、`log.enable-timestamp`。如果 TiKV log 参数为非默认值则保持兼容；如果同时配置 TiKV log 参数和 TiDB log 命名方式的参数，使用 TiDB log 命名方式的参数。详情参见[统一各组件的日志格式和日志归档轮转规则](#统一各组件的日志格式和日志归档轮转规则)。 |
| TiKV | `raft-engine` | 新增 | 包含 `enable`、`dir`、`batch-compression-threshold`、`bytes-per-sync`、`target-file-size`、`purge-threshold`、`recovery-mode`、`recovery-read-block-size`、`recovery-read-block-size`、`recovery-threads`，详情参见 [TiKV 配置文件：raft-engine](/tikv-configuration-file.md#raft-engine)。

> | PD | `log.level` | 修改 | 默认值由 "INFO" 改为 "info"，保证大小写不敏感 |

Reviewer 26: Add a configuration:
| TiKV | `cdc.incremental-scan-ts-filter-ratio` | 控制是否对 CDC 启用 TsFilter 优化。默认值 0.2，设置为 0 会禁用这个特性。 一般不建议调整。｜

    - 增加了磁盘保护机制，尽量避免磁盘空间耗尽导致 panic [#10537](https://github.com/tikv/tikv/issues/10537)
    - 日志支持存档和轮替 [#11651](https://github.com/tikv/tikv/issues/11651)
    - 减少 Raft 客户端的系统调用并提高 CPU 效率 [#11309](https://github.com/tikv/tikv/issues/11309)
    - Coprocessor 支持下推 substring 到 TiKV [#11495](https://github.com/tikv/tikv/issues/11495)
    - 通过跳过读锁提高在 RC 隔离级别中扫描的性能 [#11485](https://github.com/tikv/tikv/issues/11485)
    - 减少备份使用的默认线程池大小，并在压力大时限制其使用 [#11000](https://github.com/tikv/tikv/issues/11000)
    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)
    - 优化在文件数较多且读写频繁的场景下 RocksDB 的全局锁争用问题 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)

+ PD

+ TiFlash

+ Tools

    + Backup & Restore (BR)

    + TiCDC

        - 减少 "EventFeed retry rate limited" 日志的数量 [#4006](https://github.com/pingcap/tiflow/issues/4006)
        - 降低在同步大量表时的同步延时 [#3900](https://github.com/pingcap/tiflow/issues/3900)
        - 减少 TiKV 节点宕机后 kv client 恢复的时间 [#3191](https://github.com/pingcap/tiflow/issues/3191)

>     + TiDB Data Migration (DM)

Author: 已添加

>     + TiDB Data Migration (DM)

Author: 已添加

    + TiDB Lightning

    + Dumpling

>     + TiDB Binlog

Author: 已添加

>     + TiDB Binlog

Author: 已添加

| [`tidb_enable_top_sql`](/system-variables.md#tidb_enable_top_sql-从-v540-开始引入) | 新增 | 用于控制是否开启 Top SQL 特性，默认值为 OFF |
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启 [ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性，默认值为 `OFF` |
| [`tidb_read_staleness`](/system-variables.md#tidb_read_staleness-从-v540-版本开始引入) | 新增 | 用于设置当前会话允许读取的历史数据范围，默认值为 `0` |
| [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `OFF` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时 |
| [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败 (`OFF`) 还是退回使用 pseudo 的统计信息 (`ON`)，默认值为 `ON` |
| [`tidb_store_limit`](/system-variables.md#tidb_store_limit-从-v304-和-v40-版本开始引入) | 修改 | v5.4.0 前支持实例级别及集群级别的设置，现在只支持集群级别的设置 |

> 

Reviewer 14: ```
| [`tidb_regard_null_as_point `](/system-variables.md#tidb_regard_null_as_point-从-v540-版本开始引入) | 新增 | 用于控制优化器是否可以把 null 值当做点值并作为前缀条件来访问索引 |
```
Refer to https://github.com/pingcap/docs-cn/pull/8157

Author: 已添加

    - 支持 `ADMIN {SESSION | INSTANCE | GLOBAL} PLAN_CACHE` 语法，用于清空缓存的查询计划 [#30370](https://github.com/pingcap/tidb/pull/30370)

+ TiKV

    - Coprocessor 支持分页 API 进行流式处理 [#11448](https://github.com/tikv/tikv/issues/11448)
    - 支持 `read-through-lock`，使读操作不需要等待清理 secondary lock [#11402](https://github.com/tikv/tikv/issues/11402)
    - 增加了磁盘保护机制，尽量避免磁盘空间耗尽导致 panic [#10537](https://github.com/tikv/tikv/issues/10537)
    - 日志支持存档和轮替 [#11651](https://github.com/tikv/tikv/issues/11651)
    - 减少 Raft 客户端的系统调用并提高 CPU 效率 [#11309](https://github.com/tikv/tikv/issues/11309)
    - Coprocessor 支持下推 substring 到 TiKV [#11495](https://github.com/tikv/tikv/issues/11495)
    - 通过跳过读锁提高在 RC 隔离级别中扫描的性能 [#11485](https://github.com/tikv/tikv/issues/11485)
    - 减少备份使用的默认线程池大小，并在压力大时限制其使用 [#11000](https://github.com/tikv/tikv/issues/11000)
    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)
    - 优化在文件数较多且读写频繁的场景下 RocksDB 的全局锁争用问题 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)

+ PD

> 

Reviewer 27: - 默认开启历史热点记录功能[#4521](https://github.com/tikv/pd/pull/4521)
- 新增 HTTP Component 的签名，用于标识请求来源[#4491](https://github.com/tikv/pd/pull/4491)
- TiDB Dashboard 更新至 v2021.12.31, 新增 Top SQL, 改善 Profiling 页面 [#4502](https://github.com/tikv/pd/pull/4502)

Author: 已添加

+ TiFlash

    - 优化本地算子通讯
    - 调高 grpc 非临时线程数，避免频繁创建/销毁线程

+ Tools

>     + Backup & Restore (BR)

Reviewer 28: - 增加 BR 加密备份时，对密钥的合法性检查[#29991](https://github.com/pingcap/tidb/pull/29991)

Author: 已添加

    + TiCDC

        - 减少 "EventFeed retry rate limited" 日志的数量 [#4006](https://github.com/pingcap/tiflow/issues/4006)
        - 降低在同步大量表时的同步延时 [#3900](https://github.com/pingcap/tiflow/issues/3900)
        - 减少 TiKV 节点宕机后 kv client 恢复的时间 [#3191](https://github.com/pingcap/tiflow/issues/3191)

    + TiDB Data Migration (DM)

>     + TiDB Lightning

Reviewer 28: - 在 tidb backend 中，默认改用乐观事务进行写入，来提升性能。

Author: 已添加

    - 修复当有子查询时 IndexMerge 的查询结果错误的问题 [#30913](https://github.com/pingcap/tidb/issues/30913)
    - 修复当客户端设置过大的 FetchSize 后 TiDB 执行会 panic 的问题 [#30896](https://github.com/pingcap/tidb/issues/30896)
    - 修复 LEFT JOIN 有时会被错误地转换成 INNER JOIN 的问题 [#20510](https://github.com/pingcap/tidb/issues/20510)
    - 修复当 `CASE-WHEN` 表达式与 collation 同时使用时可能 panic 的问题 [#30245](https://github.com/pingcap/tidb/issues/30245)
    - 修复当 `IN` 的值中带有二进制常量时查询结果错误的问题 [#31261](https://github.com/pingcap/tidb/issues/31261)
    - 修复当 CTE 中带有子查询时查询结果错误的问题 [#31255](https://github.com/pingcap/tidb/issues/31255)
    - 修复 `INSERT ... SELECT ... ON DUPLICATE KEY UPDATE` 语句 panic 的问题 [#28078](https://github.com/pingcap/tidb/issues/28078)
    - 修复 INDEX HASH JOIN 报 `send on closed channel` 的问题 [#31129](https://github.com/pingcap/tidb/issues/31129)

+ TiKV

    - 修复 MVCC 删除记录可能不会被 GC 删除的问题 [#11217](https://github.com/tikv/tikv/issues/11217)
    - 修复悲观事务中 prewrite 请求重试在极少数情况下影响数据一致性的风险 [#11187](https://github.com/tikv/tikv/issues/11187)
    - 修复 GC 扫描导致的内存溢出 [#11410](https://github.com/tikv/tikv/issues/11410)
    - 修复当达到磁盘容量满时 RocksDB flush 或 compaction 导致的 panic [#11224](https://github.com/tikv/tikv/issues/11224)

+ PD

> 

Reviewer 27: - 修复 Region 统计不受 `flow-round-by-digit` 影响的问题[#4304](https://github.com/tikv/pd/pull/4304) 
- 修复调度 Operator 因为目标 Store 处于 Down 的状态而无法快速失败的问题 [#4223](https://github.com/tikv/pd/pull/4223)
- 修复不能 Merge 在 Offline Store 上面的 Region 的问题[#4357](https://github.com/tikv/pd/pull/4357)
- 修复热点统计中无法剔除冷热点数据的问题[#4396](https://github.com/tikv/pd/pull/4396)

Author: 已添加

+ TiFlash

    - 修复查询报错 `Tree struct based executor must have executor id`
    - 修复查询报错 `Illegal column type UInt32 for the second argument of function tidbRoundWithFrac`
    - 修复查询报错 `Unexpected type of column: Nullable(Nothing)`
    - 修复查询报错 `source region at right may be considered at left when merging`
    - 修复 `collation` 对 `nullable` 类型失效的问题
    - 修复 `coalesce` 错误地从结果列中删除 `nullable` 标志的问题
    - 修复当 MPP 查询被终止时，TiFlash 偶发的崩溃问题
    - 修复 `where <string>` 查询结果出错的问题
    - 修复整数类型主键的列类型设置为较大范围后数据可能不一致的问题
    - 修复输入早于 1970-01-01 00:00:01 UTC 时，`unix_timestamp` 行为与 TiDB/MySQL 不一致的问题
    - 修复 TiFlash 重启时偶发的 `EstablishMPPConnection` 错误
    - 修复在 TiFlash 与 TiDB/TiKV 之间 `CastStringAsDecimal` 行为不一致的问题
    - 修复查询报错 `DB::Exception: Encode type of coprocessor response is not CHBlock`
    - 修复在 TiFlash 与 TiDB/TiKV 之间 `castStringAsReal ` 行为不一致的问题。
    - 修复 TiFlash 的 `date_add_string_xxx` 函数返回值与 mysql 不一致的问题

+ Tools

>     + Backup & Restore (BR)

Author: 已添加

        - 修复 `mq sink write row` 没有监控数据的问题 [#3431](https://github.com/pingcap/tiflow/issues/3431)
        - 修复 `sql mode` 兼容性问题 [#3810](https://github.com/pingcap/tiflow/issues/3810)
        - 修复在移除同步任务后潜在的 panic 问题 [#3128](https://github.com/pingcap/tiflow/issues/3128)
        - 修复输出默认列值时的 panic 问题和数据不一致的问题 [#3929](https://github.com/pingcap/tiflow/issues/3929)
        - 修复不支持同步默认值的问题 [#3793](https://github.com/pingcap/tiflow/issues/3793)
        - 修复潜在的同步流控死锁问题 [#4055](https://github.com/pingcap/tiflow/issues/4055)
        - 修复在磁盘写满时无日志输出的问题 [#3362](https://github.com/pingcap/tiflow/issues/3362)
        - 修复 DDL 特殊注释导致的同步停止的问题 [#3755](https://github.com/pingcap/tiflow/issues/3755)
        - 修复在某些 RHEL 发行版上因时区问题导致服务无法启动的问题 [#3584](https://github.com/pingcap/tiflow/issues/3584)
        - 修复因 checkpoint 不准确导致的潜在的数据丢失问题 [#3545](https://github.com/pingcap/tiflow/issues/3545)
        - 修复在容器环境中 OOM 的问题 [#1798](https://github.com/pingcap/tiflow/issues/1798)
        - 修复 `config.Metadata.Timeout` 没有正确配置而导致的同步停止问题 [#3352](https://github.com/pingcap/tiflow/issues/3352)

    + TiDB Data Migration (DM)

>     + TiDB Lightning

Reviewer 28: - 修复当 lightning 没有权限访问系统表 `mysql.tidb` 时，可能生成错误 kv 格式的问题。[#31089](https://github.com/pingcap/tidb/pull/31089)
- 修复 lightning 重启时，跳过某些检查的问题 [#31033](https://github.com/pingcap/tidb/pull/31033)

Author: 已添加

Reviewer 28: 又加了两个。。。
- 修复 lightning local backend 导入后，在某些数据集上会生成大量空 region 的问题#30428
- 修复 lightning 不识别 gs 前缀的存储地址的问题#30254

Reviewer 24: LGTM

    [用户文档](/br/br-auto-tune.md)

- **支持 Azure Blob Storage 作为备份目标存储**

    Backup & Restore (BR) 支持 Azure Blob Storage 作为备份的远端目标存储。在 Azure Cloud 环境部署 TiDB 的用户，可以支持使用该功能将集群数据备份到 Azure Blob Storage 服务中。

    [用户文档](/br/backup-and-restore-azblob.md)

### 数据迁移

- **为 TiDB Lightning 增加已存在数据表是否允许导入的开关**

    为 TiDB Lightning 增加 `incremental-import` 开关。默认值为 `false`，表明目标表已存在数据时将不会执行导入。将默认值改为 `true` 则继续导入。注意，当使用并行导入特性时，需要将该配置项设为 `true`。

    [用户文档](/tidb-lightning/tidb-lightning-configuration.md#tidb-lightning-任务配置)

- **增加新配置允许自定义用于保存 TiDB Lightning 并行导入特性的元信息的 schema 名称**

     为 TiDB Lightning 增加 `meta-schema-name` 配置。在并行导入模式下，该参数用于在目标集群保存各个 TiDB Lightning 实例元信息的 schema 名称，默认值为 "lightning_metadata"。对于参与同一批并行导入的每个 TiDB Lightning 实例，必须将此配置项设置为相同的值，否则将无法确保导入数据的正确性。

    [用户文档](/tidb-lightning/tidb-lightning-configuration.md#tidb-lightning-任务配置)

- **在 TiDB Lightning 中添加重复数据的检测**

    在 `backend=local` 模式下，数据导入完成之前 TiDB Lightning 会输出冲突数据，然后从数据库中删除这些冲突数据。用户可以在导入完成后解析冲突数据，并根据业务规则选择适合的数据进行插入。建议根据冲突数据清洗上游数据源，避免在后续增量数据迁移阶段遇到冲突数据而造成数据不一致。

    [用户文档](/tidb-lightning/tidb-lightning-error-resolution.md)

- **在 TiDB Data Migration (DM) 中 优化 relay log 的使用方式**

    - 恢复 `source` 配置中 `enable-relay` 开关。

>     - 增加 `start-relay` 或 `stop-relay` 命令中动态开启或关闭 relay log 的功能。

Reviewer 3: ```suggestion
    - 增加 `start-relay` 和 `stop-relay` 命令动态开启和关闭 relay log 的功能。
```

+ PD

    - 默认开启历史热点记录功能 [#25281](https://github.com/pingcap/tidb/issues/25281)
    - 新增 HTTP Component 的签名，用于标识请求来源 [#4490](https://github.com/tikv/pd/issues/4490)
    - TiDB Dashboard 更新至 v2021.12.31 [#4257](https://github.com/tikv/pd/issues/4257)

+ TiFlash

    - 优化本地算子通讯
    - 调高 gRPC 非临时线程数，避免频繁创建/销毁线程

+ Tools

    + Backup & Restore (BR)

        - 增加 BR 加密备份时，对密钥的合法性检查 [#29794](https://github.com/pingcap/tidb/issues/29794)

    + TiCDC

        - 减少 "EventFeed retry rate limited" 日志的数量 [#4006](https://github.com/pingcap/tiflow/issues/4006)
        - 降低在同步大量表时的同步延时 [#3900](https://github.com/pingcap/tiflow/issues/3900)
        - 减少 TiKV 节点宕机后 KV client 恢复的时间 [#3191](https://github.com/pingcap/tiflow/issues/3191)

    + TiDB Data Migration (DM)

        - 降低开启 relay 时的 CPU 使用率 [#2214](https://github.com/pingcap/dm/issues/2214)

    + TiDB Lightning

        - 在 TiDB-backend 模式下，默认改用乐观事务进行写入来提升性能 [#30953](https://github.com/pingcap/tidb/pull/30953)

    + Dumpling

        - 提升 Dumpling 检查数据库版本时的兼容性 [#29500](https://github.com/pingcap/tidb/pull/29500)
        - 在导出 `CREATE DATABASE` 和 `CREATE TABLE` 时添加默认的 collation [#3420](https://github.com/pingcap/tiflow/issues/3420)

> 

Author: ```suggestion
```

| [`tidb_regard_null_as_point `](/system-variables.md#tidb_regard_null_as_point-从-v540-版本开始引入) | 新增 | 用于控制优化器是否可以把 null 值当做点值并作为前缀条件来访问索引 |
| [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `OFF` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时 |
| [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败 (`OFF`) 还是退回使用 pseudo 的统计信息 (`ON`)，默认值为 `ON` |
| [`tidb_store_limit`](/system-variables.md#tidb_store_limit-从-v304-和-v40-版本开始引入) | 修改 | v5.4.0 前支持实例级别及集群级别的设置，现在只支持集群级别的设置 |

### 配置文件参数

|  配置文件    |  配置项    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- | :----------- |
| TiDB | [`stats-load-concurrency`](/tidb-configuration-file.md#stats-load-concurrency-从-v540-版本开始引入) | 新增 |  用于设置 TiDB 统计信息同步加载功能可以并发处理的最大列数，默认值为 `5`             |
| TiDB | [`stats-load-queue-size`](/tidb-configuration-file.md#stats-load-queue-size-从-v540-版本开始引入)   | 新增 |  用于设置 TiDB 统计信息同步加载功能最多可以缓存多少列的请求，默认值为 `1000`             |

> | TiDB | [`txn-total-size-limit`](/tidb-configuration-file.md#txn-total-size-limit) | 修改 | 配置 TiDB 单个事务大小的限制。该配置项的最大值由 `10737418240`（表示 10GB）调整至 `1099511627776`（表示 1TB）。|

Author: ```suggestion
```

    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)
    - 优化在文件数较多且读写频繁的场景下 RocksDB 的全局锁争用问题 [rocksdb#250](https://github.com/tikv/rocksdb/pull/250)

+ PD

    - 默认开启历史热点记录功能 [#25281](https://github.com/pingcap/tidb/issues/25281)
    - 新增 HTTP Component 的签名，用于标识请求来源 [#4490 ](https://github.com/tikv/pd/issues/4490)
    - TiDB Dashboard 更新至 v2021.12.31，新增 Top SQL，改善 Profiling 页面 [#4257](https://github.com/tikv/pd/issues/4257)

+ TiFlash

    - 优化本地算子通讯
    - 调高 gRPC 非临时线程数，避免频繁创建/销毁线程

+ Tools

    + Backup & Restore (BR)

        - 增加 BR 加密备份时，对密钥的合法性检查 [#29794](https://github.com/pingcap/tidb/issues/29794)

    + TiCDC

        - 减少 "EventFeed retry rate limited" 日志的数量 [#4006](https://github.com/pingcap/tiflow/issues/4006)
        - 降低在同步大量表时的同步延时 [#3900](https://github.com/pingcap/tiflow/issues/3900)
        - 减少 TiKV 节点宕机后 kv client 恢复的时间 [#3191](https://github.com/pingcap/tiflow/issues/3191)

    + TiDB Data Migration (DM)

    + TiDB Lightning

        - 在 TiDB-backend 模式下，默认改用乐观事务进行写入来提升性能 [#30953](https://github.com/pingcap/tidb/pull/30953)

>     + Dumpling
> 

Reviewer 29: ```suggestion
    + Dumpling

        - 提升 dumpling 检查数据库版本时的兼容性 [#29500](https://github.com/pingcap/tidb/pull/29500)
        - 支持 collation/charset 严格模式导出 [#30292](https://github.com/pingcap/tidb/pull/30292)
```

Author: 已添加

---
title: TiDB 5.4 Release Notes
---

# TiDB 5.4 Release Notes

发版日期：2022 年 x 月 x 日

TiDB 版本：5.4.0

在 v5.4.0 版本中，你可以获得以下关键特性：

+ 支持 GBK 字符集
+ 支持索引合并 (Index Merge) 数据访问方法，能够合并多个列上索引的条件过滤结果
+ 支持通过 session 变量实现有界限过期数据读取

> + 支持统计信息采集配置持久化

Author: ```suggestion
+ 支持统计信息采集配置持久化
+ 支持使用 Raft Engine 作为 TiKV 的日志存储引擎
```

| [`tidb_enable_index_merge`](/system-variables.md#tidb_enable_index_merge-从-v40-版本开始引入) | 修改 | 默认值由 `OFF` 改为 `ON`。如果从低于 v4.0.0 版本升级到 v5.4.0 及以上版本的集群，该变量值默认保持 `OFF`。如果从 v4.0.0 及以上版本升级到 v5.4.0 及以上版本的集群，该变量开关保持升级前的状态。对于 v5.4.0 及以上版本的新建集群，该变量开关默认保持 `ON`。 |
| [`tidb_enable_paging`](/system-variables.md#tidb_enable_paging-从-v540-版本开始引入)  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能会出现读请求的延迟高、TiKV 的 Unified read pool CPU 使用率高的情况。在这种情况下，由于 `Limit` 算子只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，从而降低延迟、减少资源消耗。 |
| [`tidb_enable_top_sql`](/system-variables.md#tidb_enable_top_sql-从-v540-开始引入) | 新增 | 用于控制是否开启 Top SQL 特性，默认值为 OFF |
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启 [ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性，默认值为 `OFF` |
| [`tidb_read_staleness`](/system-variables.md#tidb_read_staleness-从-v540-版本开始引入) | 新增 | 用于设置当前会话允许读取的历史数据范围，默认值为 `0` |
| [`tidb_regard_null_as_point `](/system-variables.md#tidb_regard_null_as_point-从-v540-版本开始引入) | 新增 | 用于控制优化器是否可以把 null 值当做点值并作为前缀条件来访问索引 |

> | [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `OFF` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时 |

Reviewer 30: 这是个数值类型，默认为 0

Author: ```suggestion
| [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `0` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时 |
```

| TiKV | [`snap-generator-pool-size`](/tidb-configuration-file.md#snap-generator-pool-size) | 新增 | `snap-generator` 线程池大小，默认值为 `2` |
| TiKV | `log.file.max-size`、`log.file.max-days`、`log.file.max-backups` | 新增  | 参数说明见 [TiKV 配置文件 - log.file](/tikv-configuration-file.md#logfile-从-v540-版本开始引入)。 |
| TiKV | `raft-engine` | 新增 | 包含 `enable`、`dir`、`batch-compression-threshold`、`bytes-per-sync`、`target-file-size`、`purge-threshold`、`recovery-mode`、`recovery-read-block-size`、`recovery-read-block-size`、`recovery-threads`，详情参见 [TiKV 配置文件：raft-engine](/tikv-configuration-file.md#raft-engine)。|
| TiKV | [`backup.enable-auto-tune`](/tikv-configuration-file.md#enable-auto-tune-从-v54-版本开始引入) | 修改 | 在 v5.3.0 中默认值为 `false`，自 v5.4.0 起默认值改为 `true`。表示在集群资源占用率较高的情况下，是否允许 BR 自动限制备份使用的资源，减少对集群的影响。在默认配置下，备份速度可能下降。 |
| TiKV | `log-level`、`log-format`、`log-file`、`log-rotation-size` | 修改 | 将 TiKV log 参数名替换为与 TiDB log 参数一致的参数名，即 `log.level`、`log.format`、`log.file.filename`、`log.enable-timestamp`。如果只设置了原参数、且把其值设为非默认值，原参数与新参数会保持兼容；如果同时设置了原参数和新参数，则会使用新参数。详情参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。 |

> | TiKV | [`raftstore.raft-log-gc-tick-interval`](/tikv-configuration-file.md#raft-log-gc-tick-interval) | 修改 | 默认值修改为 `3s` |

Author: ```suggestion
```

| TiDB Data Migration (DM)  | [`collation_compatible`](/dm/task-configuration-file-full.md#完整配置文件示例) | 新增 | 同步 CREATE 语句中缺省 Collation 的方式，可选 "loose" 和 "strict"，默认为 "loose"。 |
| TiCDC | `max-message-bytes` | 修改 | 将 Kafka sink 模块的 `max-message-bytes` 默认值设置为 `104857601`（10MB）  |
| TiCDC | `partition-num`      | 修改 | 将 Kafka Sink `partition-num` 的默认值改由 `4` 为 `3`，使 TiCDC 更加平均地分发消息到各个 Kafka partition |
| TiDB Lightning | `meta-schema-name` | 修改 | 此配置项控制 TiDB Lightning 在目标 TiDB 中保存 metadata 对应的 schema name。从 v5.4.0 开始，只在开启了并行导入功能时（对应配置为 `tikv-importer.incremental-import = true` ），才会在目标 TiDB 中创建此库。 |
| TiDB Lightning | `task-info-schema-name` |  新增  | 用于配置当 Lightning 检测到冲突数据时，对应冲突数据存储的库名，默认值为 "lightning_task_info"。如果没有开启冲突检测功能，则无需配置此参数。  |
| TiDB Lightning | `incremental-import` | 新增 | 是否允许向已存在数据的表导入数据。默认值为 false |

### 其他

- TiDB Dashboard 默认不再使用 `root` + 空密码登录。

    从 v5.4.0开始，使用 TiUP 启动集群时推荐使用 `start --initial`。执行该操作启动集群后，会为 `root` 账号自动生成一个随机密码，`root` 账号登录 Dashboard 需要使用这个密码。

- 为 TiDB 和 PD 之间新增接口。使用 `information_schema.TIDB_HOT_REGIONS_HISTORY` 系统表时，TiDB 需要使用匹配的 PD 版本。

> - TiDB Server、PD Server 和 TiKV Server 将采用统一的参数命名方式来管理日志命名、输出格式、轮转和过期的规则。参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。

Reviewer 19: ```suggestion
- 对 log 相关参数，TiDB Server、PD Server 和 TiKV Server 将采用统一的参数命名方式来管理日志命名、输出格式、轮转和过期的规则。参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。
```

- 自 v5.4.0 起，对于通过 Plan Cache 已经缓存的执行计划，如果为其创建绑定 (Binding)，会使得对应查询已经缓存的计划失效。v5.4.0 前已经缓存的计划不受新 Binding 的影响。

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

    在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。

    为了更好的支持中文用户，TiDB 从 v5.4.0 起支持 GBK 字符集。在初次初始化 TiDB 集群时开启 TiDB 配置项 [`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) 后，TiDB GBK 字符集同时支持 `gbk_bin` 和 `gbk_chinese_ci` 这两种排序规则。

在使用 GBK 字符集时，需要注意兼容性限制，详情参考[字符集和排序 - GBK](/character-set-gbk.md)。

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

    [用户文档](/tispark-overview.md#安全)

### 性能

- **持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能**

    - 支持将更多函数下推至 MPP 引擎
        - 字符串函数：`LPAD()`、`RPAD()`、`STRCMP()`
        - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
    - 引入动态线程池，提升资源利用率（实验特性）

>     - 新增或修改一些 TiFlash 已有配置的默认值，提升 TiFlash 的性能和稳定性
>     - 提升由行存到列存数据同步处理时对 raft log 的解码 (decoding) 效率，数据同步性能提升 50%

Reviewer 20: ```suggestion
    - 提升从 TiKV 同步数据时，由行存格式到列存格式的数据转换效率，整体的数据同步性能提升 50%
    - 调整一些配置项的默认值，提升 TiFlash 的性能和稳定性。HTAP 混合负载下，单表简单查询的性能最高提升 20%
```

    用户文档：[TiFlash 支持的计算下推](/tiflash/use-tiflash.md#tiflash-支持的计算下推)，[TiFlash 配置文件](/tiflash/tiflash-configuration.md#配置文件-tiflashtoml)

- **通过 session 变量实现有界限过期数据读取**

>     TiDB 是基于 Raft 协议的多副本分布式数据库。面对高并发，高吞吐业务场景，可以通过follower 节点实现读性能扩展，构建读写分离架构。
> 
>     针对不同的业务场景，follower 提供强一致读和弱一致过期读两种读模式。强一致读满足数据实时性要求严格的业务场景，但是因为 leader 和 follower 的数据同步延迟、吞吐较低、延迟较高，特别是在跨机房架构下延迟问题被进一步放大。
> 
>     在对数据实时性要求不高的业务场景下，可以选择过期读模式。使用该模式可以降低延迟和提升吞吐。TiDB 目前支持通过显示只读事务或 SQL 语句的方式实现过期读。两种方式均支持指定时间的精确过期读和指定时间边界的过期读两种模式，详细用法请参考[过期读文档](/read-historical-data.md)。
> 
>     从 v5.4.0 版本开始 TiDB 支持通过 session 变量设置有界限过期读，进一步提升易用性，具体设置示例如下：
> 
>     ```sql
>     set @@tidb_replica_read=leader_and_follower
>     set @@tidb_read_staleness="-5"
>     ```
> 
>     通过该设置，TiDB 可以实现就近选取 leader 或 follower 节点，并读取 5 秒钟内的最新过期数据，满足准实时场景下低延迟高吞吐数据访问的业务诉求，降低研发门槛，提升易用性。
> 
>     [用户文档](/tidb-read-staleness.md)

Reviewer 19: ```suggestion
    TiDB 是基于 Raft 协议的多副本分布式数据库。面对高并发，高吞吐业务场景，可以通过 follower 节点实现读性能扩展，构建读写分离架构。

    针对不同的业务场景，follower 提供强一致读和弱一致过期读两种读模式。强一致读模式可以满足数据实时性要求严格的业务场景。然而，当采用该模式时，特别是在跨机房架构下，由于会出现 leader 和 follower 的数据同步延迟、吞吐量降低等情况，会出现延迟问题。

    在对数据实时性要求不高的业务场景下，可以选择过期读模式。使用该模式可以降低延迟和提升吞吐。TiDB 目前支持通过显示只读事务或 SQL 语句的方式实现过期读。通过这两种方式，你均可以从指定时间点或时间范围内读取对应的历史数据。具体用法，请参考[通过系统变量 `tidb_snapshot` 读取历史数据](/read-historical-data.md)。

    从 v5.4.0 版本开始 TiDB 支持通过 session 变量设置有界限过期读，进一步提升易用性，具体设置示例如下：

    ```sql
    set @@tidb_replica_read=leader_and_follower
    set @@tidb_read_staleness="-5"
    ```

    通过该设置，TiDB 可以实现就近选取 leader 或 follower 节点，并读取 5 秒钟内的最新过期数据，满足准实时场景下低延迟高吞吐数据访问的业务诉求，降低研发门槛，提升易用性。

    [用户文档](/read-historical-data.md)
```

- **TiDB 正式发布索引合并 (Index Merge) 功能**

    _索引合并_ 是在 TiDB v4.0 版本中作为实验特性引入的一种查询执行方式的优化，可以大幅提高查询在扫描多列数据时条件过滤的效率。例如对以下的查询，若 `WHERE` 子句中两个 `OR` 连接的过滤条件在各自包含的 _key1_ 与 _key2_ 两个列上都存在索引，则 _索引合并_ 可以同时利用  _key1_ 与 _key2_ 上的索引分别进行过滤，然后合并出最终的结果。

    ```sql
    SELECT * FROM table WHERE key1 <= 100 OR key2 = 200;
    ```

    以往 TiDB 在一个表上的查询只能使用一个索引，无法同时使用多个索引进行条件过滤。相较以往，_索引合并_ 避免了此情况下可能不必要的大量数据扫描，也可以使得需要灵活查询不特定多列数据组合的用户利用单列上的索引达到高效稳定的查询，无需大量构建多列复合索引。

    本版本正式发布了 _索引合并 (Index Merge)_ 特性，但仍存在以下的使用条件和限制：

    - 目前 TiDB 的 _索引合并_ 优化只限于 _析取范式_ (X<sub>1</sub> ⋁ X<sub>2</sub> ⋁ …X<sub>n</sub>)，即 `WHERE` 子句中过滤条件连接词为 `OR`。

    - 如果全新部署的集群版本为 v5.4.0 或以上，此特性默认开启。如果从 v5.4.0 以前的版本升级到 v5.4.0 或以上，默认保持升级前此特性的开关状态（v4.0.0 之前无此项特性的版本默认关闭），由用户决定是否开启。

    [用户文档](/explain-index-merge.md)

> 

Author: ```suggestion

- **新增 Raft Engine（实验特性）**

    支持使用 [Raft Engine](https://github.com/tikv/raft-engine) 作为 TiKV 的日志存储引擎。与使用 RocksDB 相比，Raft Engine 可以减少至多 40% 的 TiKV I/O 写流量和 10% 的 CPU 使用，同时在特定负载下提升 5% 左右前台吞吐，减少 20% 尾延迟。此外，Raft Engine 提升了日志回收效率，修复了极端条件下日志堆积的问题。

    Raft Engine 目前仍属于实验特性，并默认关闭。另外请注意 v5.4.0 版本的 Raft Engine 数据格式与之前版本不兼容，对集群做升级或者降级操作之前，需要确保所有 TiKV 节点上的 Raft Engine 已被关闭。只建议在 v5.4.0 及以后的版本使用 Raft Engine。

    [用户文档](/tikv-configuration-file.md#raft-engine)

```

+ 优化持续性能分析（实验特性）
+ TiSpark 支持用户认证与鉴权

## 兼容性变化

> **注意：**
>
> 当从一个早期的 TiDB 版本升级到 TiDB v5.4.0 时，如需了解所有中间版本对应的兼容性更改说明，请查看对应版本的 [Release Notes](/releases/release-notes.md)。

### 系统变量

|  变量名    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- |
|  [`tidb_enable_column_tracking`](/system-variables.md#tidb_enable_column_tracking-从-v540-版本开始引入) | 新增 | 用于控制是否开启 TiDB 对 `PREDICATE COLUMNS` 的收集，默认值为 `OFF` |
| [`tidb_enable_paging`](/system-variables.md#tidb_enable_paging-从-v540-版本开始引入)  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能会出现读请求的延迟高、TiKV 的 Unified read pool CPU 使用率高的情况。在这种情况下，由于 `Limit` 算子只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，从而降低延迟、减少资源消耗。 |
| [`tidb_enable_top_sql`](/system-variables.md#tidb_enable_top_sql-从-v540-开始引入) | 新增 | 用于控制是否开启 Top SQL 特性，默认值为 OFF |
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启 [ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性，默认值为 `OFF` |
| [`tidb_read_staleness`](/system-variables.md#tidb_read_staleness-从-v540-版本开始引入) | 新增 | 用于设置当前会话允许读取的历史数据范围，默认值为 `0` |

> | [`tidb_regard_null_as_point`](/system-variables.md#tidb_regard_null_as_point-从-v540-版本开始引入) | 新增 | 用于控制优化器是否可以把 null 值当做点值并作为前缀条件来访问索引 |

Reviewer 11: ```suggestion
| [`tidb_regard_null_as_point`](/system-variables.md#tidb_regard_null_as_point-从-v540-版本开始引入) | 新增 | 用于控制优化器是否可以将包含 null 的等值条件作为前缀条件来访问索引 |
```

| TiFlash | [`raftstore.apply-pool-size`](/tiflash/tiflash-configuration.md#配置文件-tiflash-learnertoml) | 新增 | 处理 Raft 数据落盘的线程池中线程的数量，默认值为 `4`。 |
| TiFlash | [`raftstore.store-pool-size`](/tiflash/tiflash-configuration.md#配置文件-tiflash-learnertoml) | 新增 | 处理 Raft 的线程池中线程的数量，即 Raftstore 线程池的大小，默认值为 `4`。 |
| TiDB Data Migration (DM)  | [`collation_compatible`](/dm/task-configuration-file-full.md#完整配置文件示例) | 新增 | 同步 CREATE 语句中缺省 Collation 的方式，可选 "loose" 和 "strict"，默认为 "loose"。 |
| TiCDC | `max-message-bytes` | 修改 | 将 Kafka sink 模块的 `max-message-bytes` 默认值设置为 `104857601`（10MB）  |
| TiCDC | `partition-num`      | 修改 | 将 Kafka Sink `partition-num` 的默认值改由 `4` 为 `3`，使 TiCDC 更加平均地分发消息到各个 Kafka partition |
| TiDB Lightning | `meta-schema-name` | 修改 | 此配置项控制 TiDB Lightning 在目标 TiDB 中保存 metadata 对应的 schema name。从 v5.4.0 开始，只在开启了并行导入功能时（对应配置为 `tikv-importer.incremental-import = true` ），才会在目标 TiDB 中创建此库。 |
| TiDB Lightning | `task-info-schema-name` |  新增  | 用于配置当 Lightning 检测到冲突数据时，对应冲突数据存储的库名，默认值为 "lightning_task_info"。如果没有开启冲突检测功能，则无需配置此参数。  |
| TiDB Lightning | `incremental-import` | 新增 | 是否允许向已存在数据的表导入数据。默认值为 false |

### 其他

> - TiDB Dashboard 默认不再使用 `root` + 空密码登录。

Reviewer 11: 如果使用 TiUP 启动集群时没有使用 `start --initial`，而是普通方式，TiDB Dashboard 依然默认使用 `root` + 空密码登录是吗？ @qqqdan

- 对 log 相关参数，TiDB Server、PD Server 和 TiKV Server 将采用统一的参数命名方式来管理日志命名、输出格式、轮转和过期的规则。参见 [TiKV 配置文件 - log](/tikv-configuration-file.md#log-从-v540-版本开始引入)。
- 自 v5.4.0 起，对于通过 Plan Cache 已经缓存的执行计划，如果为其创建绑定 (Binding)，会使得对应查询已经缓存的计划失效。v5.4.0 前已经缓存的计划不受新 Binding 的影响。
- 在 v5.3 及更早版本中，[TiDB Data Migration (DM)](https://docs.pingcap.com/zh/tidb-data-migration/v5.3/) 文档独立于 TiDB 文档。自 v5.4 起，TiDB Data Migration 的文档已合并入相同版本的 TiDB 文档，无需跳转到 DM 文档站，你可以直接在 TiDB 文档站阅读 [DM 文档](/dm/dm-overview.md)。

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

    在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。

    为了更好的支持中文用户，TiDB 从 v5.4.0 起支持 GBK 字符集。在初次初始化 TiDB 集群时开启 TiDB 配置项 [`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) 后，TiDB GBK 字符集同时支持 `gbk_bin` 和 `gbk_chinese_ci` 这两种排序规则。

在使用 GBK 字符集时，需要注意兼容性限制，详情参考[字符集和排序 - GBK](/character-set-gbk.md)。

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

    [用户文档](/tispark-overview.md#安全)

> 

Reviewer 31: - ** TiUP 部署方式支持为 root 用户生成初始密码**
         对于使用 TiUP 部署集群的场景，对启动集群的命令提供 --init 参数，支持为数据库 root 用户生成初始密码，避免使用密码为空的 root 用户所带来的安全风险，保证数据库的安全性。

> 

Reviewer 13: ```suggestion

- **TiUP 部署方式支持为 root 用户生成初始密码**

    集群启动命令增加了 `--init` 参数，有了该参数，在 TiUP 部署场景，TiUP 会为数据库 root 用户生成一个原始密码，避免使用密码为空的 root 用户所带来的安全风险，保证数据库的安全性。

    [User document](/production-deployment-using-tiup.md##第-7-步启动集群)

```

Reviewer 31: @shichun-0415 内容再改一下。。

    集群启动命令增加了 `--init` 参数，有了该参数，在 TiUP 部署场景，TiUP 会为数据库 root 用户生成一个初始的强密码，避免 root 用户使用空密码所带来的安全风险，增强数据库的安全性。

---
title: TiDB 5.4 Release Notes
---

# TiDB 5.4 Release Notes

> 发版日期：2022 年 x 月 x 日

Author: ```suggestion
发版日期：2022 年 2 月 15 日
```

| TiFlash | [`logger.count`](/tiflash/tiflash-configuration.md#配置文件-tiflashtoml) | 修改 | 默认值修改为 `10` |
| TiFlash | [`status.metrics_port`](/tiflash/tiflash-configuration.md#配置文件-tiflashtoml) | 修改 | 默认值修改为 `8234` |
| TiFlash | [`raftstore.apply-pool-size`](/tiflash/tiflash-configuration.md#配置文件-tiflash-learnertoml) | 新增 | 处理 Raft 数据落盘的线程池中线程的数量，默认值为 `4`。 |
| TiFlash | [`raftstore.store-pool-size`](/tiflash/tiflash-configuration.md#配置文件-tiflash-learnertoml) | 新增 | 处理 Raft 的线程池中线程的数量，即 Raftstore 线程池的大小，默认值为 `4`。 |
| TiDB Data Migration (DM)  | [`collation_compatible`](/dm/task-configuration-file-full.md#完整配置文件示例) | 新增 | 同步 CREATE 语句中缺省 Collation 的方式，可选 "loose" 和 "strict"，默认为 "loose"。 |
| TiCDC | `max-message-bytes` | 修改 | 将 Kafka sink 模块的 `max-message-bytes` 默认值设置为 `104857601`（10MB）  |
| TiCDC | `partition-num`      | 修改 | 将 Kafka Sink `partition-num` 的默认值改由 `4` 为 `3`，使 TiCDC 更加平均地分发消息到各个 Kafka partition |
| TiDB Lightning | `meta-schema-name` | 修改 | 此配置项控制 TiDB Lightning 在目标 TiDB 中保存 metadata 对应的 schema name。从 v5.4.0 开始，只在开启了并行导入功能时（对应配置为 `tikv-importer.incremental-import = true` ），才会在目标 TiDB 中创建此库。 |

> | TiDB Lightning | `task-info-schema-name` |  新增  | 用于配置当 Lightning 检测到冲突数据时，对应冲突数据存储的库名，默认值为 "lightning_task_info"。如果没有开启冲突检测功能，则无需配置此参数。  |

Author: ```suggestion
| TiDB Lightning | `task-info-schema-name` |  新增  | 用于配置当 TiDB Lightning 检测到冲突数据时，对应冲突数据存储的库名，默认值为 "lightning_task_info"。如果没有开启冲突检测功能，则无需配置此参数。  |
```

在使用 GBK 字符集时，需要注意兼容性限制，详情参考[字符集和排序 - GBK](/character-set-gbk.md)。

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

    [用户文档](/tispark-overview.md#安全)

### 性能

- **持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能**

    - 支持将更多函数下推至 MPP 引擎
        - 字符串函数：`LPAD()`、`RPAD()`、`STRCMP()`
        - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
    - 引入动态线程池，提升资源利用率（实验特性）
    - 提升从 TiKV 同步数据时，由行存格式到列存格式的数据转换效率，整体的数据同步性能提升 50%
    - 调整一些配置项的默认值，提升 TiFlash 的性能和稳定性。HTAP 混合负载下，单表简单查询的性能最高提升 20%

    用户文档：[TiFlash 支持的计算下推](/tiflash/use-tiflash.md#tiflash-支持的计算下推)，[TiFlash 配置文件](/tiflash/tiflash-configuration.md#配置文件-tiflashtoml)

- **通过 session 变量实现有界限过期数据读取**

    TiDB 是基于 Raft 协议的多副本分布式数据库。面对高并发，高吞吐业务场景，可以通过 follower 节点实现读性能扩展，构建读写分离架构。

    针对不同的业务场景，follower 提供强一致读和弱一致过期读两种读模式。强一致读模式可以满足数据实时性要求严格的业务场景。然而，当采用该模式时，特别是在跨机房架构下，由于会出现 leader 和 follower 的数据同步延迟、吞吐量降低等情况，会出现延迟问题。

>     在对数据实时性要求不高的业务场景下，可以选择过期读模式。使用该模式可以降低延迟和提升吞吐。TiDB 目前支持通过以下方式实现过期读：使用 SQL 语句读取一个基于历史时间的数据和开启基于历史时间的只读事务。通过这两种方式，你均可以从指定时间点或时间范围内读取对应的历史数据。具体用法，请参考[使用 `AS OF TIMESTAMP` 语法读取历史数据](/as-of-timestamp.md)。。

Reviewer 19: ```suggestion
    在对数据实时性要求不高的业务场景下，可以选择过期读模式。使用该模式可以降低延迟和提升吞吐。TiDB 目前支持通过以下方式实现过期读：使用 SQL 语句读取一个基于历史时间的数据和开启基于历史时间的只读事务。通过这两种方式，你均可以从指定时间点或时间范围内读取对应的历史数据。具体用法，请参考[使用 `AS OF TIMESTAMP` 语法读取历史数据](/as-of-timestamp.md)。
```

- 自 v5.4.0 起，对于通过 Plan Cache 已经缓存的执行计划，如果为其创建绑定 (Binding)，会使得对应查询已经缓存的计划失效。v5.4.0 前已经缓存的计划不受新 Binding 的影响。
- 在 v5.3 及更早版本中，[TiDB Data Migration (DM)](https://docs.pingcap.com/zh/tidb-data-migration/v5.3/) 文档独立于 TiDB 文档。自 v5.4 起，TiDB Data Migration 的文档已合并入相同版本的 TiDB 文档，无需跳转到 DM 文档站，你可以直接在 TiDB 文档站阅读 [DM 文档](/dm/dm-overview.md)。

## 新功能

### SQL

- **TiDB 从 v5.4.0 起支持 GBK 字符集**

    在 v5.4.0 前，TiDB 支持 `ascii`、`binary`、`latin1`、`utf8` 和 `utf8mb4` 字符集。

    为了更好的支持中文用户，TiDB 从 v5.4.0 起支持 GBK 字符集。在初次初始化 TiDB 集群时开启 TiDB 配置项 [`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) 后，TiDB GBK 字符集同时支持 `gbk_bin` 和 `gbk_chinese_ci` 这两种排序规则。

在使用 GBK 字符集时，需要注意兼容性限制，详情参考[字符集和排序 - GBK](/character-set-gbk.md)。

### 安全

- **TiSpark 支持用户认证与鉴权**

    TiSpark 提供数据库和表级别的读写授权验证以及数据库用户认证验证。开启该功能后，能避免业务侧未经授权运行抽数等批量任务获取数据，提高线上集群的稳定性和数据安全性。从 TiSpark v2.5.0 起开始支持。

    该功能默认关闭。开启后，如果用户没有对应的权限，通过 TiSpark 操作会抛出对应的异常。

    [用户文档](/tispark-overview.md#安全)

- **TiUP 部署方式支持为 root 用户生成初始密码**

    集群启动命令增加了 `--init` 参数，有了该参数，在 TiUP 部署场景，TiUP 会为数据库 root 用户生成一个初始的强密码，避免 root 用户使用空密码所带来的安全风险，增强数据库的安全性。

>     [User document](/production-deployment-using-tiup.md##第-7-步启动集群)

Reviewer 13: ```suggestion
    [User document](/production-deployment-using-tiup.md#第-7-步启动集群)
```

        - 日期时间函数：`ADDDATE()`、`DATE_ADD()`、`DATE_SUB()`、`SUBDATE()`、`QUARTER()`
    - 引入动态线程池，提升资源利用率（实验特性）
    - 提升从 TiKV 同步数据时，由行存格式到列存格式的数据转换效率，整体的数据同步性能提升 50%
    - 调整一些配置项的默认值，提升 TiFlash 的性能和稳定性。HTAP 混合负载下，单表简单查询的性能最高提升 20%

    用户文档：[TiFlash 支持的计算下推](/tiflash/use-tiflash.md#tiflash-支持的计算下推)，[TiFlash 配置文件](/tiflash/tiflash-configuration.md#配置文件-tiflashtoml)

- **通过 session 变量实现有界限过期数据读取**

    TiDB 是基于 Raft 协议的多副本分布式数据库。面对高并发，高吞吐业务场景，可以通过 follower 节点实现读性能扩展，构建读写分离架构。

    针对不同的业务场景，follower 提供强一致读和弱一致过期读两种读模式。强一致读模式可以满足数据实时性要求严格的业务场景。然而，当采用该模式时，特别是在跨机房架构下，由于会出现 leader 和 follower 的数据同步延迟、吞吐量降低等情况，会出现延迟问题。

>     在对数据实时性要求不高的业务场景下，可以选择过期读模式。使用该模式可以降低延迟和提升吞吐。TiDB 目前支持通过显示只读事务或 SQL 语句的方式实现过期读。通过这两种方式，你均可以从指定时间点或时间范围内读取对应的历史数据。具体用法，请参考[通过系统变量 `tidb_snapshot` 读取历史数据](/read-historical-data.md)。

Reviewer 19: ```suggestion
    在对数据实时性要求不高的业务场景下，可以选择过期读模式。使用该模式可以降低延迟和提升吞吐。TiDB 目前支持通过以下方式实现过期读：使用 SQL 语句读取一个基于历史时间的数据和开启基于历史时间的只读事务。通过这两种方式，你均可以从指定时间点或时间范围内读取对应的历史数据。具体用法，请参考[使用 `AS OF TIMESTAMP` 语法读取历史数据](/as-of-timestamp.md)。。
```

    - 修复 INDEX HASH JOIN 报 `send on closed channel` 的问题 [#31129](https://github.com/pingcap/tidb/issues/31129)

+ TiKV

    - 修复 MVCC 删除记录可能不会被 GC 删除的问题 [#11217](https://github.com/tikv/tikv/issues/11217)
    - 修复悲观事务中 prewrite 请求重试在极少数情况下影响数据一致性的风险 [#11187](https://github.com/tikv/tikv/issues/11187)
    - 修复 GC 扫描导致的内存溢出 [#11410](https://github.com/tikv/tikv/issues/11410)
    - 修复当达到磁盘容量满时 RocksDB flush 或 compaction 导致的 panic [#11224](https://github.com/tikv/tikv/issues/11224)

+ PD

    - 修复 Region 统计不受 `flow-round-by-digit` 影响的问题 [#4295](https://github.com/tikv/pd/issues/4295)
    - 修复调度 Operator 因为目标 Store 处于 Down 的状态而无法快速失败的问题 [#3353](https://github.com/tikv/pd/issues/3353)
    - 修复不能 Merge 在 Offline Store 上面的 Region 的问题 [#4119](https://github.com/tikv/pd/issues/4119)
    - 修复热点统计中无法剔除冷热点数据的问题 [#4390](https://github.com/tikv/pd/issues/4390)

+ TiFlash

    - 修复查询报错 `Tree struct based executor must have executor id`
    - 修复查询报错 `Illegal column type UInt32 for the second argument of function tidbRoundWithFrac`
    - 修复查询报错 `Unexpected type of column: Nullable(Nothing)`

>     - 修复查询报错 `source region at right may be considered at left when merging`

Author: ```suggestion
```

    - 修复 `collation` 对 `nullable` 类型失效的问题
    - 修复 `coalesce` 错误地从结果列中删除 `nullable` 标志的问题
    - 修复当 MPP 查询被终止时，TiFlash 偶发的崩溃问题
    - 修复 `where <string>` 查询结果出错的问题
    - 修复整数类型主键的列类型设置为较大范围后数据可能不一致的问题
    - 修复输入早于 1970-01-01 00:00:01 UTC 时，`unix_timestamp` 行为与 TiDB/MySQL 不一致的问题
    - 修复 TiFlash 重启时偶发的 `EstablishMPPConnection` 错误
    - 修复在 TiFlash 与 TiDB/TiKV 之间 `CastStringAsDecimal` 行为不一致的问题
    - 修复查询报错 `DB::Exception: Encode type of coprocessor response is not CHBlock`

>     - 修复在 TiFlash 与 TiDB/TiKV 之间 `castStringAsReal` 行为不一致的问题。

Reviewer 12: ```suggestion
    - 修复在 TiFlash 与 TiDB/TiKV 之间 `castStringAsReal` 行为不一致的问题
```

---
title: TiDB 5.4 Release Notes
---

# TiDB 5.4 Release Notes

发版日期：2022 年 x 月 x 日

TiDB 版本：5.4.0

在 v5.4.0 版本中，你可以获得以下关键特性：

+ 支持 GBK 字符集
+ 支持索引合并 (Index Merge) 数据访问方法，能够合并多个列上索引的条件过滤结果
+ 支持通过 session 变量实现有界限过期数据读取
+ 支持统计信息采集配置持久化
+ 支持使用 Raft Engine 作为 TiKV 的日志存储引擎
+ 优化备份对集群的影响
+ 支持 Azure Blob Storage 作为备份目标存储
+ 持续提升 TiFlash 列式存储引擎和 MPP 计算引擎的稳定性和性能
+ 为 TiDB Lightning 增加已存在数据表是否允许导入的开关
+ 优化持续性能分析实验特性
+ TiSpark 支持用户认证与鉴权

## 兼容性变化

> **注意：**
>
> 当从一个早期的 TiDB 版本升级到 TiDB v5.4.0 时，如需了解所有中间版本对应的兼容性更改说明，请查看对应版本的 [Release Notes](/releases/release-notes.md)。

### 系统变量

|  变量名    |  修改类型    |  描述    |
| :---------- | :----------- | :----------- |

> |  [`tidb_backoff_lock_fast`](/system-variables.md#tidb_backoff_lock_fast) | 修改 | 默认值由 `100` 修改为 `10` |
> |  [`tidb_enable_column_tracking`](/system-variables.md#tidb_enable_column_tracking-从-v540-版本开始引入) | 新增 | 用于控制是否开启 TiDB 对 `PREDICATE COLUMNS` 的收集，默认值为 `OFF` |
> | [`tidb_enable_index_merge`](/system-variables.md#tidb_enable_index_merge-从-v40-版本开始引入) | 修改 | 默认值由 `OFF` 改为 `ON`。如果从低于 v4.0.0 版本升级到 v5.4.0 及以上版本的集群，该变量值默认保持 `OFF`。如果从 v4.0.0 及以上版本升级到 v5.4.0 及以上版本的集群，该变量开关保持升级前的状态。对于 v5.4.0 及以上版本的新建集群，该变量开关默认保持 `ON`。 |
> | [`tidb_enable_paging`](/system-variables.md#tidb_enable_paging-从-v540-版本开始引入)  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能会出现读请求的延迟高、TiKV 的 Unified read pool CPU 使用率高的情况。在这种情况下，由于 `Limit` 算子只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，从而降低延迟、减少资源消耗。 |
> | [`tidb_enable_top_sql`](/system-variables.md#tidb_enable_top_sql-从-v540-开始引入) | 新增 | 用于控制是否开启 Top SQL 特性，默认值为 OFF |
> | [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启 [ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性，默认值为 `OFF` |
> | [`tidb_read_staleness`](/system-variables.md#tidb_read_staleness-从-v540-版本开始引入) | 新增 | 用于设置当前会话允许读取的历史数据范围，默认值为 `0` |
> | [`tidb_regard_null_as_point`](/system-variables.md#tidb_regard_null_as_point-从-v540-版本开始引入) | 新增 | 用于控制优化器是否可以把 null 值当做点值并作为前缀条件来访问索引 |
> | [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `0` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时 |
> | [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败 (`OFF`) 还是退回使用 pseudo 的统计信息 (`ON`)，默认值为 `ON` |
> | [`tidb_store_limit`](/system-variables.md#tidb_store_limit-从-v304-和-v40-版本开始引入) | 修改 | v5.4.0 前支持实例级别及集群级别的设置，现在只支持集群级别的设置 |

Reviewer 19: ```suggestion
|  [`tidb_enable_column_tracking`](/system-variables.md#tidb_enable_column_tracking-从-v540-版本开始引入) | 新增 | 用于控制是否开启 TiDB 对 `PREDICATE COLUMNS` 的收集，默认值为 `OFF` |
| [`tidb_enable_paging`](/system-variables.md#tidb_enable_paging-从-v540-版本开始引入)  | 新增 | 此变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求，默认值为 `OFF`。对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能会出现读请求的延迟高、TiKV 的 Unified read pool CPU 使用率高的情况。在这种情况下，由于 `Limit` 算子只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，从而降低延迟、减少资源消耗。 |
| [`tidb_enable_top_sql`](/system-variables.md#tidb_enable_top_sql-从-v540-开始引入) | 新增 | 用于控制是否开启 Top SQL 特性，默认值为 OFF |
| [`tidb_persist_analyze_options`](/system-variables.md#tidb_persist_analyze_options-从-v540-版本开始引入)  | 新增  | 用于控制是否开启 [ANALYZE 配置持久化](/statistics.md#analyze-配置持久化)特性，默认值为 `OFF` |
| [`tidb_read_staleness`](/system-variables.md#tidb_read_staleness-从-v540-版本开始引入) | 新增 | 用于设置当前会话允许读取的历史数据范围，默认值为 `0` |
| [`tidb_regard_null_as_point`](/system-variables.md#tidb_regard_null_as_point-从-v540-版本开始引入) | 新增 | 用于控制优化器是否可以把 null 值当做点值并作为前缀条件来访问索引 |
| [`tidb_stats_load_sync_wait`](/system-variables.md#tidb_stats_load_sync_wait-从-v540-版本开始引入) | 新增 | 这个变量用于控制是否开启统计信息的同步加载模式（默认为 `0` 代表不开启，即为异步加载模式），以及开启的情况下，SQL 执行同步加载完整统计信息等待多久后会超时 |
| [`tidb_stats_load_pseudo_timeout`](/system-variables.md#tidb_stats_load_pseudo_timeout-从-v540-版本开始引入) | 新增 | 用于控制统计信息同步加载超时后，SQL 是执行失败 (`OFF`) 还是退回使用 pseudo 的统计信息 (`ON`)，默认值为 `ON` |
|  [`tidb_backoff_lock_fast`](/system-variables.md#tidb_backoff_lock_fast) | 修改 | 默认值由 `100` 修改为 `10` |
| [`tidb_enable_index_merge`](/system-variables.md#tidb_enable_index_merge-从-v40-版本开始引入) | 修改 | 默认值由 `OFF` 改为 `ON`。如果从低于 v4.0.0 版本升级到 v5.4.0 及以上版本的集群，该变量值默认保持 `OFF`。如果从 v4.0.0 及以上版本升级到 v5.4.0 及以上版本的集群，该变量开关保持升级前的状态。对于 v5.4.0 及以上版本的新建集群，该变量开关默认保持 `ON`。 |
| [`tidb_store_limit`](/system-variables.md#tidb_store_limit-从-v304-和-v40-版本开始引入) | 修改 | v5.4.0 前支持实例级别及集群级别的设置，现在只支持集群级别的设置 |
```

    - 如果对排序规则要求不严格，允许上下游查询结果排序规则不一致，使用默认的 loose 模式可以避免报错。
    - 如果对排序规则要求严格，业务要求排序规则必须一致，则应当使用 strict 模式。但如果下游不支持上游缺省的 collation，同步可能会报错。

    [用户文档](/dm/task-configuration-file-full.md#完整配置文件示例)

- **在 DM 中优化 `transfer source`，支持平滑执行同步任务**

    当 DM-worker 所在各节点负载不均衡时，`transfer source` 命令可用于手动将某 `source` 配置迁移到其他节点。优化后的 `transfer source` 简化了用户操作步骤，不再要求先暂停所有关联 task 而是直接执行平滑迁移，DM 将在内部完成所需操作。

- **DM OpenAPI 特性 GA**

    DM 支持通过 API 的方式进行日常管理，包括增加数据源、管理任务等。本次更新 OpenAPI 从实验特性转为正式特性。

    [用户文档](/dm/dm-open-api.md)

### 问题诊断效率

- **Top SQL（实验特性）**

    新推出实验性特性 Top SQL（默认关闭），帮助用户轻松找到节点中消耗负载较大的查询。

    [用户文档](/dashboard/top-sql.md)

### TiDB 数据共享订阅

- **优化 TiCDC 对集群的影响**

    大幅降低了 TiCDC 启用后，对 TiDB 集群的性能影响。在实验室环境中，TiCDC 对 TiDB 的性能影响可以降低到 5% 以下。

### 部署及运维

- **增强持续性能分析（实验特性）**

    - 支持更多组件：除了 TiDB、PD 和 TiKV 外，v5.4.0 版本中还支持查看 TiFlash CPU Profiling。
    - 支持更方便的查看形式：支持以火焰图形式查看 CPU Profiling 和 Goroutine 结果。
    - 支持更多部署环境：支持在 TiDB Operator 部署环境下启用持续性能分析功能。

    该功能默认关闭，需进入 TiDB Dashboard 持续性能分析页面开启。

    持续性能分析仅支持由 v1.9.0 及以上版本 TiUP 或 v1.3.0 及以上版本 TiDB Operator 升级或安装的集群。

    [用户文档](/dashboard/continuous-profiling.md)

## 提升改进

+ TiDB

>     - 新增系统变量 `tidb_enable_paging`，开启该功能可显著降低使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 数据较小且无法下推到 `IndexScan` 上的读请求的延迟 [#30578](https://github.com/pingcap/tidb/issues/30578)

Reviewer 12: 这句话看不懂。建议优化下。

Author: ```suggestion
    - 新增系统变量 `tidb_enable_paging`，用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求。开启该功能可能够减少处理数据的数量，从而降低延迟、减少资源消耗  [#30578](https://github.com/pingcap/tidb/issues/30578)
```

    - Coprocessor 支持下推 substring 到 TiKV [#11495](https://github.com/tikv/tikv/issues/11495)
    - 通过跳过读锁提高在 RC 隔离级别中扫描的性能 [#11485](https://github.com/tikv/tikv/issues/11485)
    - 减少备份使用的默认线程池大小，并在压力大时限制其使用 [#11000](https://github.com/tikv/tikv/issues/11000)
    - 支持动态调整 Apply 和 Store 线程池大小 [#11159](https://github.com/tikv/tikv/issues/11159)
    - 支持配置 `snap-generator` 线程池大小 [#11247](https://github.com/tikv/tikv/issues/11247)
    - 优化在文件数较多且读写频繁的场景下 RocksDB 的全局锁争用问题 [#250](https://github.com/tikv/rocksdb/pull/250)

+ PD

    - 默认开启历史热点记录功能 [#25281](https://github.com/pingcap/tidb/issues/25281)
    - 新增 HTTP Component 的签名，用于标识请求来源 [#4490](https://github.com/tikv/pd/issues/4490)
    - TiDB Dashboard 更新至 v2021.12.31 [#4257](https://github.com/tikv/pd/issues/4257)

+ TiFlash

    - 优化本地算子通讯
    - 调高 gRPC 非临时线程数，避免频繁创建/销毁线程

+ Tools

    + Backup & Restore (BR)

        - 增加 BR 加密备份时，对密钥的合法性检查 [#29794](https://github.com/pingcap/tidb/issues/29794)

    + TiCDC

        - 减少 "EventFeed retry rate limited" 日志的数量 [#4006](https://github.com/pingcap/tiflow/issues/4006)
        - 降低在同步大量表时的同步延时 [#3900](https://github.com/pingcap/tiflow/issues/3900)

>         - 减少 TiKV 节点宕机后 kv client 恢复的时间 [#3191](https://github.com/pingcap/tiflow/issues/3191)

Reviewer 12: ```suggestion
        - 减少 TiKV 节点宕机后 KV client 恢复的时间 [#3191](https://github.com/pingcap/tiflow/issues/3191)
```

    + TiDB Data Migration (DM)

        - 降低开启 relay 时的 CPU 使用率 [#2214](https://github.com/pingcap/dm/issues/2214)

    + TiDB Lightning

        - 在 TiDB-backend 模式下，默认改用乐观事务进行写入来提升性能 [#30953](https://github.com/pingcap/tidb/pull/30953)

    + Dumpling

        - 提升 Dumpling 检查数据库版本时的兼容性 [#29500](https://github.com/pingcap/tidb/pull/29500)
        - 在导出 `CREATE DATABASE` 和 `CREATE TABLE` 时添加默认的 collation [#3420](https://github.com/pingcap/tiflow/issues/3420)

>     + TiDB Binlog

Reviewer 12: ```suggestion
```

    - 修复当 `IN` 的值中带有二进制常量时查询结果错误的问题 [#31261](https://github.com/pingcap/tidb/issues/31261)
    - 修复当 CTE 中带有子查询时查询结果错误的问题 [#31255](https://github.com/pingcap/tidb/issues/31255)
    - 修复 `INSERT ... SELECT ... ON DUPLICATE KEY UPDATE` 语句 panic 的问题 [#28078](https://github.com/pingcap/tidb/issues/28078)
    - 修复 INDEX HASH JOIN 报 `send on closed channel` 的问题 [#31129](https://github.com/pingcap/tidb/issues/31129)

+ TiKV

    - 修复 MVCC 删除记录可能不会被 GC 删除的问题 [#11217](https://github.com/tikv/tikv/issues/11217)
    - 修复悲观事务中 prewrite 请求重试在极少数情况下影响数据一致性的风险 [#11187](https://github.com/tikv/tikv/issues/11187)
    - 修复 GC 扫描导致的内存溢出 [#11410](https://github.com/tikv/tikv/issues/11410)
    - 修复当达到磁盘容量满时 RocksDB flush 或 compaction 导致的 panic [#11224](https://github.com/tikv/tikv/issues/11224)

+ PD

    - 修复 Region 统计不受 `flow-round-by-digit` 影响的问题 [#4295](https://github.com/tikv/pd/issues/4295)
    - 修复调度 Operator 因为目标 Store 处于 Down 的状态而无法快速失败的问题 [#3353](https://github.com/tikv/pd/issues/3353)
    - 修复不能 Merge 在 Offline Store 上面的 Region 的问题 [#4119](https://github.com/tikv/pd/issues/4119)
    - 修复热点统计中无法剔除冷热点数据的问题 [#4390](https://github.com/tikv/pd/issues/4390)

+ TiFlash

>     - 修复查询报错 `Tree struct based executor must have executor id`

Reviewer 32: 这几个修复查询报错的，能否改成实际触发的问题？这个条目描述似乎对读者没有太大意义。

Author: 确认这几条都不写进 release notes 中

>     - 修复查询报错 `Illegal column type UInt32 for the second argument of function tidbRoundWithFrac`

Reviewer 33: - 修复查询报错 `Tree struct based executor must have executor id`
- 修复查询报错 `Illegal column type UInt32 for the second argument of function tidbRoundWithFrac`
- 修复查询报错 `Unexpected type of column: Nullable(Nothing)`
- 修复 `collation` 对 `nullable` 类型失效的问题
- 修复 `coalesce` 错误地从结果列中删除 `nullable` 标志的问题

这几个都已经同步到 5.3 分支，不需要写了

- 修复查询报错 `source region at right may be considered at left when merging`
这个不严重，不需要体现在 release-note中

Author: Deleted
